{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# MDAnalysis Tutorial "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Preparations"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Loading MDAnalysis"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import MDAnalysis\n",
    "from MDAnalysis.tests.datafiles import PSF, DCD\n",
    "\n",
    "import os\n",
    "%matplotlib inline\n",
    "import matplotlib.pyplot as plt"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'0.15.0'"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "MDAnalysis.__version__"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Basics"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Universe and AtomGroup"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "MDAnalysis is object oriented. Molecular systems consist of Atom objects (instances of the class MDAnalysis.core.AtomGroup.Atom), which are grouped in AtomGroup instances. You build the AtomGroup of your system by loading a topology (list of atoms and possibly their connectivity) together with a trajectory (coordinate information) into the central data structure, the Universe object:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<Universe with 3341 atoms and 3365 bonds>\n"
     ]
    }
   ],
   "source": [
    "u = MDAnalysis.Universe(PSF, DCD)\n",
    "print(u)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The atoms are stored in the attribute atoms of the MDAnalysis.core.AtomGroup.Universe:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "<AtomGroup with 3341 atoms>\n"
     ]
    }
   ],
   "source": [
    "print(u.atoms)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[<Atom 1: N of type 56 of resname MET, resid 1 and segid 4AKE>,\n",
       " <Atom 2: HT1 of type 2 of resname MET, resid 1 and segid 4AKE>,\n",
       " <Atom 3: HT2 of type 2 of resname MET, resid 1 and segid 4AKE>,\n",
       " <Atom 4: HT3 of type 2 of resname MET, resid 1 and segid 4AKE>,\n",
       " <Atom 5: CA of type 22 of resname MET, resid 1 and segid 4AKE>]"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(u.atoms[:5])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Any AtomGroup knows the residues that the atoms belong to via the attribute residues, which produces a ResidueGroup. A ResidueGroup acts like a list of Residue objects:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/anaconda2-4.1.1/lib/python2.7/site-packages/MDAnalysis/core/AtomGroup.py:1135: DeprecationWarning: In version 0.16.0 this will select residue names, not atom names \n",
      "  \"residue names, not atom names \", DeprecationWarning)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "<ResidueGroup [<Residue LEU, 6>, <Residue GLY, 7>, <Residue ALA, 8>]>"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "u.atoms[100:130].residues"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on AtomGroup in module MDAnalysis.core.AtomGroup object:\n",
      "\n",
      "class AtomGroup(__builtin__.object)\n",
      " |  A group of atoms.\n",
      " |  \n",
      " |    ag = universe.select_atoms(atom-list)\n",
      " |  \n",
      " |  The AtomGroup contains a list of atoms; typically, a AtomGroup is generated\n",
      " |  from a selection. It is build from any list-like collection of\n",
      " |  :class:`Atom` instances. It is also possible to create an empty AtomGroup\n",
      " |  from an empty list.\n",
      " |  \n",
      " |  An AtomGroup can be indexed and sliced like a list::\n",
      " |  \n",
      " |     ag[0], ag[-1]\n",
      " |  \n",
      " |  will return the first and the last :class:`Atom` in the group\n",
      " |  whereas the slice ::\n",
      " |  \n",
      " |     ag[0:6:2]\n",
      " |  \n",
      " |  returns every second element, corresponding to indices 0, 2, and 4.\n",
      " |  \n",
      " |  It also supports \"advanced slicing\" when the argument is a\n",
      " |  :class:`numpy.ndarray` or a :class:`list`::\n",
      " |  \n",
      " |     aslice = [0, 3, -1, 10, 3]\n",
      " |     ag[aslice]\n",
      " |  \n",
      " |  will return a new :class:`AtomGroup` containing (ag[0], ag[3], ag[-1],\n",
      " |  ag[10], ag[3]).\n",
      " |  \n",
      " |  .. Note::\n",
      " |  \n",
      " |     AtomGroups originating from a selection are sorted and\n",
      " |     duplicate elements are removed. This is not true for AtomGroups\n",
      " |     produced by slicing. Thus slicing can be used when the order of\n",
      " |     atoms is crucial (for instance, in order to define angles or\n",
      " |     dihedrals).\n",
      " |  \n",
      " |  Atoms can also be accessed in a Pythonic fashion by using the atom name as\n",
      " |  an attribute. For instance, ::\n",
      " |  \n",
      " |     ag.CA\n",
      " |  \n",
      " |  will provide a :class:`AtomGroup` of all CA atoms in the\n",
      " |  group. These *instant selector* attributes are auto-generated for\n",
      " |  each atom name encountered in the group.\n",
      " |  \n",
      " |  .. Note::\n",
      " |  \n",
      " |     The name-attribute instant selector access to atoms is mainly\n",
      " |     meant for quick interactive work. Thus it either returns a\n",
      " |     single :class:`Atom` if there is only one matching atom, *or* a\n",
      " |     new :class:`AtomGroup` for multiple matches.  This makes it\n",
      " |     difficult to use the feature consistently in scripts but it is\n",
      " |     much better for interactive work.\n",
      " |  \n",
      " |  AtomGroup instances are bound to a :class:`Universe`, but only through the\n",
      " |  weak reference :class:`Atom` has. The connection is lost as soon as the\n",
      " |  :class:`Universe` goes out of scope.\n",
      " |  \n",
      " |  AtomGroups can also be pickled and unpickled. When pickling, the Atom\n",
      " |  indices are serialized alongside the Universe number-of-atoms, and topology\n",
      " |  and trajectory filenames. When unpickling, all built Universes will be\n",
      " |  searched for one with matching number-of-atoms and filenames. Finer control\n",
      " |  over which :class:`Universe` gets chosen to base the unpickling on can\n",
      " |  be exerted using the *is_anchor* and *anchor_name* flags upon Universe creation\n",
      " |  or the methods/attributes :meth:`Universe.make_anchor`, :meth:`Universe.remove_anchor`,\n",
      " |  and :attr:`Universe.anchor_name`.\n",
      " |  \n",
      " |  \n",
      " |  .. rubric:: References for analysis methods\n",
      " |  \n",
      " |  .. [Dima2004] Dima, R. I., & Thirumalai, D. (2004). Asymmetry in the\n",
      " |                shapes of folded and denatured states of proteins. *J\n",
      " |                Phys Chem B*, 108(21),\n",
      " |                6564-6570. doi:`10.1021/jp037128y`_\n",
      " |  \n",
      " |  .. _10.1021/jp037128y: http://dx.doi.org/10.1021/jp037128y\n",
      " |  \n",
      " |  \n",
      " |  .. SeeAlso:: :class:`Universe`\n",
      " |  .. versionchanged:: 0.7.6\n",
      " |     An empty AtomGroup can be created and no longer raises a\n",
      " |     :exc:`NoDataError`.\n",
      " |  .. versionchanged:: 0.9.0\n",
      " |     The size at which cache is used for atom lookup is now stored as variable\n",
      " |     _atomcache_size within the class.\n",
      " |     Added fragments manged property. Is a lazily built, cached entry, similar to residues.\n",
      " |  .. versionchanged:: 0.11.0\n",
      " |     AtomGroups can now be pickled and unpickled provided compatible Universes\n",
      " |     are available.\n",
      " |     The follow methods were changed to properties: ``indices``, ``masses``, ``charges``, ``names``,\n",
      " |     ``types``, ``radii``, ``resids``, ``resnames``, ``resnums``, ``segids``.\n",
      " |     Added ``altLocs`` and ``serials`` properties and setters.\n",
      " |     ``Torsions`` and ``torsion`` renamed to ``dihedral``.\n",
      " |     The ``bond``, ``angle``, ``dihedral`` and ``improper`` methods were removed and replaced\n",
      " |     with properties of the same name which return the corresponding object.\n",
      " |     Deprecated ``selectAtoms`` in favour of ``select_atoms``.\n",
      " |     Setters are now plural to match property names.\n",
      " |     Properties referring to residue (``resids``, ``resnames``, ``resnums``)\n",
      " |     or segment [``segids``] properties now yield arrays of length equal to\n",
      " |     ``n_atoms``\n",
      " |  .. versionchanged:: 0.11.1\n",
      " |     Added occupancies property\n",
      " |  \n",
      " |  Methods defined here:\n",
      " |  \n",
      " |  __add__(self, other)\n",
      " |  \n",
      " |  __contains__(self, other)\n",
      " |  \n",
      " |  __getattr__(self, name)\n",
      " |  \n",
      " |  __getitem__(self, item)\n",
      " |      Return element (index) or group (slicing).\n",
      " |      \n",
      " |      .. versionchanged:: 0.8\n",
      " |         :class:`ResidueGroup` and :class:`SegmentGroup`:\n",
      " |         return groups themselves and allow advanced slicing\n",
      " |      .. versionchanged:: 0.9.0\n",
      " |         This method now used by all subclasses.  These subclasses override\n",
      " |         :attr:`_cls` to define the returned class.\n",
      " |      .. versionchanged:: 0.10.0\n",
      " |         Now supports indexing via boolean numpy arrays\n",
      " |  \n",
      " |  __getstate__(self)\n",
      " |  \n",
      " |  __init__(self, atoms)\n",
      " |  \n",
      " |  __iter__(self)\n",
      " |  \n",
      " |  __len__(self)\n",
      " |      Number of atoms in the group\n",
      " |  \n",
      " |  __repr__(self)\n",
      " |  \n",
      " |  __setstate__(self, state)\n",
      " |  \n",
      " |  align_principalAxis(*args, **kwds)\n",
      " |      `align_principalAxis` is deprecated, use `align_principal_axis` instead!\n",
      " |      This will be removed in version 0.16.0\n",
      " |      \n",
      " |      Align principal axis with index *axis* with *vector*.\n",
      " |      \n",
      " |              :Arguments:\n",
      " |                *axis*\n",
      " |                  Index of the principal axis (0, 1, or 2), as produced by\n",
      " |                  :meth:`~MDAnalysis.core.AtomGroup.AtomGroup.principal_axes`.\n",
      " |                *vector*\n",
      " |                  A 3D vector such as the z-axis (``[0,0,1]``); can be\n",
      " |                  anything that looks like a list with three entries.\n",
      " |      \n",
      " |              To align the long axis of a channel (the first principal axis,\n",
      " |              i.e. *axis* = 0) with the z-axis::\n",
      " |      \n",
      " |                u.atoms.align_principal_axis(0, [0,0,1])\n",
      " |                u.atoms.write(\"aligned.pdb\")\n",
      " |  \n",
      " |  align_principal_axis(self, axis, vector)\n",
      " |      Align principal axis with index *axis* with *vector*.\n",
      " |      \n",
      " |      :Arguments:\n",
      " |        *axis*\n",
      " |          Index of the principal axis (0, 1, or 2), as produced by\n",
      " |          :meth:`~MDAnalysis.core.AtomGroup.AtomGroup.principal_axes`.\n",
      " |        *vector*\n",
      " |          A 3D vector such as the z-axis (``[0,0,1]``); can be\n",
      " |          anything that looks like a list with three entries.\n",
      " |      \n",
      " |      To align the long axis of a channel (the first principal axis,\n",
      " |      i.e. *axis* = 0) with the z-axis::\n",
      " |      \n",
      " |        u.atoms.align_principal_axis(0, [0,0,1])\n",
      " |        u.atoms.write(\"aligned.pdb\")\n",
      " |  \n",
      " |  asphericity(self, **kwargs)\n",
      " |      Asphericity.\n",
      " |      \n",
      " |      See [Dima2004]_ for background information.\n",
      " |      \n",
      " |      :Keywords:\n",
      " |        *pbc*\n",
      " |          ``True``: Move all atoms within primary unit cell before calculation [``False``]\n",
      " |      \n",
      " |      .. Note::\n",
      " |          The :class:`MDAnalysis.core.flags` flag *use_pbc* when set to ``True`` allows the *pbc*\n",
      " |          flag to be used by default.\n",
      " |      \n",
      " |      .. versionadded:: 0.7.7\n",
      " |      .. versionchanged:: 0.8 Added *pbc* keyword\n",
      " |  \n",
      " |  bbox(self, **kwargs)\n",
      " |      Return the bounding box of the selection.\n",
      " |      \n",
      " |      The lengths A,B,C of the orthorhombic enclosing box are ::\n",
      " |      \n",
      " |        L = AtomGroup.bbox()\n",
      " |        A,B,C = L[1] - L[0]\n",
      " |      \n",
      " |      :Keywords:\n",
      " |        *pbc*\n",
      " |          ``True``: Move all atoms within the primary unit cell before calculation [``False``]\n",
      " |      \n",
      " |      .. Note::\n",
      " |          The :class:`MDAnalysis.core.flags` flag *use_pbc* when set to ``True`` allows the *pbc*\n",
      " |          flag to be used by default.\n",
      " |      \n",
      " |      :Returns: [[xmin, ymin, zmin], [xmax, ymax, zmax]]\n",
      " |      \n",
      " |      .. versionadded:: 0.7.2\n",
      " |      .. versionchanged:: 0.8 Added *pbc* keyword\n",
      " |  \n",
      " |  bsphere(self, **kwargs)\n",
      " |      Return the bounding sphere of the selection.\n",
      " |      \n",
      " |      The sphere is calculated relative to the centre of geometry.\n",
      " |      \n",
      " |      :Keywords:\n",
      " |        *pbc*\n",
      " |          ``True``: Move all atoms within primary unit cell before calculation [``False``]\n",
      " |      \n",
      " |      .. Note::\n",
      " |          The :class:`MDAnalysis.core.flags` flag *use_pbc* when set to ``True`` allows the *pbc*\n",
      " |          flag to be used by default.\n",
      " |      \n",
      " |      :Returns: `(R, [xcen,ycen,zcen])`\n",
      " |      \n",
      " |      .. versionadded:: 0.7.3\n",
      " |      .. versionchanged:: 0.8 Added *pbc* keyword\n",
      " |  \n",
      " |  centerOfGeometry(*args, **kwds)\n",
      " |      `centerOfGeometry` is deprecated, use `center_of_geometry` instead!\n",
      " |      This will be removed in version 0.16.0\n",
      " |      \n",
      " |      Center of geometry (also known as centroid) of the selection.\n",
      " |      \n",
      " |              :Keywords:\n",
      " |                *pbc*\n",
      " |                  ``True``: Move all atoms within the primary unit cell before calculation [``False``]\n",
      " |      \n",
      " |              .. Note::\n",
      " |                  The :class:`MDAnalysis.core.flags` flag *use_pbc* when set to ``True`` allows the *pbc*\n",
      " |                  flag to be used by default.\n",
      " |      \n",
      " |              .. versionchanged:: 0.8 Added *pbc* keyword\n",
      " |  \n",
      " |  centerOfMass(*args, **kwds)\n",
      " |      `centerOfMass` is deprecated, use `center_of_mass` instead!\n",
      " |      This will be removed in version 0.16.0\n",
      " |      \n",
      " |      Center of mass of the selection.\n",
      " |      \n",
      " |              :Keywords:\n",
      " |                *pbc*\n",
      " |                  ``True``: Move all atoms within the primary unit cell before calculation [``False``]\n",
      " |      \n",
      " |              .. Note::\n",
      " |                  The :class:`MDAnalysis.core.flags` flag *use_pbc* when set to ``True`` allows the *pbc*\n",
      " |                  flag to be used by default.\n",
      " |      \n",
      " |              .. versionchanged:: 0.8 Added *pbc* keyword\n",
      " |  \n",
      " |  center_of_geometry(self, **kwargs)\n",
      " |      Center of geometry (also known as centroid) of the selection.\n",
      " |      \n",
      " |      :Keywords:\n",
      " |        *pbc*\n",
      " |          ``True``: Move all atoms within the primary unit cell before calculation [``False``]\n",
      " |      \n",
      " |      .. Note::\n",
      " |          The :class:`MDAnalysis.core.flags` flag *use_pbc* when set to ``True`` allows the *pbc*\n",
      " |          flag to be used by default.\n",
      " |      \n",
      " |      .. versionchanged:: 0.8 Added *pbc* keyword\n",
      " |  \n",
      " |  center_of_mass(self, **kwargs)\n",
      " |      Center of mass of the selection.\n",
      " |      \n",
      " |      :Keywords:\n",
      " |        *pbc*\n",
      " |          ``True``: Move all atoms within the primary unit cell before calculation [``False``]\n",
      " |      \n",
      " |      .. Note::\n",
      " |          The :class:`MDAnalysis.core.flags` flag *use_pbc* when set to ``True`` allows the *pbc*\n",
      " |          flag to be used by default.\n",
      " |      \n",
      " |      .. versionchanged:: 0.8 Added *pbc* keyword\n",
      " |  \n",
      " |  centroid = center_of_geometry(self, **kwargs)\n",
      " |  \n",
      " |  coordinates(*args, **kwds)\n",
      " |      `coordinates` is deprecated!\n",
      " |      This will be removed in version 0.16.0; use `positions` property instead\n",
      " |      \n",
      " |      Np array of the coordinates.\n",
      " |      \n",
      " |              .. SeeAlso:: :attr:`~AtomGroup.positions` and :meth:`~AtomGroup.get_positions`\n",
      " |      \n",
      " |              .. deprecated:: 0.7.6\n",
      " |              In new scripts use :meth:`AtomGroup.get_positions` preferrably.\n",
      " |  \n",
      " |  get_forces(*args, **kwds)\n",
      " |      `get_forces` is deprecated!\n",
      " |      This will be removed in version 0.16.0; use `forces` property instead\n",
      " |      \n",
      " |      \n",
      " |              Get a numpy array of the atomic forces (if available).\n",
      " |              Currently only supported for Gromacs .trr trajectories.\n",
      " |      \n",
      " |              :Keywords:\n",
      " |                 *ts*\n",
      " |                     If *ts* is provided then positions are read from that\n",
      " |                     :class:`~MDAnalysis.coordinates.base.Timestep` instead of\n",
      " |                     the one from the current trajectory belonging to this universe.\n",
      " |                     The *ts* is indexed with the indices returned by\n",
      " |                     :meth:`~AtomGroup.indices` and it is the user's responsibility\n",
      " |                     to provide a time step that has the appropriate dimensions.\n",
      " |                 *copy*\n",
      " |                     ``True``: always make a copy (slow), ``False``: Try to\n",
      " |                     return a array view or reference (faster); note that for\n",
      " |                     passing coordinates to C-code it can be necessary to use\n",
      " |                     a copy [``False``]\n",
      " |                 *dtype*\n",
      " |                     numpy data type of the array; the default is usually\n",
      " |                     entirely appropriate. Most C-code actually requires the\n",
      " |                     default  [:class:`np.float32`]\n",
      " |      \n",
      " |              Forces can also be directly obtained from the attribute\n",
      " |              :attr:`~AtomGroup.forces`.\n",
      " |      \n",
      " |              Forces can be directly set with :meth:`~AtomGroup.set_forces` or\n",
      " |              by assigning to :attr:`~AtomGroup.forces`.\n",
      " |      \n",
      " |              .. versionadded:: 0.7.7\n",
      " |  \n",
      " |  get_positions(*args, **kwds)\n",
      " |      `get_positions` is deprecated!\n",
      " |      This will be removed in version 0.16.0; use `positions` property instead\n",
      " |      \n",
      " |      Get a numpy array of the coordinates.\n",
      " |      \n",
      " |              :Keywords:\n",
      " |                 *ts*\n",
      " |                     If *ts* is provided then positions are read from that\n",
      " |                     :class:`~MDAnalysis.coordinates.base.Timestep` instead of\n",
      " |                     the one from the current trajectory belonging to this universe.\n",
      " |                     The *ts* is indexed with the indices returned by\n",
      " |                     :meth:`~AtomGroup.indices` and it is the user's responsibility\n",
      " |                     to provide a time step that has the appropriate dimensions.\n",
      " |                 *copy*\n",
      " |                     ``True``: always make a copy (slow), ``False``: Try to\n",
      " |                     return a array view or reference (faster); note that for\n",
      " |                     passing coordinates to C-code it can be necessary to use\n",
      " |                     a copy [``False``]\n",
      " |                 *dtype*\n",
      " |                     numpy Data type of the array; the default is usually\n",
      " |                     entirely appropriate. Most C-code actually requires the\n",
      " |                     default  [:class:`np.float32`]\n",
      " |      \n",
      " |              Coordinates can also be directly obtained from the attribute\n",
      " |              :attr:`~AtomGroup.positions`.\n",
      " |      \n",
      " |              Coordinates can be directly set with :meth:`~AtomGroup.set_positions` or\n",
      " |              by assigning to :attr:`~AtomGroup.positions`.\n",
      " |      \n",
      " |              This method is identical with :meth:`~AtomGroup.coordinates` but named\n",
      " |              differently for symmetry with with :meth:`~AtomGroup.set_positions`.\n",
      " |      \n",
      " |              .. versionadded:: 0.7.6\n",
      " |  \n",
      " |  get_velocities(*args, **kwds)\n",
      " |      `get_velocities` is deprecated!\n",
      " |      This will be removed in version 0.16.0; use `velocities` property instead\n",
      " |      \n",
      " |      numpy array of the velocities.\n",
      " |      \n",
      " |              Raises a :exc:`NoDataError` if the underlying\n",
      " |              :class:`~MDAnalysis.coordinates.base.Timestep` does not contain\n",
      " |              :attr:`~MDAnalysis.coordinates.base.Timestep.velocities`.\n",
      " |      \n",
      " |              See also :meth:`AtomGroup.set_velocities` and attribute access through\n",
      " |              :attr:`AtomGroup.velocities`.\n",
      " |      \n",
      " |              .. versionadded:: 0.7.6\n",
      " |  \n",
      " |  guess_bonds = outfunc(self, *args, **kwargs)\n",
      " |  \n",
      " |  momentOfInertia(*args, **kwds)\n",
      " |      `momentOfInertia` is deprecated, use `moment_of_inertia` instead!\n",
      " |      This will be removed in version 0.16.0\n",
      " |      \n",
      " |      Tensor of inertia as 3x3 numpy array.\n",
      " |      \n",
      " |              :Keywords:\n",
      " |                *pbc*\n",
      " |                  ``True``: Move all atoms within the primary unit cell before calculation [``False``]\n",
      " |      \n",
      " |              .. Note::\n",
      " |                  The :class:`MDAnalysis.core.flags` flag *use_pbc* when set to ``True`` allows the *pbc*\n",
      " |                  flag to be used by default.\n",
      " |      \n",
      " |              .. versionchanged:: 0.8 Added *pbc* keyword\n",
      " |  \n",
      " |  moment_of_inertia(self, **kwargs)\n",
      " |      Tensor of inertia as 3x3 numpy array.\n",
      " |      \n",
      " |      :Keywords:\n",
      " |        *pbc*\n",
      " |          ``True``: Move all atoms within the primary unit cell before calculation [``False``]\n",
      " |      \n",
      " |      .. Note::\n",
      " |          The :class:`MDAnalysis.core.flags` flag *use_pbc* when set to ``True`` allows the *pbc*\n",
      " |          flag to be used by default.\n",
      " |      \n",
      " |      .. versionchanged:: 0.8 Added *pbc* keyword\n",
      " |  \n",
      " |  packIntoBox(*args, **kwds)\n",
      " |      `packIntoBox` is deprecated, use `pack_into_box` instead!\n",
      " |      This will be removed in version 0.16.0\n",
      " |      \n",
      " |      Shift all atoms in this group to be within the primary unit cell.\n",
      " |      \n",
      " |              AtomGroup.packintobox([box, [inplace=True]])\n",
      " |      \n",
      " |              :Keywords:\n",
      " |                *box*\n",
      " |                  Unit cell to move atoms inside of.\n",
      " |                *inplace*\n",
      " |                  ``True``: Change coordinates in place and return\n",
      " |                  ``False``: Only return the coordinates\n",
      " |      \n",
      " |              All atoms will be moved so that they lie between 0 and\n",
      " |              boxlength :math:`L_i` in all dimensions, i.e. the lower left\n",
      " |              corner of the simulation box is taken to be at (0,0,0):\n",
      " |      \n",
      " |              .. math::\n",
      " |      \n",
      " |                 x_i' = x_i - \\left\\lfloor\\frac{x_i}{L_i}\\right\\rfloor\n",
      " |      \n",
      " |              The default is to take unit cell information from the\n",
      " |              underlying :class:`~MDAnalysis.coordinates.base.Timestep`\n",
      " |              instance. The optional argument *box* can be used to provide\n",
      " |              alternative unit cell information (in the MDAnalysis standard\n",
      " |              format ``[Lx, Ly, Lz, alpha, beta, gamma]``).\n",
      " |      \n",
      " |              Works with either orthogonal or triclinic box types.\n",
      " |      \n",
      " |              By default the coordinates are changed in place and returned\n",
      " |      \n",
      " |              .. versionadded:: 0.8\n",
      " |  \n",
      " |  pack_into_box(self, box=None, inplace=True)\n",
      " |      Shift all atoms in this group to be within the primary unit cell.\n",
      " |      \n",
      " |      AtomGroup.packintobox([box, [inplace=True]])\n",
      " |      \n",
      " |      :Keywords:\n",
      " |        *box*\n",
      " |          Unit cell to move atoms inside of.\n",
      " |        *inplace*\n",
      " |          ``True``: Change coordinates in place and return\n",
      " |          ``False``: Only return the coordinates\n",
      " |      \n",
      " |      All atoms will be moved so that they lie between 0 and\n",
      " |      boxlength :math:`L_i` in all dimensions, i.e. the lower left\n",
      " |      corner of the simulation box is taken to be at (0,0,0):\n",
      " |      \n",
      " |      .. math::\n",
      " |      \n",
      " |         x_i' = x_i - \\left\\lfloor\\frac{x_i}{L_i}\\right\\rfloor\n",
      " |      \n",
      " |      The default is to take unit cell information from the\n",
      " |      underlying :class:`~MDAnalysis.coordinates.base.Timestep`\n",
      " |      instance. The optional argument *box* can be used to provide\n",
      " |      alternative unit cell information (in the MDAnalysis standard\n",
      " |      format ``[Lx, Ly, Lz, alpha, beta, gamma]``).\n",
      " |      \n",
      " |      Works with either orthogonal or triclinic box types.\n",
      " |      \n",
      " |      By default the coordinates are changed in place and returned\n",
      " |      \n",
      " |      .. versionadded:: 0.8\n",
      " |  \n",
      " |  principalAxes(*args, **kwds)\n",
      " |      `principalAxes` is deprecated, use `principal_axes` instead!\n",
      " |      This will be removed in version 0.16.0\n",
      " |      \n",
      " |      Calculate the principal axes from the moment of inertia.\n",
      " |      \n",
      " |              e1,e2,e3 = AtomGroup.principal_axes()\n",
      " |      \n",
      " |              The eigenvectors are sorted by eigenvalue, i.e. the first one\n",
      " |              corresponds to the highest eigenvalue and is thus the first principal axes.\n",
      " |      \n",
      " |              :Keywords:\n",
      " |                *pbc*\n",
      " |                  ``True``: Move all atoms within primary unit cell before calculation\n",
      " |      \n",
      " |              .. Note::\n",
      " |                  The :class:`MDAnalysis.core.flags` flag *use_pbc* when set to ``True`` allows the *pbc*\n",
      " |                  flag to be used by default.\n",
      " |      \n",
      " |              :Returns: numpy.array ``v`` with ``v[0]`` as first, ``v[1]`` as second,\n",
      " |                        and ``v[2]`` as third eigenvector.\n",
      " |      \n",
      " |              .. versionchanged:: 0.8 Added *pbc* keyword\n",
      " |  \n",
      " |  principal_axes(self, **kwargs)\n",
      " |      Calculate the principal axes from the moment of inertia.\n",
      " |      \n",
      " |      e1,e2,e3 = AtomGroup.principal_axes()\n",
      " |      \n",
      " |      The eigenvectors are sorted by eigenvalue, i.e. the first one\n",
      " |      corresponds to the highest eigenvalue and is thus the first principal axes.\n",
      " |      \n",
      " |      :Keywords:\n",
      " |        *pbc*\n",
      " |          ``True``: Move all atoms within primary unit cell before calculation\n",
      " |      \n",
      " |      .. Note::\n",
      " |          The :class:`MDAnalysis.core.flags` flag *use_pbc* when set to ``True`` allows the *pbc*\n",
      " |          flag to be used by default.\n",
      " |      \n",
      " |      :Returns: numpy.array ``v`` with ``v[0]`` as first, ``v[1]`` as second,\n",
      " |                and ``v[2]`` as third eigenvector.\n",
      " |      \n",
      " |      .. versionchanged:: 0.8 Added *pbc* keyword\n",
      " |  \n",
      " |  radiusOfGyration(*args, **kwds)\n",
      " |      `radiusOfGyration` is deprecated, use `radius_of_gyration` instead!\n",
      " |      This will be removed in version 0.16.0\n",
      " |      \n",
      " |      Radius of gyration.\n",
      " |      \n",
      " |              :Keywords:\n",
      " |                *pbc*\n",
      " |                  ``True``: Move all atoms within the primary unit cell before calculation [``False``]\n",
      " |      \n",
      " |              .. Note::\n",
      " |                  The :class:`MDAnalysis.core.flags` flag *use_pbc* when set to ``True`` allows the *pbc*\n",
      " |                  flag to be used by default.\n",
      " |      \n",
      " |              .. versionchanged:: 0.8 Added *pbc* keyword\n",
      " |  \n",
      " |  radius_of_gyration(self, **kwargs)\n",
      " |      Radius of gyration.\n",
      " |      \n",
      " |      :Keywords:\n",
      " |        *pbc*\n",
      " |          ``True``: Move all atoms within the primary unit cell before calculation [``False``]\n",
      " |      \n",
      " |      .. Note::\n",
      " |          The :class:`MDAnalysis.core.flags` flag *use_pbc* when set to ``True`` allows the *pbc*\n",
      " |          flag to be used by default.\n",
      " |      \n",
      " |      .. versionchanged:: 0.8 Added *pbc* keyword\n",
      " |  \n",
      " |  rotate(self, R)\n",
      " |      Apply a rotation matrix *R* to the selection's coordinates.\n",
      " |      \n",
      " |      AtomGroup.rotate(R)\n",
      " |      \n",
      " |      :math:`\\mathsf{R}` is a 3x3 orthogonal matrix that transforms a vector\n",
      " |      :math:`\\mathbf{x} \\rightarrow \\mathbf{x}'`:\n",
      " |      \n",
      " |      .. math::\n",
      " |      \n",
      " |          \\mathbf{x}' = \\mathsf{R}\\mathbf{x}\n",
      " |  \n",
      " |  rotateby(self, angle, axis, point=None)\n",
      " |      Apply a rotation to the selection's coordinates.\n",
      " |      \n",
      " |      AtomGroup.rotateby(angle,axis[,point])\n",
      " |      \n",
      " |      The transformation from current coordinates :math:`\\mathbf{x}`\n",
      " |      to new coordinates :math:`\\mathbf{x}'` is\n",
      " |      \n",
      " |      .. math::\n",
      " |      \n",
      " |        \\mathbf{x}' = \\mathsf{R}\\,(\\mathbf{x}-\\mathbf{p}) + \\mathbf{p}\n",
      " |      \n",
      " |      where :math:`\\mathsf{R}` is the rotation by *angle* around the\n",
      " |      *axis* going through *point* :math:`\\mathbf{p}`.\n",
      " |      \n",
      " |      :Arguments:\n",
      " |        *angle*\n",
      " |           rotation angle in degrees\n",
      " |        *axis*\n",
      " |           rotation axis vector, a 3-tuple, list, or array, or a 2-tuple of\n",
      " |           two MDAnalysis objects from which the axis is calculated as the\n",
      " |           vector from the first to the second center of geometry.\n",
      " |        *point*\n",
      " |           point on the rotation axis; by default (``None``) the center of\n",
      " |           geometry of the selection is chosen, or, if *axis* is a tuple of\n",
      " |           selections, it defaults to the first point of the axis. *point*\n",
      " |           can be a 3-tuple, list, or array or a MDAnalysis object (in which\n",
      " |           case its :meth:`centroid` is used).\n",
      " |      \n",
      " |      :Returns: The 4x4 matrix which consists of the rotation matrix ``M[:3,:3]``\n",
      " |                and the translation vector ``M[:3,3]``.\n",
      " |  \n",
      " |  selectAtoms(*args, **kwds)\n",
      " |      `selectAtoms` is deprecated, use `select_atoms` instead!\n",
      " |      This will be removed in version 0.16.0\n",
      " |      \n",
      " |      Select atoms using a MDAnalysis selection string.\n",
      " |      \n",
      " |              Parameters\n",
      " |              ----------\n",
      " |              sel : str\n",
      " |                  The selection string to be parsed (following the language below)\n",
      " |              *othersel : str, optional\n",
      " |                  More selection strings to be parsed\n",
      " |              **selgroups : AtomGroup(s), optional\n",
      " |                  groups to be referred to using the 'group' keyword\n",
      " |      \n",
      " |              Returns\n",
      " |              -------\n",
      " |              A :class:`AtomGroup`, with atoms sorted according to their\n",
      " |              index.  This AtomGroup will not contain any duplicate atoms,\n",
      " |              even if the original selection had so.\n",
      " |      \n",
      " |              Note\n",
      " |              ----\n",
      " |              If exact ordering of atoms is required (for instance, for\n",
      " |              :meth:`~AtomGroup.angle` or :meth:`~AtomGroup.dihedral` calculations)\n",
      " |              then one supplies selections *separately* in the required order.\n",
      " |              For example::\n",
      " |                 ag.select_atoms('name HT*', 'name N')\n",
      " |              Gives anything matching HT* before anything matching N\n",
      " |      \n",
      " |              Examples\n",
      " |              --------\n",
      " |              The selection parser understands the following *keywords*:\n",
      " |      \n",
      " |              Simple field selections\n",
      " |              -----------------------\n",
      " |              The following selections allow implicit OR of any number of values after\n",
      " |              their keyword. For example::\n",
      " |                  'name HT1 HT2' == 'name HT1 or name HT2'\n",
      " |              They also allow the use of the wildcard character '*', for example::\n",
      " |                  'name HT*'\n",
      " |              Selects all atoms whose name starts with 'HT'\n",
      " |      \n",
      " |              name *atom-name*\n",
      " |                  select by atom name (as given in the topology). Often, this is\n",
      " |                  force field dependent.\n",
      " |                  e.g. ``name CA`` (for C alpha atoms)\n",
      " |                  or ``name OW`` (for SPC water oxygen)\n",
      " |              type *atom-type*\n",
      " |                  select by atom type; this is either a string or a number and\n",
      " |                  depends on the force field; it is read from the topology file\n",
      " |                  (e.g. the CHARMM PSF file contains numeric atom types).\n",
      " |              resname *residue-name*\n",
      " |                  select by residue name, e.g. ``resname LYS``\n",
      " |              segid *seg-name*\n",
      " |                  select by segid (as given in the topology),\n",
      " |                  e.g. ``segid 4AKE`` or ``segid DMPC``\n",
      " |              altloc *alternative-location*\n",
      " |                  a selection for atoms where alternative locations are available,\n",
      " |                  which is often the case with high-resolution crystal structures\n",
      " |                  e.g. `resid 4 and resname ALA and altloc B` selects only the\n",
      " |                  atoms of ALA-4 that have an altloc B record.\n",
      " |      \n",
      " |              Numerical ranges\n",
      " |              ----------------\n",
      " |              These selections support implicit OR as above, and also selecting\n",
      " |              a range of numerical values using either ':' or '-'\n",
      " |              For example::\n",
      " |                  'resid 1:10'\n",
      " |              Selects all resid from 1 to 10 inclusive\n",
      " |                  'resid 1 3 5 20-30'\n",
      " |              Selects 1, 3, 5, and 20 through 30\n",
      " |      \n",
      " |              resid *residue-number-range*\n",
      " |                  resid can take a single residue number or a range of numbers.\n",
      " |                  A residue number (\"resid\") is taken directly from the topology.\n",
      " |              resnum *resnum-number-range*\n",
      " |                  resnum is the canonical residue number; typically it is set to\n",
      " |                  the residue id in the original PDB structure.\n",
      " |              bynum *index-range*\n",
      " |                  selects all atoms within a range of (1-based) inclusive indices,\n",
      " |                  e.g. ``bynum 1`` selects the first atom in the universe;\n",
      " |                  ``bynum 5:10`` selects atoms 5 through 10 inclusive. All atoms\n",
      " |                  in the :class:`MDAnalysis.Universe` are consecutively numbered,\n",
      " |                  and the index runs from 1 up to the total number of atoms.\n",
      " |      \n",
      " |              Convienience selections\n",
      " |              -----------------------\n",
      " |              protein, backbone, nucleic, nucleicbackbone, nucleicbase, nucleicsugar\n",
      " |                  selects all atoms that belong to a standard set of residues;\n",
      " |                  a protein is identfied by a hard-coded set of residue names so\n",
      " |                  it  may not work for esoteric residues.\n",
      " |              atom *seg-name*  *residue-number*  *atom-name*\n",
      " |                  a selector for a single atom consisting of segid resid atomname,\n",
      " |                  e.g. ``DMPC 1 C2`` selects the C2 carbon of the first residue of\n",
      " |                  the DMPC segment\n",
      " |      \n",
      " |              Boolean modifiers\n",
      " |              -----------------\n",
      " |              not\n",
      " |                  all atoms not in the selection. e.g. ``not protein``\n",
      " |                  selects all atoms that aren't part of a protein\n",
      " |              and, or\n",
      " |                  combine two selections according to the rules of boolean\n",
      " |                  algebra, e.g. ``protein and not resname ALA LYS`` selects\n",
      " |                  all atoms that belong to a protein, but are not in a lysine\n",
      " |                  or alanine residue\n",
      " |              parentheses '(' ')'\n",
      " |                  Subselections can be grouped with parentheses to evaluate\n",
      " |                  this before other parts of the selection\n",
      " |      \n",
      " |              Geometric\n",
      " |              ---------\n",
      " |              around *distance*  *selection*\n",
      " |                  selects all atoms a certain cutoff away from another selection,\n",
      " |                  e.g. ``around 3.5 protein`` selects all atoms not belonging to\n",
      " |                  protein that are within 3.5 Angstroms from the protein\n",
      " |              point *x* *y* *z*  *distance*\n",
      " |                  selects all atoms within a cutoff of a point in space, make sure\n",
      " |                  coordinate is separated by spaces,\n",
      " |                  e.g. ``point 5.0 5.0 5.0  3.5`` selects all atoms within 3.5\n",
      " |                  Angstroms of the coordinate (5.0, 5.0, 5.0)\n",
      " |              prop [abs] *property*  *operator*  *value*\n",
      " |                  selects atoms based on position, using *property*  **x**, **y**,\n",
      " |                  or **z** coordinate. Supports the **abs** keyword (for absolute\n",
      " |                  value) and the following *operators*: **<, >, <=, >=, ==, !=**.\n",
      " |                  For example, ``prop z >= 5.0`` selects all atoms with z\n",
      " |                  coordinate greater than 5.0; ``prop abs z <= 5.0`` selects all\n",
      " |                  atoms within -5.0 <= z <= 5.0.\n",
      " |              sphzone *radius* *selection*\n",
      " |                  Selects all atoms that are within *radius* of the center of\n",
      " |                  geometry of *selection*\n",
      " |              sphlayer *inner radius* *outer radius* *selection*\n",
      " |                  Similar to sphzone, but also excludes atoms that are within\n",
      " |                  *inner radius* of the selection COG\n",
      " |      \n",
      " |              Connectivity\n",
      " |              ------------\n",
      " |              bonded *selection*\n",
      " |                  selects all atoms that are bonded to selection\n",
      " |                  eg: ``select name H bonded name O`` selects only hydrogens\n",
      " |                  bonded to oxygens\n",
      " |              same *keyword* as *selection*\n",
      " |                  Selects all atoms that have the same property as those in\n",
      " |                  the other *selection*\n",
      " |                  *keyword* can be one of:\n",
      " |                      x, y, z (to match the position)\n",
      " |                      residue, fragment, segment\n",
      " |                      type, segid, mass, charge, radius, bfactor\n",
      " |                      resid, resnum, resname, segid\n",
      " |                  e.g. 'same mass as name CT1' would select all atoms that\n",
      " |                  had the same mass as those called CT1\n",
      " |              byres *selection*\n",
      " |                  selects all atoms that are in the same segment and residue as\n",
      " |                  selection,\n",
      " |                  e.g. specify the subselection after the byres keyword\n",
      " |      \n",
      " |              Preexisting selections\n",
      " |              ----------------------\n",
      " |              group *group-name*\n",
      " |                  selects the atoms in the :class:`AtomGroup` passed to the\n",
      " |                  function as an argument named *group-name*. Only the atoms\n",
      " |                  common to *group-name* and the instance :meth:`~select_atoms`\n",
      " |                  was called from will be considered. *group-name* will be\n",
      " |                  included in the parsing just by comparison of atom indices.\n",
      " |                  This means that it is up to the user to make sure they were\n",
      " |                  defined in an appropriate :class:`Universe`.\n",
      " |              fullgroup *group-name*\n",
      " |                  just like the ``group`` keyword with the difference that all the\n",
      " |                  atoms of *group-name* are included. The resulting selection may\n",
      " |                  therefore have atoms that were initially absent from the\n",
      " |                  instance :meth:`~select_atoms` was called from.\n",
      " |                  Deprecated in 0.11, use `global group` instead\n",
      " |      \n",
      " |              Global selections\n",
      " |              -----------------\n",
      " |              global *selection*\n",
      " |                  By default, select_atoms works on the AtomGroup it is called\n",
      " |                  from.  By placing the global prefix, the selection is instead\n",
      " |                  applied to the entire Universe the AtomGroup belongs to.\n",
      " |                  e.g.\n",
      " |                      lipids.select_atoms('around 10 global protein')\n",
      " |                  Selects all atoms from group `lipids` which are within 10 of\n",
      " |                  the selection u.select_atoms('protein')\n",
      " |                  Without the *global* modifier, the selection `protein` would\n",
      " |                  have only been applied to the lipids group itself.\n",
      " |      \n",
      " |              .. SeeAlso:: :ref:`selection-commands-label` for further details.\n",
      " |      \n",
      " |              .. versionchanged:: 0.7.4\n",
      " |                 Added *resnum* selection.\n",
      " |              .. versionchanged:: 0.8.1\n",
      " |                 Added *group* and *fullgroup* selections.\n",
      " |              .. versionchanged:: 0.13.0\n",
      " |                 Added *bonded* selection\n",
      " |              .. versionchanged:: 0.13.1\n",
      " |                 Added implicit OR syntax to field and range selections\n",
      " |  \n",
      " |  select_atoms(self, selstr, *othersel, **selgroups)\n",
      " |      Select atoms using a MDAnalysis selection string.\n",
      " |      \n",
      " |      Parameters\n",
      " |      ----------\n",
      " |      sel : str\n",
      " |          The selection string to be parsed (following the language below)\n",
      " |      *othersel : str, optional\n",
      " |          More selection strings to be parsed\n",
      " |      **selgroups : AtomGroup(s), optional\n",
      " |          groups to be referred to using the 'group' keyword\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      A :class:`AtomGroup`, with atoms sorted according to their\n",
      " |      index.  This AtomGroup will not contain any duplicate atoms,\n",
      " |      even if the original selection had so.\n",
      " |      \n",
      " |      Note\n",
      " |      ----\n",
      " |      If exact ordering of atoms is required (for instance, for\n",
      " |      :meth:`~AtomGroup.angle` or :meth:`~AtomGroup.dihedral` calculations)\n",
      " |      then one supplies selections *separately* in the required order.\n",
      " |      For example::\n",
      " |         ag.select_atoms('name HT*', 'name N')\n",
      " |      Gives anything matching HT* before anything matching N\n",
      " |      \n",
      " |      Examples\n",
      " |      --------\n",
      " |      The selection parser understands the following *keywords*:\n",
      " |      \n",
      " |      Simple field selections\n",
      " |      -----------------------\n",
      " |      The following selections allow implicit OR of any number of values after\n",
      " |      their keyword. For example::\n",
      " |          'name HT1 HT2' == 'name HT1 or name HT2'\n",
      " |      They also allow the use of the wildcard character '*', for example::\n",
      " |          'name HT*'\n",
      " |      Selects all atoms whose name starts with 'HT'\n",
      " |      \n",
      " |      name *atom-name*\n",
      " |          select by atom name (as given in the topology). Often, this is\n",
      " |          force field dependent.\n",
      " |          e.g. ``name CA`` (for C alpha atoms)\n",
      " |          or ``name OW`` (for SPC water oxygen)\n",
      " |      type *atom-type*\n",
      " |          select by atom type; this is either a string or a number and\n",
      " |          depends on the force field; it is read from the topology file\n",
      " |          (e.g. the CHARMM PSF file contains numeric atom types).\n",
      " |      resname *residue-name*\n",
      " |          select by residue name, e.g. ``resname LYS``\n",
      " |      segid *seg-name*\n",
      " |          select by segid (as given in the topology),\n",
      " |          e.g. ``segid 4AKE`` or ``segid DMPC``\n",
      " |      altloc *alternative-location*\n",
      " |          a selection for atoms where alternative locations are available,\n",
      " |          which is often the case with high-resolution crystal structures\n",
      " |          e.g. `resid 4 and resname ALA and altloc B` selects only the\n",
      " |          atoms of ALA-4 that have an altloc B record.\n",
      " |      \n",
      " |      Numerical ranges\n",
      " |      ----------------\n",
      " |      These selections support implicit OR as above, and also selecting\n",
      " |      a range of numerical values using either ':' or '-'\n",
      " |      For example::\n",
      " |          'resid 1:10'\n",
      " |      Selects all resid from 1 to 10 inclusive\n",
      " |          'resid 1 3 5 20-30'\n",
      " |      Selects 1, 3, 5, and 20 through 30\n",
      " |      \n",
      " |      resid *residue-number-range*\n",
      " |          resid can take a single residue number or a range of numbers.\n",
      " |          A residue number (\"resid\") is taken directly from the topology.\n",
      " |      resnum *resnum-number-range*\n",
      " |          resnum is the canonical residue number; typically it is set to\n",
      " |          the residue id in the original PDB structure.\n",
      " |      bynum *index-range*\n",
      " |          selects all atoms within a range of (1-based) inclusive indices,\n",
      " |          e.g. ``bynum 1`` selects the first atom in the universe;\n",
      " |          ``bynum 5:10`` selects atoms 5 through 10 inclusive. All atoms\n",
      " |          in the :class:`MDAnalysis.Universe` are consecutively numbered,\n",
      " |          and the index runs from 1 up to the total number of atoms.\n",
      " |      \n",
      " |      Convienience selections\n",
      " |      -----------------------\n",
      " |      protein, backbone, nucleic, nucleicbackbone, nucleicbase, nucleicsugar\n",
      " |          selects all atoms that belong to a standard set of residues;\n",
      " |          a protein is identfied by a hard-coded set of residue names so\n",
      " |          it  may not work for esoteric residues.\n",
      " |      atom *seg-name*  *residue-number*  *atom-name*\n",
      " |          a selector for a single atom consisting of segid resid atomname,\n",
      " |          e.g. ``DMPC 1 C2`` selects the C2 carbon of the first residue of\n",
      " |          the DMPC segment\n",
      " |      \n",
      " |      Boolean modifiers\n",
      " |      -----------------\n",
      " |      not\n",
      " |          all atoms not in the selection. e.g. ``not protein``\n",
      " |          selects all atoms that aren't part of a protein\n",
      " |      and, or\n",
      " |          combine two selections according to the rules of boolean\n",
      " |          algebra, e.g. ``protein and not resname ALA LYS`` selects\n",
      " |          all atoms that belong to a protein, but are not in a lysine\n",
      " |          or alanine residue\n",
      " |      parentheses '(' ')'\n",
      " |          Subselections can be grouped with parentheses to evaluate\n",
      " |          this before other parts of the selection\n",
      " |      \n",
      " |      Geometric\n",
      " |      ---------\n",
      " |      around *distance*  *selection*\n",
      " |          selects all atoms a certain cutoff away from another selection,\n",
      " |          e.g. ``around 3.5 protein`` selects all atoms not belonging to\n",
      " |          protein that are within 3.5 Angstroms from the protein\n",
      " |      point *x* *y* *z*  *distance*\n",
      " |          selects all atoms within a cutoff of a point in space, make sure\n",
      " |          coordinate is separated by spaces,\n",
      " |          e.g. ``point 5.0 5.0 5.0  3.5`` selects all atoms within 3.5\n",
      " |          Angstroms of the coordinate (5.0, 5.0, 5.0)\n",
      " |      prop [abs] *property*  *operator*  *value*\n",
      " |          selects atoms based on position, using *property*  **x**, **y**,\n",
      " |          or **z** coordinate. Supports the **abs** keyword (for absolute\n",
      " |          value) and the following *operators*: **<, >, <=, >=, ==, !=**.\n",
      " |          For example, ``prop z >= 5.0`` selects all atoms with z\n",
      " |          coordinate greater than 5.0; ``prop abs z <= 5.0`` selects all\n",
      " |          atoms within -5.0 <= z <= 5.0.\n",
      " |      sphzone *radius* *selection*\n",
      " |          Selects all atoms that are within *radius* of the center of\n",
      " |          geometry of *selection*\n",
      " |      sphlayer *inner radius* *outer radius* *selection*\n",
      " |          Similar to sphzone, but also excludes atoms that are within\n",
      " |          *inner radius* of the selection COG\n",
      " |      \n",
      " |      Connectivity\n",
      " |      ------------\n",
      " |      bonded *selection*\n",
      " |          selects all atoms that are bonded to selection\n",
      " |          eg: ``select name H bonded name O`` selects only hydrogens\n",
      " |          bonded to oxygens\n",
      " |      same *keyword* as *selection*\n",
      " |          Selects all atoms that have the same property as those in\n",
      " |          the other *selection*\n",
      " |          *keyword* can be one of:\n",
      " |              x, y, z (to match the position)\n",
      " |              residue, fragment, segment\n",
      " |              type, segid, mass, charge, radius, bfactor\n",
      " |              resid, resnum, resname, segid\n",
      " |          e.g. 'same mass as name CT1' would select all atoms that\n",
      " |          had the same mass as those called CT1\n",
      " |      byres *selection*\n",
      " |          selects all atoms that are in the same segment and residue as\n",
      " |          selection,\n",
      " |          e.g. specify the subselection after the byres keyword\n",
      " |      \n",
      " |      Preexisting selections\n",
      " |      ----------------------\n",
      " |      group *group-name*\n",
      " |          selects the atoms in the :class:`AtomGroup` passed to the\n",
      " |          function as an argument named *group-name*. Only the atoms\n",
      " |          common to *group-name* and the instance :meth:`~select_atoms`\n",
      " |          was called from will be considered. *group-name* will be\n",
      " |          included in the parsing just by comparison of atom indices.\n",
      " |          This means that it is up to the user to make sure they were\n",
      " |          defined in an appropriate :class:`Universe`.\n",
      " |      fullgroup *group-name*\n",
      " |          just like the ``group`` keyword with the difference that all the\n",
      " |          atoms of *group-name* are included. The resulting selection may\n",
      " |          therefore have atoms that were initially absent from the\n",
      " |          instance :meth:`~select_atoms` was called from.\n",
      " |          Deprecated in 0.11, use `global group` instead\n",
      " |      \n",
      " |      Global selections\n",
      " |      -----------------\n",
      " |      global *selection*\n",
      " |          By default, select_atoms works on the AtomGroup it is called\n",
      " |          from.  By placing the global prefix, the selection is instead\n",
      " |          applied to the entire Universe the AtomGroup belongs to.\n",
      " |          e.g.\n",
      " |              lipids.select_atoms('around 10 global protein')\n",
      " |          Selects all atoms from group `lipids` which are within 10 of\n",
      " |          the selection u.select_atoms('protein')\n",
      " |          Without the *global* modifier, the selection `protein` would\n",
      " |          have only been applied to the lipids group itself.\n",
      " |      \n",
      " |      .. SeeAlso:: :ref:`selection-commands-label` for further details.\n",
      " |      \n",
      " |      .. versionchanged:: 0.7.4\n",
      " |         Added *resnum* selection.\n",
      " |      .. versionchanged:: 0.8.1\n",
      " |         Added *group* and *fullgroup* selections.\n",
      " |      .. versionchanged:: 0.13.0\n",
      " |         Added *bonded* selection\n",
      " |      .. versionchanged:: 0.13.1\n",
      " |         Added implicit OR syntax to field and range selections\n",
      " |  \n",
      " |  sequence(self, **kwargs)\n",
      " |      Returns the amino acid sequence.\n",
      " |      \n",
      " |      The format of the sequence is selected with the keyword *format*:\n",
      " |      \n",
      " |      ============== ============================================\n",
      " |      *format*       description\n",
      " |      ============== ============================================\n",
      " |      'SeqRecord'    :class:`Bio.SeqRecord.SeqRecord` (default)\n",
      " |      'Seq'          :class:`Bio.Seq.Seq`\n",
      " |      'string'       string\n",
      " |      ============== ============================================\n",
      " |      \n",
      " |      The sequence is returned by default (keyword ``format = 'SeqRecord'``)\n",
      " |      as a :class:`Bio.SeqRecord.SeqRecord` instance, which can then be\n",
      " |      further processed. In this case, all keyword arguments (such as the\n",
      " |      *id* string or the *name* or the *description*) are directly passed to\n",
      " |      :class:`Bio.SeqRecord.SeqRecord`.\n",
      " |      \n",
      " |      If the keyword *format* is set to ``'Seq'``, all *kwargs* are ignored\n",
      " |      and a :class:`Bio.Seq.Seq` instance is returned. The difference to the\n",
      " |      record is that the record also contains metadata and can be directly\n",
      " |      used as an input for other functions in :mod:`Bio`.\n",
      " |      \n",
      " |      If the keyword *format* is set to ``'string'``, all *kwargs* are\n",
      " |      ignored and a Python string is returned.\n",
      " |      \n",
      " |      .. rubric:: Example: Write FASTA file\n",
      " |      \n",
      " |      Use :func:`Bio.SeqIO.write`, which takes sequence records::\n",
      " |      \n",
      " |         import Bio.SeqIO\n",
      " |      \n",
      " |         # get the sequence record of a protein component of a Universe\n",
      " |         protein = u.select_atoms(\"protein\")\n",
      " |         record = protein.sequence(id=\"myseq1\", name=\"myprotein\")\n",
      " |      \n",
      " |         Bio.SeqIO.write(record, \"single.fasta\", \"fasta\")\n",
      " |      \n",
      " |      A FASTA file with multiple entries can be written with ::\n",
      " |      \n",
      " |         Bio.SeqIO.write([record1, record2, ...], \"multi.fasta\", \"fasta\")\n",
      " |      \n",
      " |      :Keywords:\n",
      " |          *format*\n",
      " |      \n",
      " |              - ``\"string\"``: return sequence as a string of 1-letter codes\n",
      " |              - ``\"Seq\"``: return a :class:`Bio.Seq.Seq` instance\n",
      " |              - ``\"SeqRecord\"``: return a :class:`Bio.SeqRecord.SeqRecord`\n",
      " |                instance\n",
      " |      \n",
      " |              Default is ``\"SeqRecord\"``\n",
      " |      \n",
      " |           *id*\n",
      " |              Sequence ID for SeqRecord (should be different for different\n",
      " |              sequences)\n",
      " |           *name*\n",
      " |              Name of the protein.\n",
      " |           *description*\n",
      " |              Short description of the sequence.\n",
      " |           *kwargs*\n",
      " |              Any other keyword arguments that are understood by\n",
      " |              :class:`Bio.SeqRecord.SeqRecord`.\n",
      " |      \n",
      " |      :Raises: :exc:`ValueError` if a residue name cannot be converted to a\n",
      " |               1-letter IUPAC protein amino acid code; make sure to only\n",
      " |               select protein residues. Raises :exc:`TypeError` if an unknown\n",
      " |               *format* is selected.\n",
      " |      \n",
      " |      .. versionadded:: 0.9.0\n",
      " |  \n",
      " |  set = _set_atoms(self, name, value, **kwargs)\n",
      " |  \n",
      " |  set_altLoc(*args, **kwds)\n",
      " |      `set_altLoc` is deprecated, use `set_altLocs` instead!\n",
      " |      This will be removed in version 0.16.0\n",
      " |      \n",
      " |      `set_altLocs` is deprecated!\n",
      " |      This will be removed in version 0.16.0; use `altLocs` property instead\n",
      " |      \n",
      " |      Set the altLocs to *altLoc for **all atoms** in the AtomGroup.\n",
      " |      \n",
      " |              If *altLoc* is a sequence of the same length as the :class:`AtomGroup`\n",
      " |              then each :attr:`Atom.altLoc` is set to the corresponding value. If\n",
      " |              *value* is neither of length 1 (or a scalar) nor of the length of the\n",
      " |              :class:`AtomGroup` then a :exc:`ValueError` is raised.\n",
      " |      \n",
      " |              .. versionadded:: 0.11.0\n",
      " |  \n",
      " |  set_altLocs(*args, **kwds)\n",
      " |      `set_altLocs` is deprecated!\n",
      " |      This will be removed in version 0.16.0; use `altLocs` property instead\n",
      " |      \n",
      " |      Set the altLocs to *altLoc for **all atoms** in the AtomGroup.\n",
      " |      \n",
      " |              If *altLoc* is a sequence of the same length as the :class:`AtomGroup`\n",
      " |              then each :attr:`Atom.altLoc` is set to the corresponding value. If\n",
      " |              *value* is neither of length 1 (or a scalar) nor of the length of the\n",
      " |              :class:`AtomGroup` then a :exc:`ValueError` is raised.\n",
      " |      \n",
      " |              .. versionadded:: 0.11.0\n",
      " |  \n",
      " |  set_bfactor(*args, **kwds)\n",
      " |      `set_bfactor` is deprecated, use `set_bfactors` instead!\n",
      " |      This will be removed in version 0.16.0\n",
      " |      \n",
      " |      `set_bfactors` is deprecated!\n",
      " |      This will be removed in version 0.16.0; use `bfactors` property instead\n",
      " |      \n",
      " |      Set the atom bfactors to float *bfactor* for **all atoms** in the AtomGroup.\n",
      " |      \n",
      " |              If *bfactor* is a sequence of the same length as the :class:`AtomGroup`\n",
      " |              then each :attr:`Atom.bfactor` is set to the corresponding value. If\n",
      " |              *value* is neither of length 1 (or a scalar) nor of the length of the\n",
      " |              :class:`AtomGroup` then a :exc:`ValueError` is raised.\n",
      " |      \n",
      " |              .. versionadded:: 0.7.4\n",
      " |              .. versionchanged:: 0.8\n",
      " |                 Can set atoms and residues to distinct values by providing a sequence or iterable.\n",
      " |              .. versionchanged:: 0.11.0\n",
      " |                 Made plural to make consistent with corresponding property\n",
      " |  \n",
      " |  set_bfactors(*args, **kwds)\n",
      " |      `set_bfactors` is deprecated!\n",
      " |      This will be removed in version 0.16.0; use `bfactors` property instead\n",
      " |      \n",
      " |      Set the atom bfactors to float *bfactor* for **all atoms** in the AtomGroup.\n",
      " |      \n",
      " |              If *bfactor* is a sequence of the same length as the :class:`AtomGroup`\n",
      " |              then each :attr:`Atom.bfactor` is set to the corresponding value. If\n",
      " |              *value* is neither of length 1 (or a scalar) nor of the length of the\n",
      " |              :class:`AtomGroup` then a :exc:`ValueError` is raised.\n",
      " |      \n",
      " |              .. versionadded:: 0.7.4\n",
      " |              .. versionchanged:: 0.8\n",
      " |                 Can set atoms and residues to distinct values by providing a sequence or iterable.\n",
      " |              .. versionchanged:: 0.11.0\n",
      " |                 Made plural to make consistent with corresponding property\n",
      " |  \n",
      " |  set_charge(*args, **kwds)\n",
      " |      `set_charge` is deprecated, use `set_charges` instead!\n",
      " |      This will be removed in version 0.16.0\n",
      " |      \n",
      " |      `set_charges` is deprecated!\n",
      " |      This will be removed in version 0.16.0; use `charges` property instead\n",
      " |      \n",
      " |      Set the partial charges to float *charge* for **all atoms** in the AtomGroup.\n",
      " |      \n",
      " |              If *charge* is a sequence of the same length as the :class:`AtomGroup`\n",
      " |              then each :attr:`Atom.charge` is set to the corresponding value. If\n",
      " |              *value* is neither of length 1 (or a scalar) nor of the length of the\n",
      " |              :class:`AtomGroup` then a :exc:`ValueError` is raised.\n",
      " |      \n",
      " |              .. versionadded:: 0.7.4\n",
      " |              .. versionchanged:: 0.8\n",
      " |                 Can set atoms and residues to distinct values by providing a sequence or iterable.\n",
      " |              .. versionchanged:: 0.11.0\n",
      " |                 Made plural to make consistent with corresponding property\n",
      " |  \n",
      " |  set_charges(*args, **kwds)\n",
      " |      `set_charges` is deprecated!\n",
      " |      This will be removed in version 0.16.0; use `charges` property instead\n",
      " |      \n",
      " |      Set the partial charges to float *charge* for **all atoms** in the AtomGroup.\n",
      " |      \n",
      " |              If *charge* is a sequence of the same length as the :class:`AtomGroup`\n",
      " |              then each :attr:`Atom.charge` is set to the corresponding value. If\n",
      " |              *value* is neither of length 1 (or a scalar) nor of the length of the\n",
      " |              :class:`AtomGroup` then a :exc:`ValueError` is raised.\n",
      " |      \n",
      " |              .. versionadded:: 0.7.4\n",
      " |              .. versionchanged:: 0.8\n",
      " |                 Can set atoms and residues to distinct values by providing a sequence or iterable.\n",
      " |              .. versionchanged:: 0.11.0\n",
      " |                 Made plural to make consistent with corresponding property\n",
      " |  \n",
      " |  set_forces(*args, **kwds)\n",
      " |      `set_forces` is deprecated!\n",
      " |      This will be removed in version 0.16.0; use `forces` property instead\n",
      " |      \n",
      " |      Set the forces for all atoms in the group.\n",
      " |      \n",
      " |              :Arguments:\n",
      " |                 *forces*\n",
      " |                     a Nx3 numpy :class:`numpy.ndarray` where N is the number of\n",
      " |                     atoms in this atom group.\n",
      " |      \n",
      " |              :Keywords:\n",
      " |                 *ts*\n",
      " |                    :class:`~MDAnalysis.coordinates.base.Timestep`, defaults\n",
      " |                    to ``None`` and then the current time step is used.\n",
      " |      \n",
      " |              .. Note::\n",
      " |      \n",
      " |                 If the group contains N atoms and *force* is a single vector (i.e. an\n",
      " |                 array of length 3) then all N atom positions are set to *force* (due\n",
      " |                 to numpy's broadcasting rules), as described for\n",
      " |                 :attr:`~AtomGroup.forces`.\n",
      " |      \n",
      " |              See also :meth:`~AtomGroup.get_forces` and attribute access through\n",
      " |              :attr:`~AtomGroup.forces`.\n",
      " |      \n",
      " |              .. versionadded:: 0.7.7\n",
      " |  \n",
      " |  set_mass(*args, **kwds)\n",
      " |      `set_mass` is deprecated, use `set_masses` instead!\n",
      " |      This will be removed in version 0.16.0\n",
      " |      \n",
      " |      `set_masses` is deprecated!\n",
      " |      This will be removed in version 0.16.0; use `masses` property instead\n",
      " |      \n",
      " |      Set the atom masses to float *mass* for **all atoms** in the AtomGroup.\n",
      " |      \n",
      " |              If *mass* is a sequence of the same length as the :class:`AtomGroup`\n",
      " |              then each :attr:`Atom.mass` is set to the corresponding value. If\n",
      " |              *value* is neither of length 1 (or a scalar) nor of the length of the\n",
      " |              :class:`AtomGroup` then a :exc:`ValueError` is raised.\n",
      " |      \n",
      " |              .. versionadded:: 0.7.4\n",
      " |              .. versionchanged:: 0.8\n",
      " |                 Can set atoms and residues to distinct values by providing a sequence or iterable.\n",
      " |              .. versionchanged:: 0.11.0\n",
      " |                 Made plural to make consistent with corresponding property\n",
      " |  \n",
      " |  set_masses(*args, **kwds)\n",
      " |      `set_masses` is deprecated!\n",
      " |      This will be removed in version 0.16.0; use `masses` property instead\n",
      " |      \n",
      " |      Set the atom masses to float *mass* for **all atoms** in the AtomGroup.\n",
      " |      \n",
      " |              If *mass* is a sequence of the same length as the :class:`AtomGroup`\n",
      " |              then each :attr:`Atom.mass` is set to the corresponding value. If\n",
      " |              *value* is neither of length 1 (or a scalar) nor of the length of the\n",
      " |              :class:`AtomGroup` then a :exc:`ValueError` is raised.\n",
      " |      \n",
      " |              .. versionadded:: 0.7.4\n",
      " |              .. versionchanged:: 0.8\n",
      " |                 Can set atoms and residues to distinct values by providing a sequence or iterable.\n",
      " |              .. versionchanged:: 0.11.0\n",
      " |                 Made plural to make consistent with corresponding property\n",
      " |  \n",
      " |  set_name(*args, **kwds)\n",
      " |      `set_name` is deprecated, use `set_names` instead!\n",
      " |      This will be removed in version 0.16.0\n",
      " |      \n",
      " |      `set_names` is deprecated!\n",
      " |      This will be removed in version 0.16.0; use `names` property instead\n",
      " |      \n",
      " |      Set the atom names to string for *all atoms* in the AtomGroup.\n",
      " |      \n",
      " |              If *value* is a sequence of the same length as the :class:`AtomGroup`\n",
      " |              then each :attr:`Atom.name` is set to the corresponding value. If\n",
      " |              *value* is neither of length 1 (or a scalar) nor of the length of the\n",
      " |              :class:`AtomGroup` then a :exc:`ValueError` is raised.\n",
      " |      \n",
      " |              .. versionadded:: 0.7.4\n",
      " |              .. versionchanged:: 0.8\n",
      " |                 Can set atoms to distinct values by providing a sequence or iterable.\n",
      " |              .. versionchanged:: 0.11.0\n",
      " |                 Made plural to make consistent with corresponding property\n",
      " |  \n",
      " |  set_names(*args, **kwds)\n",
      " |      `set_names` is deprecated!\n",
      " |      This will be removed in version 0.16.0; use `names` property instead\n",
      " |      \n",
      " |      Set the atom names to string for *all atoms* in the AtomGroup.\n",
      " |      \n",
      " |              If *value* is a sequence of the same length as the :class:`AtomGroup`\n",
      " |              then each :attr:`Atom.name` is set to the corresponding value. If\n",
      " |              *value* is neither of length 1 (or a scalar) nor of the length of the\n",
      " |              :class:`AtomGroup` then a :exc:`ValueError` is raised.\n",
      " |      \n",
      " |              .. versionadded:: 0.7.4\n",
      " |              .. versionchanged:: 0.8\n",
      " |                 Can set atoms to distinct values by providing a sequence or iterable.\n",
      " |              .. versionchanged:: 0.11.0\n",
      " |                 Made plural to make consistent with corresponding property\n",
      " |  \n",
      " |  set_occupancies(self, occupancies)\n",
      " |      Set the occupancy for *all atoms* in the AtomGroup\n",
      " |      \n",
      " |      If *value* is a sequence of the same length as the :class:`AtomGroup`\n",
      " |      then each :attr:`Atom.name` is set to the corresponding value. If\n",
      " |      *value* is neither of length 1 (or a scalar) nor of the length of the\n",
      " |      :class:`AtomGroup` then a :exc:`ValueError` is raised.\n",
      " |      \n",
      " |      .. versionadded:: 0.11.1\n",
      " |  \n",
      " |  set_positions(*args, **kwds)\n",
      " |      `set_positions` is deprecated!\n",
      " |      This will be removed in version 0.16.0; use `positions` property instead\n",
      " |      \n",
      " |      Set the positions for all atoms in the group.\n",
      " |      \n",
      " |              :Arguments:\n",
      " |                 *coords*\n",
      " |                     a Nx3 :class:`numpy.ndarray` where N is the number of\n",
      " |                     atoms in this atom group.\n",
      " |      \n",
      " |              :Keywords:\n",
      " |                 *ts*\n",
      " |                    :class:`~MDAnalysis.coordinates.base.Timestep`, defaults\n",
      " |                    to ``None`` and then the current time step is used.\n",
      " |      \n",
      " |              .. Note::\n",
      " |      \n",
      " |                 If the group contains N atoms and *coord* is a single point (i.e. an\n",
      " |                 array of length 3) then all N atom positions are set to *coord* (due\n",
      " |                 to numpy's broadcasting rules), as described for\n",
      " |                 :attr:`~AtomGroup.positions`.\n",
      " |      \n",
      " |              See also :meth:`~AtomGroup.get_positions` and attribute access through\n",
      " |              :attr:`~AtomGroup.positions`.\n",
      " |      \n",
      " |              .. versionadded:: 0.7.6\n",
      " |  \n",
      " |  set_radii(*args, **kwds)\n",
      " |      `set_radii` is deprecated!\n",
      " |      This will be removed in version 0.16.0; use `radii` property instead\n",
      " |      \n",
      " |      Set the atom radii to float *radius* for **all atoms** in the AtomGroup.\n",
      " |      \n",
      " |              If *radius* is a sequence of the same length as the :class:`AtomGroup`\n",
      " |              then each :attr:`Atom.radius` is set to the corresponding value. If\n",
      " |              *value* is neither of length 1 (or a scalar) nor of the length of the\n",
      " |              :class:`AtomGroup` then a :exc:`ValueError` is raised.\n",
      " |      \n",
      " |              .. versionadded:: 0.7.4\n",
      " |              .. versionchanged:: 0.8\n",
      " |                 Can set atoms and residues to distinct values by providing a sequence or iterable.\n",
      " |              .. versionchanged:: 0.11.0\n",
      " |                 Made plural to make consistent with corresponding property\n",
      " |  \n",
      " |  set_radius(*args, **kwds)\n",
      " |      `set_radius` is deprecated, use `set_radii` instead!\n",
      " |      This will be removed in version 0.16.0\n",
      " |      \n",
      " |      `set_radii` is deprecated!\n",
      " |      This will be removed in version 0.16.0; use `radii` property instead\n",
      " |      \n",
      " |      Set the atom radii to float *radius* for **all atoms** in the AtomGroup.\n",
      " |      \n",
      " |              If *radius* is a sequence of the same length as the :class:`AtomGroup`\n",
      " |              then each :attr:`Atom.radius` is set to the corresponding value. If\n",
      " |              *value* is neither of length 1 (or a scalar) nor of the length of the\n",
      " |              :class:`AtomGroup` then a :exc:`ValueError` is raised.\n",
      " |      \n",
      " |              .. versionadded:: 0.7.4\n",
      " |              .. versionchanged:: 0.8\n",
      " |                 Can set atoms and residues to distinct values by providing a sequence or iterable.\n",
      " |              .. versionchanged:: 0.11.0\n",
      " |                 Made plural to make consistent with corresponding property\n",
      " |  \n",
      " |  set_resid(*args, **kwds)\n",
      " |      `set_resid` is deprecated, use `set_resids` instead!\n",
      " |      This will be removed in version 0.16.0\n",
      " |      \n",
      " |      `set_resids` is deprecated!\n",
      " |      This will be removed in version 0.16.0; use `resids` property instead\n",
      " |      \n",
      " |      Set the resids to integer *resid* for **all atoms** in the :class:`AtomGroup`.\n",
      " |      \n",
      " |              If *resid* is a sequence of the same length as the :class:`AtomGroup`\n",
      " |              then each :attr:`Atom.resid` is set to the corresponding value together\n",
      " |              with the :attr:`Residue.resid` of the residue the atom belongs to. If\n",
      " |              *value* is neither of length 1 (or a scalar) nor of the length of the\n",
      " |              :class:`AtomGroup` then a :exc:`ValueError` is raised.\n",
      " |      \n",
      " |              .. Note::\n",
      " |      \n",
      " |                 Changing resids can change the topology.\n",
      " |      \n",
      " |                 Assigning the same *resid* to multiple residues will\n",
      " |                 **merge** these residues. Assigning different *resid* to\n",
      " |                 atoms in the same residue will **split** a residue (and\n",
      " |                 potentially merge with another one).\n",
      " |      \n",
      " |              .. versionadded:: 0.7.4\n",
      " |              .. versionchanged:: 0.7.5\n",
      " |                 Also changes the residues.\n",
      " |              .. versionchanged:: 0.8\n",
      " |                 Can set atoms and residues to distinct values by providing a\n",
      " |                 sequence or iterable and can change the topology via\n",
      " |                 :func:`MDAnalysis.topology.core.build_residues`.\n",
      " |              .. versionchanged:: 0.11.0\n",
      " |                 Made plural to make consistent with corresponding property\n",
      " |  \n",
      " |  set_resids(*args, **kwds)\n",
      " |      `set_resids` is deprecated!\n",
      " |      This will be removed in version 0.16.0; use `resids` property instead\n",
      " |      \n",
      " |      Set the resids to integer *resid* for **all atoms** in the :class:`AtomGroup`.\n",
      " |      \n",
      " |              If *resid* is a sequence of the same length as the :class:`AtomGroup`\n",
      " |              then each :attr:`Atom.resid` is set to the corresponding value together\n",
      " |              with the :attr:`Residue.resid` of the residue the atom belongs to. If\n",
      " |              *value* is neither of length 1 (or a scalar) nor of the length of the\n",
      " |              :class:`AtomGroup` then a :exc:`ValueError` is raised.\n",
      " |      \n",
      " |              .. Note::\n",
      " |      \n",
      " |                 Changing resids can change the topology.\n",
      " |      \n",
      " |                 Assigning the same *resid* to multiple residues will\n",
      " |                 **merge** these residues. Assigning different *resid* to\n",
      " |                 atoms in the same residue will **split** a residue (and\n",
      " |                 potentially merge with another one).\n",
      " |      \n",
      " |              .. versionadded:: 0.7.4\n",
      " |              .. versionchanged:: 0.7.5\n",
      " |                 Also changes the residues.\n",
      " |              .. versionchanged:: 0.8\n",
      " |                 Can set atoms and residues to distinct values by providing a\n",
      " |                 sequence or iterable and can change the topology via\n",
      " |                 :func:`MDAnalysis.topology.core.build_residues`.\n",
      " |              .. versionchanged:: 0.11.0\n",
      " |                 Made plural to make consistent with corresponding property\n",
      " |  \n",
      " |  set_resname(*args, **kwds)\n",
      " |      `set_resname` is deprecated, use `set_resnames` instead!\n",
      " |      This will be removed in version 0.16.0\n",
      " |      \n",
      " |      `set_resnames` is deprecated!\n",
      " |      This will be removed in version 0.16.0; use `resnames` property instead\n",
      " |      \n",
      " |      Set the resnames to string *resname* for **all atoms** in the :class:`AtomGroup`.\n",
      " |      \n",
      " |              If *resname* is a sequence of the same length as the :class:`AtomGroup`\n",
      " |              then each :attr:`Atom.resname` is set to the corresponding value together\n",
      " |              with the :attr:`Residue.resname` of the residue the atom belongs to. If\n",
      " |              *value* is neither of length 1 (or a scalar) nor of the length of the\n",
      " |              :class:`AtomGroup` then a :exc:`ValueError` is raised.\n",
      " |      \n",
      " |              .. versionadded:: 0.7.4\n",
      " |              .. versionchanged:: 0.7.5\n",
      " |                 Also changes the residues.\n",
      " |              .. versionchanged:: 0.8\n",
      " |                 Can set atoms and residues to distinct values by providing a sequence or iterable.\n",
      " |              .. versionchanged:: 0.11.0\n",
      " |                 Made plural to make consistent with corresponding property\n",
      " |  \n",
      " |  set_resnames(*args, **kwds)\n",
      " |      `set_resnames` is deprecated!\n",
      " |      This will be removed in version 0.16.0; use `resnames` property instead\n",
      " |      \n",
      " |      Set the resnames to string *resname* for **all atoms** in the :class:`AtomGroup`.\n",
      " |      \n",
      " |              If *resname* is a sequence of the same length as the :class:`AtomGroup`\n",
      " |              then each :attr:`Atom.resname` is set to the corresponding value together\n",
      " |              with the :attr:`Residue.resname` of the residue the atom belongs to. If\n",
      " |              *value* is neither of length 1 (or a scalar) nor of the length of the\n",
      " |              :class:`AtomGroup` then a :exc:`ValueError` is raised.\n",
      " |      \n",
      " |              .. versionadded:: 0.7.4\n",
      " |              .. versionchanged:: 0.7.5\n",
      " |                 Also changes the residues.\n",
      " |              .. versionchanged:: 0.8\n",
      " |                 Can set atoms and residues to distinct values by providing a sequence or iterable.\n",
      " |              .. versionchanged:: 0.11.0\n",
      " |                 Made plural to make consistent with corresponding property\n",
      " |  \n",
      " |  set_resnum(*args, **kwds)\n",
      " |      `set_resnum` is deprecated, use `set_resnums` instead!\n",
      " |      This will be removed in version 0.16.0\n",
      " |      \n",
      " |      `set_resnums` is deprecated!\n",
      " |      This will be removed in version 0.16.0; use `resnums` property instead\n",
      " |      \n",
      " |      Set the resnums to *resnum* for **all atoms** in the :class:`AtomGroup`.\n",
      " |      \n",
      " |              If *resnum* is a sequence of the same length as the :class:`AtomGroup`\n",
      " |              then each :attr:`Atom.resnum` is set to the corresponding value together\n",
      " |              with the :attr:`Residue.resnum` of the residue the atom belongs to. If\n",
      " |              *value* is neither of length 1 (or a scalar) nor of the length of the\n",
      " |              :class:`AtomGroup` then a :exc:`ValueError` is raised.\n",
      " |      \n",
      " |              .. Note::\n",
      " |      \n",
      " |                 Changing *resnum* will not affect the topology: you can have\n",
      " |                 multiple residues with the same *resnum*.\n",
      " |      \n",
      " |              .. SeeAlso:: :meth:`set_resid`\n",
      " |      \n",
      " |              .. versionadded:: 0.7.4\n",
      " |              .. versionchanged:: 0.7.5\n",
      " |                 Also changes the residues.\n",
      " |              .. versionchanged:: 0.8\n",
      " |                 Can set atoms and residues to distinct values by providing a sequence or iterable.\n",
      " |              .. versionchanged:: 0.11.0\n",
      " |                 Made plural to make consistent with corresponding property\n",
      " |  \n",
      " |  set_resnums(*args, **kwds)\n",
      " |      `set_resnums` is deprecated!\n",
      " |      This will be removed in version 0.16.0; use `resnums` property instead\n",
      " |      \n",
      " |      Set the resnums to *resnum* for **all atoms** in the :class:`AtomGroup`.\n",
      " |      \n",
      " |              If *resnum* is a sequence of the same length as the :class:`AtomGroup`\n",
      " |              then each :attr:`Atom.resnum` is set to the corresponding value together\n",
      " |              with the :attr:`Residue.resnum` of the residue the atom belongs to. If\n",
      " |              *value* is neither of length 1 (or a scalar) nor of the length of the\n",
      " |              :class:`AtomGroup` then a :exc:`ValueError` is raised.\n",
      " |      \n",
      " |              .. Note::\n",
      " |      \n",
      " |                 Changing *resnum* will not affect the topology: you can have\n",
      " |                 multiple residues with the same *resnum*.\n",
      " |      \n",
      " |              .. SeeAlso:: :meth:`set_resid`\n",
      " |      \n",
      " |              .. versionadded:: 0.7.4\n",
      " |              .. versionchanged:: 0.7.5\n",
      " |                 Also changes the residues.\n",
      " |              .. versionchanged:: 0.8\n",
      " |                 Can set atoms and residues to distinct values by providing a sequence or iterable.\n",
      " |              .. versionchanged:: 0.11.0\n",
      " |                 Made plural to make consistent with corresponding property\n",
      " |  \n",
      " |  set_segid(*args, **kwds)\n",
      " |      `set_segid` is deprecated, use `set_segids` instead!\n",
      " |      This will be removed in version 0.16.0\n",
      " |      \n",
      " |      `set_segids` is deprecated!\n",
      " |      This will be removed in version 0.16.0; use `segids` property instead\n",
      " |      \n",
      " |      Set the segids to *segid* for all atoms in the :class:`AtomGroup`.\n",
      " |      \n",
      " |              If *segid* is a sequence of the same length as the :class:`AtomGroup`\n",
      " |              then each :attr:`Atom.segid` is set to the corresponding value together\n",
      " |              with the :attr:`Segment.segid` of the residue the atom belongs to. If\n",
      " |              *value* is neither of length 1 (or a scalar) nor of the length of the\n",
      " |              :class:`AtomGroup` then a :exc:`ValueError` is raised.\n",
      " |      \n",
      " |              .. Note::\n",
      " |      \n",
      " |                 :meth:`set_segid` can change the topology.\n",
      " |      \n",
      " |              .. versionadded:: 0.7.4\n",
      " |              .. versionchanged:: 0.8\n",
      " |                 Can set atoms and residues to distinct values by providing a sequence or iterable.\n",
      " |              .. versionchanged:: 0.11.0\n",
      " |                 Stale caches are problematic; though it can be expensive, changing segid\n",
      " |                 results in Segments being regenerated\n",
      " |              .. versionchanged:: 0.11.0\n",
      " |                 Made plural to make consistent with corresponding property\n",
      " |  \n",
      " |  set_segids(*args, **kwds)\n",
      " |      `set_segids` is deprecated!\n",
      " |      This will be removed in version 0.16.0; use `segids` property instead\n",
      " |      \n",
      " |      Set the segids to *segid* for all atoms in the :class:`AtomGroup`.\n",
      " |      \n",
      " |              If *segid* is a sequence of the same length as the :class:`AtomGroup`\n",
      " |              then each :attr:`Atom.segid` is set to the corresponding value together\n",
      " |              with the :attr:`Segment.segid` of the residue the atom belongs to. If\n",
      " |              *value* is neither of length 1 (or a scalar) nor of the length of the\n",
      " |              :class:`AtomGroup` then a :exc:`ValueError` is raised.\n",
      " |      \n",
      " |              .. Note::\n",
      " |      \n",
      " |                 :meth:`set_segid` can change the topology.\n",
      " |      \n",
      " |              .. versionadded:: 0.7.4\n",
      " |              .. versionchanged:: 0.8\n",
      " |                 Can set atoms and residues to distinct values by providing a sequence or iterable.\n",
      " |              .. versionchanged:: 0.11.0\n",
      " |                 Stale caches are problematic; though it can be expensive, changing segid\n",
      " |                 results in Segments being regenerated\n",
      " |              .. versionchanged:: 0.11.0\n",
      " |                 Made plural to make consistent with corresponding property\n",
      " |  \n",
      " |  set_serial(*args, **kwds)\n",
      " |      `set_serial` is deprecated, use `set_serials` instead!\n",
      " |      This will be removed in version 0.16.0\n",
      " |      \n",
      " |      `set_serials` is deprecated!\n",
      " |      This will be removed in version 0.16.0; use `serials` property instead\n",
      " |      \n",
      " |      Set the serials to *serial* for **all atoms** in the AtomGroup.\n",
      " |      \n",
      " |              If *serial* is a sequence of the same length as the :class:`AtomGroup`\n",
      " |              then each :attr:`Atom.serial` is set to the corresponding value. If\n",
      " |              *value* is neither of length 1 (or a scalar) nor of the length of the\n",
      " |              :class:`AtomGroup` then a :exc:`ValueError` is raised.\n",
      " |      \n",
      " |              .. versionadded:: 0.11.0\n",
      " |  \n",
      " |  set_serials(*args, **kwds)\n",
      " |      `set_serials` is deprecated!\n",
      " |      This will be removed in version 0.16.0; use `serials` property instead\n",
      " |      \n",
      " |      Set the serials to *serial* for **all atoms** in the AtomGroup.\n",
      " |      \n",
      " |              If *serial* is a sequence of the same length as the :class:`AtomGroup`\n",
      " |              then each :attr:`Atom.serial` is set to the corresponding value. If\n",
      " |              *value* is neither of length 1 (or a scalar) nor of the length of the\n",
      " |              :class:`AtomGroup` then a :exc:`ValueError` is raised.\n",
      " |      \n",
      " |              .. versionadded:: 0.11.0\n",
      " |  \n",
      " |  set_type(*args, **kwds)\n",
      " |      `set_type` is deprecated, use `set_types` instead!\n",
      " |      This will be removed in version 0.16.0\n",
      " |      \n",
      " |      `set_types` is deprecated!\n",
      " |      This will be removed in version 0.16.0; use `types` property instead\n",
      " |      \n",
      " |      Set the atom types to *atype* for **all atoms** in the AtomGroup.\n",
      " |      \n",
      " |              If *atype* is a sequence of the same length as the :class:`AtomGroup`\n",
      " |              then each :attr:`Atom.atype` is set to the corresponding value. If\n",
      " |              *value* is neither of length 1 (or a scalar) nor of the length of the\n",
      " |              :class:`AtomGroup` then a :exc:`ValueError` is raised.\n",
      " |      \n",
      " |              .. versionadded:: 0.7.4\n",
      " |              .. versionchanged:: 0.8\n",
      " |                 Can set atoms and residues to distinct values by providing a sequence or iterable.\n",
      " |              .. versionchanged:: 0.11.0\n",
      " |                 Made plural to make consistent with corresponding property\n",
      " |  \n",
      " |  set_types(*args, **kwds)\n",
      " |      `set_types` is deprecated!\n",
      " |      This will be removed in version 0.16.0; use `types` property instead\n",
      " |      \n",
      " |      Set the atom types to *atype* for **all atoms** in the AtomGroup.\n",
      " |      \n",
      " |              If *atype* is a sequence of the same length as the :class:`AtomGroup`\n",
      " |              then each :attr:`Atom.atype` is set to the corresponding value. If\n",
      " |              *value* is neither of length 1 (or a scalar) nor of the length of the\n",
      " |              :class:`AtomGroup` then a :exc:`ValueError` is raised.\n",
      " |      \n",
      " |              .. versionadded:: 0.7.4\n",
      " |              .. versionchanged:: 0.8\n",
      " |                 Can set atoms and residues to distinct values by providing a sequence or iterable.\n",
      " |              .. versionchanged:: 0.11.0\n",
      " |                 Made plural to make consistent with corresponding property\n",
      " |  \n",
      " |  set_velocities(*args, **kwds)\n",
      " |      `set_velocities` is deprecated!\n",
      " |      This will be removed in version 0.16.0; use `velocities` property instead\n",
      " |      \n",
      " |      Assign the velocities *v* to the timestep.\n",
      " |      \n",
      " |              Raises a :exc:`NoDataError` if the underlying\n",
      " |              :class:`~MDAnalysis.coordinates.base.Timestep` does not contain\n",
      " |              :attr:`~MDAnalysis.coordinates.base.Timestep.velocities`.\n",
      " |      \n",
      " |              See also :meth:`AtomGroup.get_velocities` and :attr:`AtomGroup.velocities` for\n",
      " |              attribute access.\n",
      " |      \n",
      " |              .. versionadded:: 0.7.6\n",
      " |  \n",
      " |  shapeParameter(*args, **kwds)\n",
      " |      `shapeParameter` is deprecated, use `shape_parameter` instead!\n",
      " |      This will be removed in version 0.16.0\n",
      " |      \n",
      " |      Shape parameter.\n",
      " |      \n",
      " |              See [Dima2004]_ for background information.\n",
      " |      \n",
      " |              :Keywords:\n",
      " |                *pbc*\n",
      " |                  ``True``: Move all atoms within the primary unit cell before calculation [``False``]\n",
      " |      \n",
      " |              .. Note::\n",
      " |                  The :class:`MDAnalysis.core.flags` flag *use_pbc* when set to ``True`` allows the *pbc*\n",
      " |                  flag to be used by default.\n",
      " |      \n",
      " |              .. versionadded:: 0.7.7\n",
      " |              .. versionchanged:: 0.8 Added *pbc* keyword\n",
      " |  \n",
      " |  shape_parameter(self, **kwargs)\n",
      " |      Shape parameter.\n",
      " |      \n",
      " |      See [Dima2004]_ for background information.\n",
      " |      \n",
      " |      :Keywords:\n",
      " |        *pbc*\n",
      " |          ``True``: Move all atoms within the primary unit cell before calculation [``False``]\n",
      " |      \n",
      " |      .. Note::\n",
      " |          The :class:`MDAnalysis.core.flags` flag *use_pbc* when set to ``True`` allows the *pbc*\n",
      " |          flag to be used by default.\n",
      " |      \n",
      " |      .. versionadded:: 0.7.7\n",
      " |      .. versionchanged:: 0.8 Added *pbc* keyword\n",
      " |  \n",
      " |  split(self, level)\n",
      " |      Split atomgroup into a list of atomgroups by *level*.\n",
      " |      \n",
      " |      *level* can be \"atom\", \"residue\", \"segment\".\n",
      " |      .. versionadded:: 0.9.0\n",
      " |  \n",
      " |  totalCharge(*args, **kwds)\n",
      " |      `totalCharge` is deprecated, use `total_charge` instead!\n",
      " |      This will be removed in version 0.16.0\n",
      " |      \n",
      " |      Sum of all partial charges (must be defined in topology).\n",
      " |  \n",
      " |  totalMass(*args, **kwds)\n",
      " |      `totalMass` is deprecated, use `total_mass` instead!\n",
      " |      This will be removed in version 0.16.0\n",
      " |      \n",
      " |      Total mass of the selection (masses are taken from the topology or guessed).\n",
      " |  \n",
      " |  total_charge(self)\n",
      " |      Sum of all partial charges (must be defined in topology).\n",
      " |  \n",
      " |  total_mass(self)\n",
      " |      Total mass of the selection (masses are taken from the topology or guessed).\n",
      " |  \n",
      " |  transform(self, M)\n",
      " |      Apply homogenous transformation matrix *M* to the coordinates.\n",
      " |      \n",
      " |      The matrix *M* must be a 4x4 matrix, with the rotation in\n",
      " |      `R = `M[:3,:3]`` and the translation in ``t = M[:3,3]``.\n",
      " |      \n",
      " |      The rotation :math:`\\mathsf{R}` is applied before the\n",
      " |      translation :math:`\\mathbf{t}`:\n",
      " |      \n",
      " |      .. math::\n",
      " |      \n",
      " |         \\mathbf{x}' = \\mathsf{R}\\mathbf{x} + \\mathbf{t}\n",
      " |      \n",
      " |      .. SeeAlso: :mod:`MDAnalysis.core.transformations`\n",
      " |  \n",
      " |  translate(self, t)\n",
      " |      Apply translation vector *t* to the selection's coordinates.\n",
      " |      \n",
      " |        >>> AtomGroup.translate(t)\n",
      " |        >>> AtomGroup.translate((A, B))\n",
      " |      \n",
      " |      The method applies a translation to the AtomGroup from current\n",
      " |      coordinates :math:`\\mathbf{x}` to new coordinates :math:`\\mathbf{x}'`:\n",
      " |      \n",
      " |      .. math::\n",
      " |      \n",
      " |          \\mathbf{x}' = \\mathbf{x} + \\mathbf{t}\n",
      " |      \n",
      " |      The translation can also be given as a tuple of two MDAnalysis objects\n",
      " |      such as two selections `(selA, selB)`, i.e. two :class:`AtomGroup`, or\n",
      " |      two :class:`Atom`. The translation vector is computed as the\n",
      " |      difference between the centers of geometry (centroid) of B and A::\n",
      " |      \n",
      " |          t = B.centroid() - A.centroid()\n",
      " |  \n",
      " |  wrap(self, compound='atoms', center='com', box=None)\n",
      " |      Shift the contents of this AtomGroup back into the unit cell.\n",
      " |      \n",
      " |      This is a more powerful version of :meth:`pack_into_box`, allowing\n",
      " |      groups of atoms to be kept together through the process.\n",
      " |      \n",
      " |      :Keywords:\n",
      " |         *compound*\n",
      " |             The group which will be kept together through the shifting\n",
      " |             process. [``atoms``]\n",
      " |      \n",
      " |             Possible options:\n",
      " |      \n",
      " |                 - ``atoms``\n",
      " |                 - ``group`` - This AtomGroup\n",
      " |                 - ``residues``\n",
      " |                 - ``segments``\n",
      " |                 - ``fragments``\n",
      " |      \n",
      " |         *center*\n",
      " |             How to define the center of a given group of atoms [``com``]\n",
      " |         *box*\n",
      " |             Unit cell information.  If not provided, the values from\n",
      " |             Timestep will be used.\n",
      " |      \n",
      " |      When specifying a *compound*, the translation is calculated based on\n",
      " |      each compound. The same translation is applied to all atoms\n",
      " |      within this compound, meaning it will not be broken by the shift.\n",
      " |      This might however mean that all atoms from the compound are not\n",
      " |      inside the unit cell, but rather the center of the compound is.\n",
      " |      Compounds available for use are *atoms*, *residues*,\n",
      " |      *segments* and *fragments*\n",
      " |      \n",
      " |      *center* allows the definition of the center of each group to be\n",
      " |      specified.  This can be either 'com' for center of mass, or 'cog'\n",
      " |      for center of geometry.\n",
      " |      \n",
      " |      *box* allows a unit cell to be given for the transformation.  If not\n",
      " |      specified, an the dimensions information from the current Timestep\n",
      " |      will be used.\n",
      " |      \n",
      " |      .. Note::\n",
      " |         wrap with all default keywords is identical to :meth:`pack_into_box`\n",
      " |      \n",
      " |      .. versionadded:: 0.9.2\n",
      " |  \n",
      " |  write(self, filename=None, format='PDB', filenamefmt='%(trjname)s_%(frame)d', **kwargs)\n",
      " |      Write AtomGroup to a file.\n",
      " |      \n",
      " |      AtomGroup.write(filename[,format])\n",
      " |      \n",
      " |      :Keywords:\n",
      " |        *filename*\n",
      " |             ``None``: create TRJNAME_FRAME.FORMAT from filenamefmt [``None``]\n",
      " |        *format*\n",
      " |              PDB, CRD, GRO, VMD (tcl), PyMol (pml), Gromacs (ndx) CHARMM (str)\n",
      " |              Jmol (spt); case-insensitive and can also be supplied as the\n",
      " |              filename extension [PDB]\n",
      " |        *filenamefmt*\n",
      " |              format string for default filename; use substitution tokens\n",
      " |              'trjname' and 'frame' [\"%(trjname)s_%(frame)d\"]\n",
      " |        *bonds*\n",
      " |              how to handle bond information, especially relevant for PDBs;\n",
      " |              default is ``\"conect\"``.\n",
      " |      \n",
      " |              * ``\"conect\"``: write only the CONECT records defined in the original\n",
      " |                file\n",
      " |      \n",
      " |              * ``\"all\"``: write out all bonds, both the original defined and those\n",
      " |                guessed by MDAnalysis\n",
      " |      \n",
      " |              * ``None``: do not write out bonds\n",
      " |      \n",
      " |      .. versionchanged:: 0.9.0\n",
      " |         Merged with write_selection.  This method can now write both\n",
      " |         selections out.\n",
      " |  \n",
      " |  write_selection(*args, **kwds)\n",
      " |      `write_selection` is deprecated!\n",
      " |      This will be removed in version 0.16.0; use `write` method instead\n",
      " |      \n",
      " |      Write AtomGroup selection to a file to be used in another programme.\n",
      " |      \n",
      " |              :Keywords:\n",
      " |                *filename*\n",
      " |                      ``None``: create TRJNAME_FRAME.FORMAT from *filenamefmt*\n",
      " |                *format*\n",
      " |                      output file format: VMD (tcl), PyMol (pml), Gromacs (ndx), CHARMM (str);\n",
      " |                      can also be supplied as the filename extension. Case insensitive. [vmd]\n",
      " |                *filenamefmt*\n",
      " |                      format string for default filename; use '%(trjname)s' and '%(frame)s'\n",
      " |                      placeholders; the extension is set according to the *format*\n",
      " |                      [\"%(trjname)s_%(frame)d\"]\n",
      " |                *kwargs*\n",
      " |                      additional keywords are passed on to the appropriate\n",
      " |                      :class:`~MDAnalysis.selections.base.SelectionWriter`\n",
      " |      \n",
      " |              .. deprecated:: 0.9.0\n",
      " |                 Use :meth:`write`\n",
      " |  \n",
      " |  ----------------------------------------------------------------------\n",
      " |  Data descriptors defined here:\n",
      " |  \n",
      " |  __dict__\n",
      " |      dictionary for instance variables (if defined)\n",
      " |  \n",
      " |  __weakref__\n",
      " |      list of weak references to the object (if defined)\n",
      " |  \n",
      " |  altLocs\n",
      " |  \n",
      " |  angle\n",
      " |      This AtomGroup represented as an Angle object\n",
      " |      \n",
      " |      :Returns:\n",
      " |        A :class:`MDAnalysis.core.topologyobjects.Angle` object\n",
      " |      \n",
      " |      :Raises:\n",
      " |        `ValueError` if the AtomGroup is not length 3\n",
      " |      \n",
      " |      .. versionadded:: 0.11.0\n",
      " |  \n",
      " |  angles\n",
      " |  \n",
      " |  atoms\n",
      " |      :class:`AtomGroup` of all atoms in this group.\n",
      " |      \n",
      " |      If this is a :class:`AtomGroup` then it returns itself. Otherwise, it\n",
      " |      will return a new :class:`AtomGroup` based on all :class:`Atom`\n",
      " |      instances contained.\n",
      " |      \n",
      " |      Apply `:func:`list` to :attr:`atoms` or use :attr:`_atoms` if you\n",
      " |      really only need a list of individual :class:`Atom` instances.\n",
      " |  \n",
      " |  bfactors\n",
      " |  \n",
      " |  bond\n",
      " |      This AtomGroup represented as a Bond object\n",
      " |      \n",
      " |      :Returns:\n",
      " |        A :class:`MDAnalysis.core.topologyobjects.Bond` object\n",
      " |      \n",
      " |      :Raises:\n",
      " |        `ValueError` if the AtomGroup is not length 2\n",
      " |      \n",
      " |      .. versionadded:: 0.11.0\n",
      " |  \n",
      " |  bonds\n",
      " |  \n",
      " |  charges\n",
      " |  \n",
      " |  dihedral\n",
      " |      This AtomGroup represented as a Dihedral object\n",
      " |      \n",
      " |      :Returns:\n",
      " |        A :class:`MDAnalysis.core.topologyobjects.Dihedral` object\n",
      " |      \n",
      " |      :Raises:\n",
      " |        `ValueError` if the AtomGroup is not length 4\n",
      " |      \n",
      " |      .. versionadded:: 0.11.0\n",
      " |  \n",
      " |  dihedrals\n",
      " |  \n",
      " |  dimensions\n",
      " |      Dimensions of the Universe to which the group belongs, at the current time step.\n",
      " |  \n",
      " |  forces\n",
      " |      Forces on the atoms in the AtomGroup.\n",
      " |      \n",
      " |      The forces can be changed by assigning an array of the appropriate\n",
      " |      shape, i.e. either Nx3 to assign individual force or 3, to assign\n",
      " |      the *same* force to all atoms (e.g. ``ag.forces = array([0,0,0])``\n",
      " |      will set all forces to (0.,0.,0.)).\n",
      " |      \n",
      " |      For more control use the :meth:`~AtomGroup.get_forces` and\n",
      " |      :meth:`~AtomGroup.set_forces` methods.\n",
      " |      \n",
      " |      .. versionadded:: 0.7.7\n",
      " |  \n",
      " |  fragments\n",
      " |      Read-only list of fragments.\n",
      " |      \n",
      " |      Contains all fragments that any Atom in this AtomGroup is part of, the contents of\n",
      " |      the fragments may extend beyond the contents of this AtomGroup.\n",
      " |      \n",
      " |      .. versionadded 0.9.0\n",
      " |  \n",
      " |  ids\n",
      " |  \n",
      " |  improper\n",
      " |      This AtomGroup represented as an ImproperDihedral object\n",
      " |      \n",
      " |      :Returns:\n",
      " |        A :class:`MDAnalysis.core.topologyobjects.ImproperDihedral` object\n",
      " |      \n",
      " |      :Raises:\n",
      " |        `ValueError` if the AtomGroup is not length 4\n",
      " |      \n",
      " |      .. versionadded:: 0.11.0\n",
      " |  \n",
      " |  impropers\n",
      " |  \n",
      " |  indices\n",
      " |  \n",
      " |  masses\n",
      " |  \n",
      " |  n_atoms\n",
      " |      Total number of atoms in the group\n",
      " |  \n",
      " |  n_residues\n",
      " |      Total number of residues in the group\n",
      " |  \n",
      " |  n_segments\n",
      " |      Total number of segments in the group\n",
      " |  \n",
      " |  names\n",
      " |  \n",
      " |  occupancies\n",
      " |  \n",
      " |  positions\n",
      " |      Coordinates of the atoms in the AtomGroup.\n",
      " |      \n",
      " |      The positions can be changed by assigning an array of the appropriate\n",
      " |      shape, i.e. either Nx3 to assign individual coordinates or 3, to assign\n",
      " |      the *same* coordinate to all atoms (e.g. ``ag.positions = array([0,0,0])``\n",
      " |      will move all particles to the origin).\n",
      " |      \n",
      " |      .. versionadded:: 0.7.6\n",
      " |  \n",
      " |  radii\n",
      " |  \n",
      " |  resids\n",
      " |  \n",
      " |  residues\n",
      " |      Read-only list of :class:`Residue` objects.\n",
      " |      \n",
      " |      A :class:`ResidueGroup` of all residues that contain atoms in\n",
      " |      this group.\n",
      " |      \n",
      " |      .. versionchanged:: 0.9.0\n",
      " |         Now returns strictly a ResidueGroup of the unique Residues that Atoms in this group\n",
      " |         belong to.\n",
      " |  \n",
      " |  resnames\n",
      " |  \n",
      " |  resnums\n",
      " |  \n",
      " |  segids\n",
      " |  \n",
      " |  segments\n",
      " |      Read-only list of :class:`Segment` objects.\n",
      " |      \n",
      " |      A :class:`SegmentGroup` of all segments that contain atoms in this group.\n",
      " |      \n",
      " |      .. versionchanged:: 0.9.0\n",
      " |         Now strictly returns a :class:`SegmentGroup` of a set of\n",
      " |         the :class:`Segment` instances from this :class:`AtomGroup`\n",
      " |  \n",
      " |  serials\n",
      " |      `serials` is deprecated!\n",
      " |      This will be removed in version 0.16.0; use `ids` property instead\n",
      " |      \n",
      " |      numpy array of the serials for all atoms in this group\n",
      " |      \n",
      " |              .. versionadded:: 0.11.0\n",
      " |  \n",
      " |  ts\n",
      " |      Returns a Timestep that contains only the group's coordinates.\n",
      " |      \n",
      " |      Returns\n",
      " |      -------\n",
      " |      A :class:`~MDAnalysis.coordinates.base.Timestep` instance,\n",
      " |      which can be passed to a trajectory writer.\n",
      " |      \n",
      " |      If the returned timestep is modified the modifications\n",
      " |      will not be reflected in the base timestep. Likewise,\n",
      " |      when the underlying timestep changes (either by loading a\n",
      " |      new frame or by setting new positions by hand) the returned\n",
      " |      timestep will not reflect those changes.\n",
      " |  \n",
      " |  types\n",
      " |  \n",
      " |  universe\n",
      " |      The universe to which the atoms belong (read-only).\n",
      " |  \n",
      " |  velocities\n",
      " |      numpy array of the velocities of the atoms in the group.\n",
      " |      \n",
      " |      If the trajectory does not contain velocity information then a\n",
      " |      :exc:`~MDAnalysis.NoDataError` is raised.\n",
      " |      \n",
      " |      .. versionadded:: 0.7.5\n",
      " |      .. deprecated:: 0.7.6\n",
      " |         In 0.8 this will become an attribute! You can already use :meth:`get_velocities`\n",
      " |         and :meth:`set_velocities`.\n",
      " |      .. versionchanged:: 0.8\n",
      " |         Became an attribute.\n",
      "\n"
     ]
    }
   ],
   "source": [
    "help(u.atoms[100:130])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Larger organizational units are Segment instances, for example one protein or all the solvent molecules or simply the whole system. Atom, AtomGroup, Residue, and ResidueGroup have an attribute segments that will list the segment IDs (segids) as a SegmentGroup:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<SegmentGroup [<Segment 4AKE>]>"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "u.atoms.segments"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The converse is also true: each higher level in the hierarchy also know about the Residue and Atom instances it contains. For example, to list the atoms of the ResidueGroup we had before:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<AtomGroup with 36 atoms>"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "r = u.atoms[100:130].residues\n",
    "r.atoms"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Exercise 1"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "What residue (resname) does the last atom belong to in the above example?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<Atom 136: O of type 70 of resname ALA, resid 8 and segid 4AKE>"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "r = u.atoms[100:130].residues\n",
    "r.atoms[-1]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Why does the expression"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "False"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(u.atoms[100:130]) == len(u.atoms[100:130].residues.atoms)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "return false? Because the complete residues contain more atoms than the arbitrary slice of atoms."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "How many residues are in the Universe u?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "214"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(u.atoms.residues)\n",
    "u.atoms.n_residues"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "How do you get a list of the residue names (such as `[\"Ala\", \"Gly\", \"Gly\", \"Asp\", ...]`) and residue numbers (resid) for atoms 1000 to 1300? And as a list of tuples `(resname, resid)` (Hint: `zip()`)?:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[('LEU', 67),\n",
       " ('VAL', 68),\n",
       " ('LYS', 69),\n",
       " ('GLU', 70),\n",
       " ('ARG', 71),\n",
       " ('ILE', 72),\n",
       " ('ALA', 73),\n",
       " ('GLN', 74),\n",
       " ('GLU', 75),\n",
       " ('ASP', 76),\n",
       " ('CYS', 77),\n",
       " ('ARG', 78),\n",
       " ('ASN', 79),\n",
       " ('GLY', 80),\n",
       " ('PHE', 81),\n",
       " ('LEU', 82),\n",
       " ('LEU', 83),\n",
       " ('ASP', 84),\n",
       " ('GLY', 85),\n",
       " ('PHE', 86)]"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "resnames = u.atoms[999:1300].residues.resnames\n",
    "resids = u.atoms[999:1300].residues.resids\n",
    "zip(resnames, resids)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "How do you obtain the resid and the resname for the 100th residue? (Hint: investigate the Residue object interactively with TAB completion)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(100, 'GLY')\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/anaconda2-4.1.1/lib/python2.7/site-packages/numpy/lib/utils.py:99: DeprecationWarning: `id` is deprecated!\n",
      "This will be removed in version 0.16.0; use `resid` property instead\n",
      "  warnings.warn(depdoc, DeprecationWarning)\n",
      "/opt/anaconda2-4.1.1/lib/python2.7/site-packages/numpy/lib/utils.py:99: DeprecationWarning: `name` is deprecated!\n",
      "This will be removed in version 0.16.0; use `resname` property instead\n",
      "  warnings.warn(depdoc, DeprecationWarning)\n"
     ]
    }
   ],
   "source": [
    "r100 = u.atoms.residues[99]\n",
    "print(r100.id, r100.name)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "How many segments are there?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1\n",
      "1\n",
      "1\n"
     ]
    }
   ],
   "source": [
    "print(len(u.segments))\n",
    "print(len(u.atoms.segments))\n",
    "print(u.atoms.n_segments)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "What is the segment identifier of the first Segment?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/anaconda2-4.1.1/lib/python2.7/site-packages/numpy/lib/utils.py:99: DeprecationWarning: `id` is deprecated!\n",
      "This will be removed in version 0.16.0; use `segid` property instead\n",
      "  warnings.warn(depdoc, DeprecationWarning)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "'4AKE'"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "s1 = u.segments[0]\n",
    "s1.id"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Selections"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "MDAnalysis comes with a fairly complete atom selection facility. Primarily, one uses the method [select_atoms()](http://docs.mdanalysis.org/documentation_pages/selections.html) of a Universe:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<AtomGroup with 214 atoms>"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "CA = u.select_atoms(\"protein and name CA\")\n",
    "CA"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "but really any AtomGroup has a selectAtoms() method:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<AtomGroup with 35 atoms>"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "acidic = CA.select_atoms(\"resname ASP or resname GLU\")\n",
    "acidic"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<ResidueGroup [<Residue GLU, 22>, <Residue ASP, 33>, <Residue GLU, 44>, <Residue ASP, 51>, <Residue ASP, 54>, <Residue ASP, 61>, <Residue GLU, 62>, <Residue GLU, 70>, <Residue GLU, 75>, <Residue ASP, 76>, <Residue ASP, 84>, <Residue ASP, 94>, <Residue GLU, 98>, <Residue ASP, 104>, <Residue GLU, 108>, <Residue ASP, 110>, <Residue ASP, 113>, <Residue GLU, 114>, <Residue ASP, 118>, <Residue GLU, 143>, <Residue ASP, 146>, <Residue ASP, 147>, <Residue GLU, 151>, <Residue GLU, 152>, <Residue ASP, 158>, <Residue ASP, 159>, <Residue GLU, 161>, <Residue GLU, 162>, <Residue GLU, 170>, <Residue GLU, 185>, <Residue GLU, 187>, <Residue ASP, 197>, <Residue GLU, 204>, <Residue ASP, 208>, <Residue GLU, 210>]>"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "acidic.residues"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Exercise 2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Select the range of resids 100 to 200 (100-200) with a selection. Compare the result to what you get by slicing the u.atoms.residues appropriately.\n",
    "\n",
    "Which approach would you prefer to use in a analysis script?\n",
    "\n",
    "Solution:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<AtomGroup with 1609 atoms>"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "u.select_atoms(\"resid 100-200\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Compare to the slicing solution (doing an element-wise comparison, i.e. residue by residue in each list()):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<AtomGroup with 1609 atoms>"
      ]
     },
     "execution_count": 26,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "r = u.atoms.residues[99:200]\n",
    "r.atoms"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "list(u.select_atoms(\"resid 100-200\").residues) == list(u.atoms.residues[99:200])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "If one wants to get specific residues in scripts one typically uses selections instead of slicing because the index in the slice might not correspond to the actual residue ids (minus 1): If a number of residues (e.g. 150-160) are missing from the structure then the selection will simply give you residues 100-149 and 151-200 but the slice 99:200 would give you residues 100-149 and 161-209."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Select all residues that do not contain a C (CB) atom. How many are there? What residue names did you find?\n",
    "\n",
    "Solution:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "20"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sel = u.select_atoms(\"(byres name CA) and not (byres name CB)\").residues\n",
    "len(sel)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "These are all Glycines, as can be seen by comparing the residue groups element-wise:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 29,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "glycines = u.select_atoms(\"resname GLY\")\n",
    "list(sel) == list(glycines.residues)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Working with AtomGroups"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "An AtomGroup has a large number of methods attributes defined that provide information about the atoms such as names, indices, or the coordinates in the positions attribute:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(214, 3)"
      ]
     },
     "execution_count": 30,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "CA = u.select_atoms(\"protein and name CA\")\n",
    "r = CA.positions\n",
    "r.shape"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The resulting output is a numpy.ndarray. The main purpose of MDAnalysis is to get trajectory data into numpy arrays!"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Important methods and attributes of AtomGroup"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The coordinates positions attribute is probably the most important information that you can get from an AtomGroup.\n",
    "\n",
    "Other quantities that can be easily calculated for a AtomGroup are\n",
    "\n",
    "* the center of mass `center_of_mass()` and the center of geometry (or centroid) `center_of_geometry()` (equivalent to `centroid()`);\n",
    "\n",
    "* the total mass `total_mass()`;\n",
    "\n",
    "* the total charge `total_charge()` (if partial charges are defined in the topology);\n",
    "\n",
    "* the radius of gyration\n",
    "\\begin{equation*}\n",
    "    R_{\\text{gyr}} = \\sqrt{\\frac{1}{M}\\sum_{i=1}^{N}m_{i}(\\textbf{r}_{i} - \\textbf{R})^{2}}\n",
    "\\end{equation*}\n",
    " with `radius_of_yration()`;\n",
    "\n",
    "* the principal axes $\\textbf{p}_{1}$,$\\textbf{p}_{2}$,$\\textbf{p}_{3}$ from `principal_axes()` via a diagonalization of the tensor of inertia `moment_of_inertia()`,\n",
    "\\begin{align*}\n",
    "    \\Lambda = U^{T}IU && \\text{with}\\,U = (\\textbf{p}_{1},\\textbf{p}_{2},\\textbf{p}_{3})\n",
    "\\end{align*}\n",
    "where $U$ is a rotation matrix whose columns are the eigenvectors that form the principal axes, $\\Lambda$ is is the diagonal matrix of eigenvalues (sorted from largest to smallest) known as the principal moments of inertia, and $I = \\sum_{i=1}^{N}m_{i}[\\textbf{(r}_{i}\\cdot\\textbf{r}_{i})\\sum_{a=1}^{3}\\textbf{e}_{\\alpha}\\otimes\\textbf{e}_{\\alpha} - \\textbf{r}_{i}\\otimes\\textbf{r}_{i}]$ is the tensor of inertia."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Exercise 3"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "AdK consists of three domains:\n",
    "\n",
    "* CORE residues 1-29, 60-121, 160-214 (gray)\n",
    "* NMP residues 30-59 (blue)\n",
    "* LID residues 122-159 (yellow)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Calculate the center of mass and the center of geometry for each of the three domains."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'LID': array([-15.16074944,   2.11599636,  -4.37305355], dtype=float32), 'CORE': array([ 4.43884087,  2.05389476,  1.63895261], dtype=float32), 'NMP': array([ -2.99990702, -13.62531662,  -2.93235731], dtype=float32)}\n"
     ]
    }
   ],
   "source": [
    "domains = {\n",
    "    'CORE': u.select_atoms(\"protein and (resid 1-29 or resid 60-121 or resid 160-214)\"),\n",
    "    'NMP': u.select_atoms(\"protein and resid 30-59\"),\n",
    "    'LID': u.select_atoms(\"protein and resid 122-159\")\n",
    "    }\n",
    "cg = dict((name, dom.centroid()) for name,dom in domains.items())\n",
    "cm = dict((name, dom.center_of_mass()) for name,dom in domains.items())\n",
    "print(cg)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "{'LID': array([-15.11337499,   2.12292226,  -4.40910485]), 'CORE': array([ 4.564116  ,  2.08700105,  1.54992649]), 'NMP': array([ -3.20330174, -13.60247613,  -3.06221538])}\n"
     ]
    }
   ],
   "source": [
    "print(cm)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "What are the distances between the centers of mass?\n",
    "\n",
    "(Hint: you can use numpy.linalg.norm() or calculate it manually.)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "18.1042626244\n"
     ]
    }
   ],
   "source": [
    "from numpy.linalg import norm\n",
    "print(norm(cm['CORE'] - cm['NMP']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "20.5600339602\n"
     ]
    }
   ],
   "source": [
    "print(norm(cm['CORE'] - cm['LID']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "19.7725089609\n"
     ]
    }
   ],
   "source": [
    "print(norm(cm['NMP'] - cm['LID']))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Does it matter to use center of mass vs center of geometry?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "17.9463\n"
     ]
    }
   ],
   "source": [
    "print(norm(cg['CORE'] - cg['NMP']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "20.501\n"
     ]
    }
   ],
   "source": [
    "print(norm(cg['CORE'] - cg['LID']))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "19.9437\n"
     ]
    }
   ],
   "source": [
    "print(norm(cg['NMP'] - cg['LID']))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "AdK undergoes a conformational transition during which the NMP and LID domain move relative to the CORE domain. The movement can be characterized by two angles, $\\theta_{\\text{NMP}}$ and $\\theta_{\\text{LID}}$, which are defined between the centers of geometry of the backbone and C atoms between groups of residues [Beckstein2009]:\n",
    "\n",
    "Definition of $\\theta_{\\text{NMP}}$\n",
    "\n",
    "    A: 115-125, B: 90-100, C: 35-55\n",
    "    \n",
    "Definition of $\\theta_{\\text{LID}}$\n",
    "\n",
    "    A: 179-185, B: 115-125, C: 125-153\n",
    "    \n",
    "The angle between vectors $\\vec{BA}$ and $\\vec{BC}$ is\n",
    "\\begin{equation*}\n",
    "    \\theta = \\arccos\\left(\\frac{\\vec{BA}\\cdot\\vec{BC}}{|\\vec{BA}||\\vec{BC}|}\\right)\n",
    "\\end{equation*}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Write a function `theta_NMP()` that takes a Universe as an argument and computes $\\theta_{\\text{NMP}}$:\n",
    "\n",
    "`theta_NMP(u)`: Calculate the NMP-CORE angle for E. coli AdK in degrees from Universe u"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### Calculation of the domain angles of AdK"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "from numpy.linalg import norm\n",
    "\n",
    "def theta_NMP(u):\n",
    "    \"\"\"Calculate the NMP-CORE angle for E. coli AdK in degrees\"\"\"\n",
    "    C = u.select_atoms(\"resid 115:125 and (backbone or name CB)\").center_of_geometry()\n",
    "    B = u.select_atoms(\"resid 90:100 and (backbone or name CB)\").center_of_geometry()\n",
    "    A = u.select_atoms(\"resid 35:55 and (backbone or name CB)\").center_of_geometry()\n",
    "    BA = A - B\n",
    "    BC = C - B\n",
    "    theta = np.arccos(np.dot(BA, BC)/(norm(BA)*norm(BC)))\n",
    "    return np.rad2deg(theta)\n",
    "\n",
    "def theta_LID(u):\n",
    "    \"\"\"Calculate the LID-CORE angle for E. coli AdK in degrees\"\"\"\n",
    "    C = u.select_atoms(\"resid 179:185 and (backbone or name CB)\").center_of_geometry()\n",
    "    B = u.select_atoms(\"resid 115:125 and (backbone or name CB)\").center_of_geometry()\n",
    "    A = u.select_atoms(\"resid 125:153 and (backbone or name CB)\").center_of_geometry()\n",
    "    BA = A - B\n",
    "    BC = C - B\n",
    "    theta = np.arccos(np.dot(BA, BC)/(norm(BA)*norm(BC)))\n",
    "    return np.rad2deg(theta)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "44.124821"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "theta_NMP(u)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "107.00881"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "theta_LID(u)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Processing AtomGroups"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "You can directly write a AtomGroup to a file with the `write()` method:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "CORE = u.select_atoms(\"resid 1:29 or resid 60:121 or resid 160:214\")\n",
    "CORE.write(\"AdK_CORE.pdb\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/anaconda2-4.1.1/lib/python2.7/site-packages/numpy/lib/utils.py:99: DeprecationWarning: `write_selection` is deprecated!\n",
      "This will be removed in version 0.16.0; use `write` method instead\n",
      "  warnings.warn(depdoc, DeprecationWarning)\n"
     ]
    }
   ],
   "source": [
    "CORE.write_selection(\"CORE.ndx\", name=\"CORE\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "(The extension determines the file type.)\n",
    "\n",
    "You can do fairly complicated things on the fly, such as writing the hydration shell around a protein to a file"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```python\n",
    "u.selectAtoms(\"byres (name OW and around 4.0 protein)\").write(\"hydration_shell.pdb\")\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "for further analysis or visualization.\n",
    "\n",
    "You can also write Gromacs index files (in case you dont like make_ndx...) with the `write_selection()` method:"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```python\n",
    "CORE.write_selection(\"CORE.ndx\", name=\"CORE\")\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Trajectory Analysis"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The `Universe` binds together the static topology (which atoms, how are they connected, what un-changing properties do the atoms possess (such as partial charge), ...) and the changing coordinate information, which is stored in the trajectory.\n",
    "\n",
    "The length of a trajectory (number of frames) is"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "98"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(u.trajectory)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The standard way to assess each time step (or frame) in a trajectory is to iterate over the Universe.trajectory attribute (which is an instance of Reader class):"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Frame:     0, Time:    0.000 ps\n",
      "Rgyr: 16.669 A\n",
      "Frame:     1, Time:    1.000 ps\n",
      "Rgyr: 16.6732 A\n",
      "Frame:     2, Time:    2.000 ps\n",
      "Rgyr: 16.7315 A\n",
      "Frame:     3, Time:    3.000 ps\n",
      "Rgyr: 16.7223 A\n",
      "Frame:     4, Time:    4.000 ps\n",
      "Rgyr: 16.744 A\n",
      "Frame:     5, Time:    5.000 ps\n",
      "Rgyr: 16.7185 A\n",
      "Frame:     6, Time:    6.000 ps\n",
      "Rgyr: 16.7741 A\n",
      "Frame:     7, Time:    7.000 ps\n",
      "Rgyr: 16.7764 A\n",
      "Frame:     8, Time:    8.000 ps\n",
      "Rgyr: 16.7894 A\n",
      "Frame:     9, Time:    9.000 ps\n",
      "Rgyr: 16.8289 A\n",
      "Frame:    10, Time:   10.000 ps\n",
      "Rgyr: 16.8521 A\n",
      "Frame:    11, Time:   11.000 ps\n",
      "Rgyr: 16.8549 A\n",
      "Frame:    12, Time:   12.000 ps\n",
      "Rgyr: 16.8723 A\n",
      "Frame:    13, Time:   13.000 ps\n",
      "Rgyr: 16.9108 A\n",
      "Frame:    14, Time:   14.000 ps\n",
      "Rgyr: 16.9494 A\n",
      "Frame:    15, Time:   15.000 ps\n",
      "Rgyr: 16.981 A\n",
      "Frame:    16, Time:   16.000 ps\n",
      "Rgyr: 17.0033 A\n",
      "Frame:    17, Time:   17.000 ps\n",
      "Rgyr: 17.0196 A\n",
      "Frame:    18, Time:   18.000 ps\n",
      "Rgyr: 17.0784 A\n",
      "Frame:    19, Time:   19.000 ps\n",
      "Rgyr: 17.1265 A\n",
      "Frame:    20, Time:   20.000 ps\n",
      "Rgyr: 17.1491 A\n",
      "Frame:    21, Time:   21.000 ps\n",
      "Rgyr: 17.1759 A\n",
      "Frame:    22, Time:   22.000 ps\n",
      "Rgyr: 17.2168 A\n",
      "Frame:    23, Time:   23.000 ps\n",
      "Rgyr: 17.2267 A\n",
      "Frame:    24, Time:   24.000 ps\n",
      "Rgyr: 17.2574 A\n",
      "Frame:    25, Time:   25.000 ps\n",
      "Rgyr: 17.3006 A\n",
      "Frame:    26, Time:   26.000 ps\n",
      "Rgyr: 17.3244 A\n",
      "Frame:    27, Time:   27.000 ps\n",
      "Rgyr: 17.3716 A\n",
      "Frame:    28, Time:   28.000 ps\n",
      "Rgyr: 17.4265 A\n",
      "Frame:    29, Time:   29.000 ps\n",
      "Rgyr: 17.4639 A\n",
      "Frame:    30, Time:   30.000 ps\n",
      "Rgyr: 17.524 A\n",
      "Frame:    31, Time:   31.000 ps\n",
      "Rgyr: 17.5426 A\n",
      "Frame:    32, Time:   32.000 ps\n",
      "Rgyr: 17.5981 A\n",
      "Frame:    33, Time:   33.000 ps\n",
      "Rgyr: 17.6389 A\n",
      "Frame:    34, Time:   34.000 ps\n",
      "Rgyr: 17.6945 A\n",
      "Frame:    35, Time:   35.000 ps\n",
      "Rgyr: 17.7392 A\n",
      "Frame:    36, Time:   36.000 ps\n",
      "Rgyr: 17.8015 A\n",
      "Frame:    37, Time:   37.000 ps\n",
      "Rgyr: 17.8213 A\n",
      "Frame:    38, Time:   38.000 ps\n",
      "Rgyr: 17.8712 A\n",
      "Frame:    39, Time:   39.000 ps\n",
      "Rgyr: 17.9209 A\n",
      "Frame:    40, Time:   40.000 ps\n",
      "Rgyr: 17.9416 A\n",
      "Frame:    41, Time:   41.000 ps\n",
      "Rgyr: 18.0115 A\n",
      "Frame:    42, Time:   42.000 ps\n",
      "Rgyr: 18.057 A\n",
      "Frame:    43, Time:   43.000 ps\n",
      "Rgyr: 18.1017 A\n",
      "Frame:    44, Time:   44.000 ps\n",
      "Rgyr: 18.1168 A\n",
      "Frame:    45, Time:   45.000 ps\n",
      "Rgyr: 18.1822 A\n",
      "Frame:    46, Time:   46.000 ps\n",
      "Rgyr: 18.2343 A\n",
      "Frame:    47, Time:   47.000 ps\n",
      "Rgyr: 18.2669 A\n",
      "Frame:    48, Time:   48.000 ps\n",
      "Rgyr: 18.3362 A\n",
      "Frame:    49, Time:   49.000 ps\n",
      "Rgyr: 18.3647 A\n",
      "Frame:    50, Time:   50.000 ps\n",
      "Rgyr: 18.3847 A\n",
      "Frame:    51, Time:   51.000 ps\n",
      "Rgyr: 18.4327 A\n",
      "Frame:    52, Time:   52.000 ps\n",
      "Rgyr: 18.4965 A\n",
      "Frame:    53, Time:   53.000 ps\n",
      "Rgyr: 18.5229 A\n",
      "Frame:    54, Time:   54.000 ps\n",
      "Rgyr: 18.6312 A\n",
      "Frame:    55, Time:   55.000 ps\n",
      "Rgyr: 18.6261 A\n",
      "Frame:    56, Time:   56.000 ps\n",
      "Rgyr: 18.665 A\n",
      "Frame:    57, Time:   57.000 ps\n",
      "Rgyr: 18.7334 A\n",
      "Frame:    58, Time:   58.000 ps\n",
      "Rgyr: 18.8223 A\n",
      "Frame:    59, Time:   59.000 ps\n",
      "Rgyr: 18.8393 A\n",
      "Frame:    60, Time:   60.000 ps\n",
      "Rgyr: 18.8433 A\n",
      "Frame:    61, Time:   61.000 ps\n",
      "Rgyr: 18.8844 A\n",
      "Frame:    62, Time:   62.000 ps\n",
      "Rgyr: 18.9093 A\n",
      "Frame:    63, Time:   63.000 ps\n",
      "Rgyr: 18.9389 A\n",
      "Frame:    64, Time:   64.000 ps\n",
      "Rgyr: 18.9433 A\n",
      "Frame:    65, Time:   65.000 ps\n",
      "Rgyr: 19.0103 A\n",
      "Frame:    66, Time:   66.000 ps\n",
      "Rgyr: 19.0643 A\n",
      "Frame:    67, Time:   67.000 ps\n",
      "Rgyr: 19.0975 A\n",
      "Frame:    68, Time:   68.000 ps\n",
      "Rgyr: 19.1824 A\n",
      "Frame:    69, Time:   69.000 ps\n",
      "Rgyr: 19.2416 A\n",
      "Frame:    70, Time:   70.000 ps\n",
      "Rgyr: 19.285 A\n",
      "Frame:    71, Time:   71.000 ps\n",
      "Rgyr: 19.2705 A\n",
      "Frame:    72, Time:   72.000 ps\n",
      "Rgyr: 19.2502 A\n",
      "Frame:    73, Time:   73.000 ps\n",
      "Rgyr: 19.2766 A\n",
      "Frame:    74, Time:   74.000 ps\n",
      "Rgyr: 19.3199 A\n",
      "Frame:    75, Time:   75.000 ps\n",
      "Rgyr: 19.3502 A\n",
      "Frame:    76, Time:   76.000 ps\n",
      "Rgyr: 19.3695 A\n",
      "Frame:    77, Time:   77.000 ps\n",
      "Rgyr: 19.4071 A\n",
      "Frame:    78, Time:   78.000 ps\n",
      "Rgyr: 19.4312 A\n",
      "Frame:    79, Time:   79.000 ps\n",
      "Rgyr: 19.4423 A\n",
      "Frame:    80, Time:   80.000 ps\n",
      "Rgyr: 19.4571 A\n",
      "Frame:    81, Time:   81.000 ps\n",
      "Rgyr: 19.4617 A\n",
      "Frame:    82, Time:   82.000 ps\n",
      "Rgyr: 19.4597 A\n",
      "Frame:    83, Time:   83.000 ps\n",
      "Rgyr: 19.5003 A\n",
      "Frame:    84, Time:   84.000 ps\n",
      "Rgyr: 19.492 A\n",
      "Frame:    85, Time:   85.000 ps\n",
      "Rgyr: 19.5407 A\n",
      "Frame:    86, Time:   86.000 ps\n",
      "Rgyr: 19.5403 A\n",
      "Frame:    87, Time:   87.000 ps\n",
      "Rgyr: 19.4934 A\n",
      "Frame:    88, Time:   88.000 ps\n",
      "Rgyr: 19.527 A\n",
      "Frame:    89, Time:   89.000 ps\n",
      "Rgyr: 19.5733 A\n",
      "Frame:    90, Time:   90.000 ps\n",
      "Rgyr: 19.5641 A\n",
      "Frame:    91, Time:   91.000 ps\n",
      "Rgyr: 19.5588 A\n",
      "Frame:    92, Time:   92.000 ps\n",
      "Rgyr: 19.5915 A\n",
      "Frame:    93, Time:   93.000 ps\n",
      "Rgyr: 19.562 A\n",
      "Frame:    94, Time:   94.000 ps\n",
      "Rgyr: 19.5606 A\n",
      "Frame:    95, Time:   95.000 ps\n",
      "Rgyr: 19.5506 A\n",
      "Frame:    96, Time:   96.000 ps\n",
      "Rgyr: 19.5684 A\n",
      "Frame:    97, Time:   97.000 ps\n",
      "Rgyr: 19.5916 A\n"
     ]
    }
   ],
   "source": [
    "for ts in u.trajectory:\n",
    "    print(\"Frame: %5d, Time: %8.3f ps\" % (ts.frame, u.trajectory.time))\n",
    "    print(\"Rgyr: %g A\" % (u.atoms.radius_of_gyration(), ))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The time attribute contains the current time step. The Reader only contains information about one time step: imagine a cursor or pointer moving along the trajectory file. Where the cursor points, theres you current coordinates, frame number, and time.\n",
    "\n",
    "Normally you will collect the data in a list or array, e.g."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "Rgyr = []\n",
    "protein = u.select_atoms(\"protein\")\n",
    "for ts in u.trajectory:\n",
    "   Rgyr.append((u.trajectory.time, protein.radius_of_gyration()))\n",
    "Rgyr = np.array(Rgyr)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The data can be plotted to give the graph below:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAZcAAAEPCAYAAACOU4kjAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAIABJREFUeJzt3Xm8VXW9//HXGycUJ1IkFIWccp5FDbSTU2ikaebwo2up\nDVopqde0rOCWt9LK9HpvpmXOyDVzwCnRq0ckNQcwUEFRExERJyQwmT+/P74LOR7OwbP3XuesPbyf\nj8d+nL3WXsNnr8fhfPjOigjMzMzy1K3oAMzMrP44uZiZWe6cXMzMLHdOLmZmljsnFzMzy52Ti5mZ\n5a6w5CKpr6T7JT0jaZKk07L9PSWNkfScpHskrdfO+YMlTZH0vKSzuzZ6MzNbGRU1zkXSx4GPR8RT\nktYGngQOB04A3o6IC7Kk0TMizml1bjfgeeAA4DXgceDYiJjSpV/CzMzaVFjJJSJej4insvfzgMlA\nX1KCuTo77GrgC22cPgCYGhHTImIRMCo7z8zMqkBVtLlI6g/sAjwK9I6IWZASELBRG6dsAkxvsf1q\nts/MzKpA4cklqxK7CRiWlWBa19N5fhozsxqzapE3l7QqKbFcGxG3ZbtnSeodEbOydpk32jh1BrBZ\ni+2+2b627uHkZGZWoohQJecXXXL5I/BsRFzcYt9o4KvZ+68At7U+idSAv6WkfpJWB47NzmtTRPgV\nwfDhwwuPoRpefg5+Fn4WK3/lociuyAOBocD+kiZIGi9pMHA+cJCk50i9wX6RHd9H0h0AEbEE+A4w\nBngGGBURk4v4HmZmtqLCqsUi4q/AKu18fGAbx88EhrTY/gvwyc6JzszMKlF0tZh1oaampqJDqAp+\nDsv5WSznZ5GvwgZRdhVJUe/f0cwsT5KIGm/QNzOzOuTkYmZmuXNyMTOz3Dm5mJlZ7pxczMwsd04u\nZmaWOycXMzPLnZOLmZnlzsnFzMxy5+RiZma5c3IxM7PcObmYmVnunFzMzCx3Ti5mZpY7JxczM8ud\nk4uZmeXOycXMzHLn5GJmZrlzcjEzq3aLFsE998C775Z/jbffzi+eDnByMTMr0muvwbx5Kz8mAk4/\nHbbYAi65JCWbjnriCRg6FPr0gaefrizWEji5mJkVYfFi+MY3YJNNYJ11YPPN4fDD4eyzVzx29dXh\nxBPhnXfgtNNghx3g5z+H665r+9ozZ8LFF8M++8Cee8LIkbBkCYwb17nfqQVFRJfd7EM3lq4AhgCz\nImKnbN9OwO+AHsDLwNCIWCGlS3oZmAMsBRZFxICV3CeK+o5m1qAWL4b33oP11mv/mDffhAEDYMaM\ntL2sNNKnTyrNtBYBo0fDWWfB1Klp3667wvjxKx77y1/C976X3q+7Lnz963DqqdCvX4fCl0REqEMH\nt3eNApPLIGAecE2L5PIYcEZEjJP0VWDziPhxG+e+BOweEbM7cB8nFzPrWlOmwPbbp0Tws59Bt3Yq\niaZMgdmzYY894PnnYdIkmD8fvvrV9q+9cCFcey1MnpwS0ZlnrnjMtGnw3e/CMcfAkCGw9tolhV/T\nyQVAUj/g9hbJZXZE9Mze9wXuiYjt2zjvH8AeEfGRLVROLmbW5aZMgW23Te+HDoU//jFVbdWIPJJL\ntbW5PCPpsOz90UDfdo4L4F5Jj0v6eteEZmYNb9o0uOCClCwefhgeewwmTFjxuG22Sb27evSA66+H\nnXeG6dO7Pt4CrVp0AK2cCFwi6UfAaGBhO8cNjIiZknqRkszkiGi3pWrEiBEfvG9qaqKpqSm/iM2s\nMfzzn6lx/M03P7y/Vy945JHUk6ulgw+GBx+EQw9NJZmzzoJRo7ou3hI0NzfT3Nyc6zWrqlqs1Wdb\nAddGxN4fcY3hwNyIuLCdz10tZmYdF5G6Bq+9NqhVzdAvfwljxsBGG6U2j5deSsnjiitgzTXbvt6L\nL6Y2lM02g6uvhlWr7f/0K6qHNpf+pOSyY7bdKyLelNQNuBJ4ICKuanXOWkC3iJgnqQcwBviPiBjT\nzj2cXMysbYsWwf33w1/+kqq5Xn8d3ngjNaovWbJiQ3xEerXXQF8n8kguhaVQSSOBJmADSa8Aw4F1\nJH2b1KZy87LEIqkP8PuIGAL0Bm6RFKT4r28vsZiZrdTChXDYYelnSz17tp1ApBVLM9amQksuXcEl\nFzPj8cfTmJC2qqS+8500HuXAA6F//1Tl1aNHl4dYTWq+WqwrOLmYNbC//S0NJhw7Fm64AY49tuiI\nakI9dkU2M6vckiVw3nkwcGBKLOutV9mkj1ay6u+2YGZWijfegKOOgoceSttnnAHDh6dpUKzLOLmY\nWX1Zd12YMydNjXL11XDQQUVH1JDc5mJm9efFF1NV2IYbFh1JTXKDfgc4uZjVqbvvTlPLn3hi0ZHU\nHSeXDnByMasz77+f1kG57rrUZXjq1FQFZrmp6UGUZmYlmzEDjjgijVvp0QN++lPYYIOio7I2OLmY\nWXFmz06LWP3rX6nh/bOfTSsytuWJJ9Jo+pkz02DH229PKzJaVXK1mJkV5957U0Jp+W90q61Stde/\n//uHj504ET71qbSw1k03ubG+E7nNpQOcXMyq3B/+kEojkybBffel0szQoW2vD//kk7DjjjW18FYt\ncnLpACcXsxqyZElKMH36wE4rrMRhXcTJpQOcXMzMSuO5xcysdkyalJYFtobg5GJmnW/6dDjkEPjM\nZ5xgGoSTi5l1riVL4Oij0xiV3Xd3W0qDqCi5SOohaZW8gjGzOnT55fDoo7DxxnDrrdC9e9ERWRco\nqUE/W9v+WGAosCewAFgDeAu4E7gsIl7ohDjL5gZ9swK9/jpss02apfimm+CLXyw6IuuAIhr0HwC2\nAL4PfDwiNo2IjYBBwKPA+ZK+XElAZlZHJkxI1WKHHgpHHll0NNaFSi25rBYRiyo9piu55GJWsOnT\n089NNy02DuuwLhvnog78hZa0U0RMrCSYzuDkYmZWmi6ZFVnSMcAqkgK4EehPqhrbssXPrbKfnpPB\nzMw61ObyHqnRfiHwn8AEYBjwCeBFYBpwGDC4k2I0s1owcWJauMs1BUYZ079I2oDUW2w14AZgSERc\nXvKNpSuAIcCsiNgp27cT8DugB/AyMDQi5rVx7mDgIlJyvCIizl/JfVwtZtbZnnkGmprgrbdg1Cg4\n5piiI7IKFDq3mKR1gG8A+wGnA7MjYnYJ5w8C5gHXtEgujwFnRMQ4SV8FNo+IH7c6rxvwPHAA8Brw\nOHBsRExp5z5OLmad6fnnYb/9YNYsGDwYbrnFY1lqXKFzi0XE3Ij4NWncyyHAHSWePw5onYy2yvYD\n3Ae01Sl+ADA1IqZlvdJGAYeXFLyZ5ePFF2H//VNi2X9/uPlmJxYDSkwuklbIZBHxfkT8D/DN9o4p\nwTOSDsveHw30beOYTYDpLbZfzfaZWVf71rfStC777gujR8OaaxYdkVWJUpc5fkDSn4HbIuKVZTsl\nrQ5sJOlq0kDLq8qM50TgEkk/AkaTOhFUbMSIER+8b2pqoqmpKY/LmtnIkfCTn8B556U17a0mNTc3\n09zcnOs1Sx1E2Z2UAIaSeou9C3QHVgHGAL+NiAklXK8fcPuyNpdWn20FXBsRe7favzcwIiIGZ9vn\nANFeo77bXMzMSlN0g/5qwIbA+xHxbpnX6E9KLjtm270i4s2s0f5K4IGIuKrVOasAz5Ea9GcCjwHH\nRcTkdu7h5GJmVoKiG/QXRcTMChLLSOBhYGtJr0g6AThO0nPAs8CMZYlFUh9Jd2T3XQJ8h1RSegYY\n1V5iMbMKRaTeX4MHp6nyTz+96IisRniZYzNr25NPwhlnwNixy/cdeijceWdxMVmX6JLpX8ysAU2b\nBgMGwNKlsMEG8OMfp7EsG29cdGRWIyoquUjaAZicVVVVJZdczMr09a/D+uvDueemn9YwCm3QzwI4\nFZgL/Ak4ARgXEU9VElDenFzMyhQBFQ1bs1pVaIN+ZmJEXBUR75EmsNy2wuuZWVd65x24/vq2P3Ni\nsQpU2ubSe9mbiLhd0qcrvJ6ZdaZHH00N9W+/nV6jR8PLL8N668GQIUVHZ3Wk0uTyrKQRwLXALGA7\n4MFKgzKzTnLppXDNNR/et8ceqZuxWY4qSi4R8bSk2cC3sl2/rTwkM+s0X/kKrLEGbLRR6gXWty98\n/vOebNJyV870LwcBf42Id1rs3wjoHRGT8g+xMm7QNzMrTREN+r8hzS12i6Q1JXWTtGZEvAFsU0kg\nZpazxYuLjsAaWKnJ5bmIOAL4EvBj0gzIz0qaT5oi38yqwYwZsNVWcPnlXnbYClFqcpkLkJVUXomI\nT0fEJ4B1I+JLuUdnZuU566zUC+yee9yl2ApRapvLVOBmoBnoHhG3tPisV0S8mXuEFXKbizWcBx9M\n69l37w6TJ0P//kVHZDWmiLnF/khas34vYA9J3wVeB/4OfBL4SiXBmFmF3n4bjj8+vT/nHCcWK0zF\nsyJL2oyUbE6OiANyiSpHLrlYQ/niF9M69nvtlWYzXn31oiOyGlQVsyJnyx2/ImlWpdcyswqNGAGz\nZqXlh51YrEBez8Ws3njCSatQNUxcaWbVxonFqoCTi1mtiEgN9osWpe333oMFC4qNyawdZbW5SFoD\n+CLQv+U1IuIn+YRlZiv48Y/hvPNSyWTDDdO+446Diy8uNi6zNpTV5iLpL8Ac4Engg1UoI+LX+YWW\nD7e5WF2YOxf69EmllW7d0vLDANttB489Bj16FBuf1ZXCVqKU9HRE7FDJjbuKk4vVjUmT0oj7734X\n3ngjLfT1yU/CaqsVHZnVmSKTy+XAJdU4C3JrTi5mZqUpMrk8C2wJ/ANYAAiIiKi6FYecXMzMSlPk\nIMpDKrkpgKQrgCHArGVJSdLOwO+A7sAi4FsR8UQb575MavNZCiyKiAGVxmNmZvkpexBllgj2zTYf\nioi/l3j+IGAecE2L5HIP8OuIGCPpEOB7EfGZNs59Cdg9ImZ34D4uuVjtWro0NeCbdaHCBlFKGgZc\nD2yUva6TdGop14iIcUDr5LAUWC97vz4wo70Q8Bgdq2czZ8Lpp8Oxx8LzzxcdjVnJym1zmQjsExHv\nZds9gEdKbXOR1A+4vUXJZRvgHlLyEPCpiJjexnkvAe+SukFfHhG/X8k9XHKx2jF/Ppx7Lvz2t+k9\nwMknw6WXFhuXNZQi21xEi/Et2fs85pw4BRgWEbdKOoo0xf9BbRw3MCJmSuoF3CtpclYSatOIESM+\neN/U1ERTU1MOoZrlLAJOOilNOglwxBEwfDjsvHOxcVnda25uprm5OddrlltyOYO0dsuyxcK+AFwV\nEReVeJ3WJZd3I2L9Fp/PiYj12r1AOmY4MDciLmznc5dcrDaMHw8DBqRFvu69F/bZp+iIrEEVVnKJ\niAslPQgMzHadEBETyrjUsuqvZWZI+nREPCjpAGCFymZJawHdImJeVh13MPAfZdzbrLrsthuMGZNG\n4TuxWI0rbMp9SSOBJmADYBYwHHgO+C9gFWA+qSvyBEl9gN9HxBBJnyCVmIKUHK+PiF+s5D4uuZiZ\nlaDLB1FKGhcRgyTNJf1x/+Aj0iDKdSsJpjM4uZiZlaawEfq1xMnFqtZbby2f3disihQ5zuX8juwz\nszYsWQLXXgv9+8PddxcdjVmnKHcgYlvdgyueEsasrk2dCj/4AfTrB8cfnxruH3ig6KjMOkVJvcUk\nnQJ8C9g8G0i5zDrAw3kGZlZX/u//4MADl29vsQWceWYaIGlWh0pt0F8P6An8HDinxUdzI+KdnGPL\nhdtcrCosXgx77pkGRJ54Iuy7r9e6t6pVaIO+pJ7AVqQZjAGIiLGVBNMZnFysaixeDKuWOymGWdcp\nbBClpK8Bw4C+wFPA3sAjwP6VBGNW15xYrIGU26A/DNgTmJZNib8raSJJM7vvvjSrsVkDKze5zI+I\n+QCS1oiIKcAn8wvLrEY991yacHLXXeHVV4uOxqww5ZbTX5W0PnAraVbi2cC0/MIyq0F//3taf2Xe\nPBgyBDbZpOiIzApTcoO+JAF9l62zIunTpAW+/hIRC/MPsTJu0LdON29emhr/4ovTAMltt4W//Q3W\nWafoyMzKUlhvMUmTImLHSm7cVZxcrNM99RTsvnt6f+qp8NOfOrFYTStysbDxkvaMiMcrublZXdhl\nl1Rq2Wef5UnGrMGVW3KZAmxJamd5j+WzIpe0zHFXcMnFzKw0RZZcPlvJTc3MrL6V2xX5MGBORExr\n+cozMLOqNG0aPPlk0VGYVb1yk0tv4AlJN0oanPUgM6tv778PRx4JAwem5YjNrF1lJZeI+CFpXrEr\ngK8CUyX9TNIWOcZmVj0i4BvfgPHj0/iVPfYoOiKzqlZuyYWslfz17LWYNFvyTZIuyCk2s+qweDFc\ncAFcdx306AG33gof+1jRUZlVtXJ7iw0DjgfeAv4A3BoRiyR1A6ZGRNWUYNxbzCp2+ulw0UXp/Z/+\nBEcdVWw8Zp2syN5iHwOObN2IHxFLJQ2pJCCzqnPAAXDXXWlxLycWsw4pez2XWuGSi1Uswgt7WUMp\ncj2XM9rYPQd4MiKeqiQgs8I89BBstlla474lJxazkpXboL8HcDKwSfb6JjAY+L2k73XkApKukDRL\n0sQW+3aW9IikCZIek9Rml5ys+/MUSc9LOrvM72C23HXXpeqvIUNg7tyiozGreeUml77AbhFxZkSc\nCewObATsR+qa3BFXsuJI/wuA4RGxKzAc+GXrk7JOA/+dnbs9cJykbcr5EmZEpIkm/+3fYNGilGDW\nWqvoqMxqXrkN+hsBC1psLwJ6R8T7kha0c86HRMQ4Sa3qH1hKmr4fYH1gRhunDiD1SJsGIGkUcDgw\npYT4zWDpUvja1+DKK1PV10UXwWmnFR2VWV0oN7lcD/xN0m3Z9ueBkZJ6AM9WEM/pwD2Sfk2aDPNT\nbRyzCTC9xfarpIRjVpobb0yJZa214IYb4LDDio7IrG6UlVwi4qeS7gYGZrtOjognsvdDK4jnFGBY\nRNwq6Sjgj8BBFVwPgBEjRnzwvqmpiaampkovafXg6KPTUsTbbw+HHFJ0NGaFaW5uprm5OddrFtoV\nOasWu33ZVP2S3o2I9Vt8Pici1mt1zt7AiIgYnG2fQ5ow4Px27uGuyGZmJcijK3LZ07/kRNlrmRnZ\nsslIOgB4vo1zHge2lNRP0urAscDoTo/UzMw6rKTkIuna7OewSm8saSTwMLC1pFcknQB8Hfi1pAnA\necA3smP7SLoDICKWAN8BxgDPAKMiYnKl8Vidmz0bRo0qOgqzhlFStZikZ4EDgbuBJj5c6iAi3skz\nuDy4WqzBLV0KP/hB6gm2YAE8/XRqYzGzdhUxQv93wP8BmwNP8uHkEtl+s+qweDGcdBJcc03aPvDA\nlGDMrNOVOyvypRFxSifEkzuXXBrUggXw//4f3Hzz8mnyDzyw6KjMakIeJZeye4tJ2hnYN9scGxET\nV3Z8UZxcGtTMmbD33vDPf6YZjffZp+iIzGpGYclF0mmkxvabs11HAJdHxCWVBNMZnFwa2AsvwLx5\nsMsuRUdiVlOKTC4TgX0i4r1suwfwyLLxKtXEycXMrDRFjnMRsKTF9hJa9Rwz61Lz5hUdgZm1UO7c\nYleS5ha7Jdv+AnBFPiGZddCcOWmq/CuvTI32Dz5YdERmlqmkQX83YFC2+VBETMgtqhy5WqxOvflm\narB/6aW0vf766X3PnsXGZVYHCu0tViucXOrQ/Plp3ZWHH4YddoBzz4UvfAG6dy86MrO6UNgyx2aF\nGjkyJZZNN4V774WPf7zoiMysFScXqz0nnJDaW/bf34nFrEq5WszMzD6ksK7Ikr4kaZ3s/Q8l3Zw1\n8JuZmZU9zuVHETFX0iDSLMlXAJfmF5ZZ5p134Jlnio7CzEpUbnJZNoDyc6RpX+4EVs8nJLPMnDlw\n8MHpNWNG0dGYWQnKTS4zJF1GWgXyLklrVHAtsxX9618wZAg8+WTqYrx0adERmVkJyp1bbC1gMDAp\nIqZK6gPsGBFj8g6wUm7Qr0ELFsBhh8GYMbDJJjBuHPTvX3RUZg2jyHEu/5793EH60P2rLrlYjYlI\n67CMGQO9esF99zmxmNWgcquy3mvxWgIcAvTPKSZrZBLsu2+azmXMGNhmm6IjMrMy5DLOJWtzuSci\nmiq+WM5cLVaDIlIDft++RUdi1pCKnHK/tbUA/yWwfEhOLGY1rqw2F0mTgGXFgVWAXsBP8grKGsi8\nebD22kVHYWY5K7e3WL8Wm4uBWRGxOLeocuRqsSp2883wrW/BnXfC7rsXHY2ZZQrrLRYR0yq5KYCk\nK4AhpMS0U7ZvFLB1dkhPYHZErDCtjKSXgTnAUmBRRAyoNB7rQq+9Btdfn6bKX7QI7r7bycWszpRU\ncpE0LiIGSZpLqhZrmdkiItYt4VqDgHnANcuSS6vPfwW8GxHntfHZS8DuETG7A/dxyaVazJoFQ4fC\n/fenRnuAYcPgN79J7SxmVhW6vOQSEYOyn+tUctPsGuNaVa+1djTwmXY+E54RoPZsuCE8+yysthp8\n7nNw/PFw+OFOLGZ1qKTkIumMlX0eERdWFs4H99kXeD0iXmzvVsC9kpaQ5jb7fR73tU62yipwyy2w\n9dZejtiszpXa5rKsxPJJYE9gdLb9eeCxvIICjgNuWMnnAyNipqRepCQzOSLGtXfwiBEjPnjf1NRE\nU1NTXnFaqfbaq+gIzKyV5uZmmpubc71mub3FxgKfi4i52fY6wJ0RsV+J1+kH3N6yzUXSKsAMYLeI\neK0D1xgOzG2v1OQ2FzOz0hQ5iLI3sLDF9sJsX6nEhzsFABwETG4vsUhaS9La2fsewMHA02Xc2zrT\n9den9pQlSz76WDOrO+Uml2uAxySNkDQC+BtwdSkXkDQSeBjYWtIrkk7IPjqGVlVikvpIuiPb7A2M\nkzQBeJRU8vGEmdVizpzUI+zLX4bRo1Mbi5k1nLLnFsuWNd432xwbERNyiypHrhbrQu++CwccAOPH\nw1prwcUXw0knuTeYWY3Jo1qskuTSE9gK6L5sX0SMrSSYzuDk0kXmzUsrRj7yCGyxBdx1V+oVZmY1\np7AR+pK+BgwjTVb5FLA38AiwfyXBWA3r1i3NEbbZZmmQ5GabFR2RmRWo3N5ik0hdkR+NiF0kbQP8\nLCKOzDvASrnk0oXmz4c33nBiMatxRfYWmx8R87Mg1oiIKaSxL9bIund3YjEzoPxljl+VtD5wK2kQ\n42yg4skszcysPpRcLSZJQN+ImJ5tfxpYD/hLRCxc6ckFcLVYzhYtglGjUjfja66BNdcsOiIzy1kh\nDfoREZLuAnbMth+sJACrEXPnwmWXwUUXpSWIAbbdFn7iNeLMbEXlVouNl7RnRDyeazRWnaZMgcMO\ng6lT0/Z228GZZ6aBkmZmbSg3uewFDJU0DXiPNIVLtLUui9WByZNTYtlxRzj/fPjsZ1PXYzOzduSx\nzPEH8lihMm9uc8nJTTfB4MFe796sARQ6Qr9WOLmYmZWmsBH6VmcWLYLnnoOnnkrTuJx8ctERmVmN\nc3JpZO+/D2ecAVdeCQsWpH0S7LFHepmZlcnJpVFFwKGHwrLV5zbfHHbZBXbdFXqXszSPmdlyTi6N\nSoLTToPp0+HPf4addy46IjOrI27Qb3QLFsAaaxQdhZlVkSInrrRaEQEPPQRvvtn2504sZtYJnFzq\n1YIFcNVVsPvusN9+8LvfFR2RmTUQt7nUo5kz04DHiRPTdq9esM46xcZkZg3FyaXevPkmDBwI//hH\nWm74hz+EY49Na62YmXURJ5d6s+GGcOCBaUDkXXelbTOzLubeYvVoyZK05HCPHkVHYmY1yHOLdUBD\nJhczswrUdFdkSVdImiVpYot9oySNz17/kDS+nXMHS5oi6XlJZ3dd1GZm1hFFdkW+Evhsyx0RcWxE\n7BYRuwF/Bm5ufZKkbsB/Z+duDxwnaZsuiLf6zJoF3/wmvP560ZGYmX1IYQ36ETGuvXVhMkcDn2lj\n/wBg6rK1YySNAg4HpuQfZRV6+WW48Ua47TZ45JE0SHLqVLj//qIjMzP7QFX2FpO0L/B6RLzYxseb\nANNbbL9KSjj174474PDDYenStL366nDQQXDhhcXGZWbWSlUmF+A44Ia8LjZixIgP3jc1NdHU1JTX\npbvWfvtBz56pq/FRR6Xlhj040swq1NzcTPOyGdJzUmhvsaxa7PaI2KnFvlWAGcBuEfFaG+fsDYyI\niMHZ9jlARMT57dyjvnqLvf8+rLlm0VGYWR2r6d5iGWWvlg4CJreVWDKPA1tK6idpdeBYYHQnxti1\nItI8YPfe2/bnTixmVgOK7Io8EngY2FrSK5JOyD46hlZVYpL6SLoDICKWAN8BxgDPAKMiYnLXRd6J\n3n0XvvQlOOUU+Pa30/LDZmY1yIMoq8Wjj6Y5wKZNS+0ol10Gxx1XdFRm1oDqoVrMAC65BAYNSoll\njz1gwgQnFjOraU4u1WCzzdJ8YGecAX/9a5rN2MyshrlarFo8+yxst13RUZiZeeLKjig0ubz+Olx+\neZr+/p//TK/LLoNddy0mHjOzDsgjuVTrIMra96tfwbnnwsKFH97/v//r5GJmdc/JpbNsuWXqSnzE\nEXDMMbDBBqkXWJ8+RUdmZtbpXC3WWZYsSb2/Nt+86+9tZlYBt7l0QKcml4UL4fHHYf31YfvtO+ce\nZmZdzONcutrYsWmyyEGDYJddUlIZNCiNUzEzsw+4zaUUs2fDmDEf3rfddrDppsXEY2ZWpVwttkwE\njBsHV10Fa60Fw4alRvmWZs1Ko+d79EivTTeFXr06JW4zs6K4zaUDOpRcxo5NE0U+/fTyfd26wX/+\nJ5xzTucGaGZWZTzOJS89e6bE0rs3nHRSKqFcfXWa58vMzErWWCWX3/8+lUhOOmnFA8eMgaamtHQw\nwIwZsPHGoIqSt5lZzXG1WAdIinjhBXjooZRUli6F8eM9St7MrB2uFuuolg3zI0Y4sZiZdbLGSC79\n+6deXUceCWefXXQ0ZmZ1rzGqxer8O5qZ5ckj9M3MrCo5uZiZWe6cXMzMLHdOLmZmlrvCkoukKyTN\nkjSx1f5TJU2WNEnSL9o592VJf5c0QdJjXROxmZl1VJEllyuBz7bcIakJ+DywY0TsCPyqnXOXAk0R\nsWtEDOghp5zZAAAGuUlEQVTUKOtIc3Nz0SFUBT+H5fwslvOzyFdhySUixgGzW+0+BfhFRCzOjnmr\nndOFq/RK5n88iZ/Dcn4Wy/lZ5Kva/kBvDewn6VFJD0hqb+bIAO6V9Likr3dhfGZm1gHVNkJ/VaBn\nROwtaU/gRqCtRegHRsRMSb1ISWZyVhIyM7MqUOgIfUn9gNsjYqds+y7g/Ih4MNt+AdgrIt5eyTWG\nA3Mj4sJ2PvfwfDOzEtX6xJXKXsvcCuwPPChpa2C11olF0lpAt4iYJ6kHcDDwH+3doNIHZGZmpSuy\nK/JI4GFga0mvSDoB+COwuaRJwEjg+OzYPpLuyE7tDYyTNAF4lFTyGbPiHczMrCh1P3GlmZl1vWrr\nLZYbSYMlTZH0vKSGmmdfUl9J90t6JhuMelq2v6ekMZKek3SPpPWKjrWrSOomabyk0dl2Qz4LSetJ\n+lM2UPkZSXs18LM4XdLTkiZKul7S6o3yLNoaxL6y7y7p+5KmZr83B3fkHnWZXCR1A/6bNEhze+A4\nSdsUG1WXWgycERHbA/sA386+/znAfRHxSeB+4PsFxtjVhgHPtthu1GdxMXBXRGwL7AxMoQGfhaSN\ngVOB3bIORasCx9E4z2KFQey0890lbQccDWwLHAL8Vvro9d/rMrkAA4CpETEtIhYBo4DDC46py0TE\n6xHxVPZ+HjAZ6Et6Bldnh10NfKGYCLuWpL7AocAfWuxuuGchaV1g34i4EiAiFkfEHBrwWWRWAXpI\nWhVYE5hBgzyLdgaxt/fdDwNGZb8vLwNTSX9jV6pek8smwPQW269m+xqOpP7ALqTOD70jYhakBARs\nVFxkXeo3wFmkwbfLNOKz+ATwlqQrsyrCy7Pelw33LCLiNeDXwCukpDInIu6jAZ9FCxu1891b/z2d\nQQf+ntZrcjFA0trATcCwrATTuvdG3ffmkPQ5YFZWkltZUb7unwWp6mc34H8iYjfgPVJVSCP+XqxP\n+p96P2BjUglmKA34LFaiou9er8llBrBZi+2+2b6GkRX1bwKujYjbst2zJPXOPv848EZR8XWhgcBh\nkl4CbgD2l3Qt8HoDPotXgekR8US2/WdSsmnE34sDgZci4p2IWALcAnyKxnwWy7T33WcAm7Y4rkN/\nT+s1uTwObCmpn6TVgWOB0QXH1NX+CDwbERe32Dca+Gr2/ivAba1PqjcR8YOI2CwiNif9HtwfEf8G\n3E7jPYtZwPRsgDLAAcAzNODvBak6bG9J3bPG6QNIHT4a6Vm0HsTe3ncfDRyb9ab7BLAl8JFLndTt\nOBdJg0k9Y7oBV0REm2vD1CNJA4GxwCRS0TaAH5B+IW4k/S9kGnB0RLxbVJxdTdKngTMj4jBJH6MB\nn4WknUkdG1YDXgJOIDVsN+KzGE76D8ciYALwNWAdGuBZZIPYm4ANgFnAcNIMKX+ije8u6fvASaRn\nNawjA9frNrmYmVlx6rVazMzMCuTkYmZmuXNyMTOz3Dm5mJlZ7pxczMwsd04uZmaWOycXs4+QTVN/\nSovtPpJu7KR7HS7ph2Wct5qkB7MZwc0K53EuZh8hm/zz9ojYsQvu9Vfg8xHxThnn/gh4MSJG5h+Z\nWWn8vxyzj/Zz0vLb4yWdn00rNAlA0lck3ZItsvSSpG9ni1CNl/RwNkEikjaXdLekx7MSxtatbyJp\nK2D+ssSSzV58aXbOlGwSTiRtJ+lv2T2ekrRFdonbgKFd8UDMPsqqRQdgVgPOAbbPZhJGUj8+PGPs\n9qRlDdYCXgDOiojdJF0IHA/8F3A58M2IeFHSAOBS0nxWLQ0Exrfa1y8i9pS0JfBAlkhOBi6KiBuy\nCUpXyY59Gtgzn69sVhknF7PKPRAR/wL+Jeld4I5s/yRgR0k9SDPu/qnFCn6rtXGdPsCbrfbdCBAR\nL0h6EdgGeAQ4N1sE7ZaIeCE7ZqmkBZJ6RMR7eX5Bs1K5WsyscgtavI8W20tJ/4HrBsyOiN0iYtfs\ntUMb13kf6N5qX8sSkoCIiBuAzwPzgbskNbU4Zo1sv1mhnFzMPtpc0my5ZYmIucA/JB21bJ+kndo4\ndDKwVat9X1KyBWklyeckfSIi/hERl5DaWXbKrvkx4K1sfRKzQjm5mH2ErIH9r5ImSjr/ow5vZ/+X\ngZOyBvinSeuStzaW1HbT0iukpRLuJLXZLASOlvS0pAmk9p5rsmM/kx1nVjh3RTarIpJ+Q+r2fL+k\nK7P3N3fw3D8DZy9rgzErkksuZtXlZ6ReZ1DCGuaSVqNF475Z0VxyMTOz3LnkYmZmuXNyMTOz3Dm5\nmJlZ7pxczMwsd04uZmaWOycXMzPL3f8HztPDWs5nj1AAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x7f54a41d0410>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "# quick plot\n",
    "import matplotlib.pyplot as plt\n",
    "ax = plt.subplot(111)\n",
    "ax.plot(Rgyr[:,0], Rgyr[:,1], 'r--', lw=2, label=r\"$R_G$\")\n",
    "ax.set_xlabel(\"time (ps)\")\n",
    "ax.set_ylabel(r\"radius of gyration $R_G$ ($\\AA$)\")\n",
    "ax.figure.savefig(\"Rgyr.pdf\")\n",
    "plt.draw()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The graph of $R_{G}(t)$ increases over time, indicating an opening up of the AdK enzyme."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Exercise 4"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "1. Take the functions to calculate $\\theta_\\text{NMP}$ and $\\theta_\\text{LID}$ and calculate the time series $\\theta_\\text{NMP}(t)$ and $\\theta_\\text{LID}(t)$. Plot them together in one plot.\n",
    "2. Plot $\\theta_\\text{NMP}(t)$ against $\\theta_\\text{LID}(t)$. What does the plot show? Why could such a plot be useful?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAbgAAADOCAYAAABMxhfbAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAIABJREFUeJztnXd8FVX2wL8nEKQXkQ4BRRAQQVDRFcQIiqgsrrq6uFgQ\nFH6LYsG14CJi2bVX1kJRsGBZQQQLLipFERVpAkIAQamhKQuhl5zfH3fey0vyXshLXsvL+X4+82Hm\n3pl7zzwmc+bee4qoKoZhGIaRbKTEWwDDMAzDiAam4AzDMIykxBScYRiGkZSYgjMMwzCSElNwhmEY\nRlJiCs4wDMNISkzBGYZhGEmJKTjDMAwjKYm5ghORV0Vki4gszlNeTUTeF5HlIvKTiJzplXcXkQwR\nWSki98RaXsMwjNJGAe/pX0XkRxFZKCJzA8oT8j0tsY5kIiKdgN3AG6raJqB8HDBLVceKSFmgonfe\nSqArsAn4AeilqhneNXWAlsBBVZ0T0xsxDMNIUgp4T68BTlPVHQFlKSToezrmIzhVnQ3sCCwTkarA\nOao61jvnsKruAjoAq1R1raoeAt4FLg24tJWqzgSyRaRCTG7AMAwjyQn2nvYQ8uuNhH1PJ8oa3PHA\ndhEZKyILRGSUiFQEGgDrA87b4JX5WCYiXYEUVd0XQ3kNwzBKIwp8LiI/iMhNXlnCvqfLxrKzAigL\ntAduVtV5IvIccC/wY0EXqeoWYEtB54iIRZM2oo6qSjTatefXiDZhPrsdVTVTRGrhFN3yQrR/1Pd0\ntEiUEdwGYL2qzvOOJwDtvPK0gPMaAhvDbVxV47I98MADparf0nrP0aak/F7R+j+wdqPXbhGexUzv\n323AJNz05EYi8J6OBvFScOJtgF/DrxeR5l5RV2AZMA84UUQai0g5oBcwJdbCGoZhlEJyvadFpKKI\nVPb2KwHdgCU4o5KEfE/HfIpSRN4G0oGaIrIOeECdccmtwHgRSQXWADeo6hERuQWYhlPGr6rqUYfE\nhmEYRtEJ9p4GZgKTvGnzssB4Vf3cOz8h39MxV3Cq+tcQ5T8CZwQp/ww4KdpyRYP09PRS1W88+47n\nPZdEovF7Rev/wNqNbrvBCPWeBk4NcX5Cvqdj7gcXa0REk/0ejfgiImgUjUzs+TWiRTSf3UQgUYxM\nDMMwDCOimIIzDMMwkhJTcIZhGEZSYgrOMAzDSEpMwRmGYSQwBw4ciLcIJRZTcIZhGAnIzJkzOffc\nc6lQoQK33HJLvMUpkZibgGEUE3MTMIrLwYMH6dOnD++88w5NmzalWrVqLFiwINc5p512Gj169KB7\n9+40btyYgwcP0rhx42L1m+xuAqbgDKOYmIIzikN2djZlypQp0rV33HEHzzzzTJH7DvXsisirQA9g\niwbkg/PqUnBhFDeoak+v7FdgJ5ANHFLVDkUWKoLYFKVhGEYcmT17dtDyMmXK0Lx586B1Pp599lna\ntm3Lli0RD9Y/FrgwRN1tuFjBgWQD6araLq9yE5E6IpIuImdHWsijYQrOMAwjjkycODFf2XXXXUdG\nRgYrVqwgOzublStX8sorrwS9fvHixdStW5e0tDQ2boxMEH8NkfBURBoCFwNj8lYRWp/ELeFpouSD\nM4wctm6FmTNh1SqoUgUaNIDjjoMvvoDp02HxYjjhBDjnHGjeHGrVgpYt4dSgYfJCs3cvrF4NW7aA\nKtSvD61auX5TU+H446Nye4bhIzs7mw8++MB//PHHH9OpUyeqVavmLxMRmjVrdtRR2vr162nYsGGB\naXBmzpzJzJkziyPys8BdQLU85b5EqEeAUao6OqDOl/B0n8Y6MXU08hYl0uZu0SgxTJ+uWr26qlM5\n4W2jRqmOHq3aoYPqSSepHnecau3aqiNGqO7Zo/rVV6oDBqj27KnapYtquXL526hUKWe/f3/V999X\n/fpr1SNHVNetc21lZ+cS2XvG7Pk1wmLGjBl6yimnqKcctEaNGnrw4MGQ57/yyiv+c0NtV199dVgy\nFPTsAo2BxQHHlwD/9vbTgY8C6up5/9YCFgGdQrUby82MTIz4k50NI0fC5MluhHboEJx9NnTsCHv2\nwPr1sHEjnH46XHoptG8Py5fD/PluBLZpE0yZAiJONRUWETcCrF/f7S9bBps3w7HHwq5dcPhwzrn1\n6rk6VZg7F844I6AZMzIxCo+q8uSTTzJkyBCys7P95X//+9958sknQ143Y8YMunTpUmDbWVlZVK5c\nudCyFPTsikhjnBJr4x3/C7gGOAxUAKoAH6jqdXmuewDIUtWiW79Einhr2Ghv2Bdw4pKdrfrjj6oX\nXaS5RlGDB6sePhxeW3ff7a4tV0515EjVn35SzcxUnTBBtV491bJlVZs1U733XtUPPlD98EPVrVvz\ny7Nhg+t76VLVfv1UL79ctVGjnLb/8hfVJUtyXYaN4IxCsmvXLr3iiityjboqVaqk9913X4GjNx9z\n587Vb775RlVVr7/++lztrFq1Kmx5Cnp2gSbAkhB15wJTvP2KQGVvvxLwDdAtVLux3GwEZ8SOXbtg\n5UqYMcOtsS1cCJmZru7YY+Hpp+H886Fhw/Dbzs6G8eOhTRto2zZ/nSoU0RSbI0dg0SJo3NitBebB\nRnBGYfj555/p2bMny5fn5ALt2LEj7733Hg0aNAi7vUOHDjFy5EgqV67M9ddfj0j4j2ABbgL+hKfA\nFnISU/vqzwXuVNWeInI8MAmnaH2JUB8LW5goEHMFVwT/iu7Ac+Rkin08zP7sBRFv/vtfeOwxp9Ty\nUru2m3a8/35o1CjmokUCU3BGYUhPT2fWrFn+40GDBvHUU09Rrly5uMlkjt6R7lCkE7AbeCOIgrsD\nOA2o6n0ZpAArga7AJuAHoJeqZnjn1wFaAgdVdU6I/uwFEU8++ACuuMLtlysHzZq5tbQLL3T/Nm0K\nKSXbW8UUnFEQP//8My+++CLPPfdcrvLDhw8X2cE7UiS7gou5m4CqzvYWL3MR4F/xT2CwV9wBWKWq\na71z3gUuBTK8+laqOkNEzhKRChprE1SjYH7+Gfr2dft33QX/+AdUy2tdbBjJx6FDh8jIyGDKlCk8\n8sgj7N+/P1d9uXLl2Lt3L1WqVImThKWDRPp09vlXBH6uNgDWBxxv8Mp8+PwrUky5RZivv4aLL4Zp\n0/LXrVsHBw7AwYPwzTfOnwxg3z4YMwZ694ZOneCkk2DnTujZEx5/3JSbkfR8+umndOjQgUqVKtGm\nTRuGDh2aS7mJCGeffTbvvvuuKbcYkBCO3iJyCW5NbpGIpBf2OlXdglsALZDhw4f799PT00lPL3QX\npZNFi+CSSyAryym4UaNyRmLPPw+33w4VKzpn6J07nYP17be7Udq2bTntlC0Lf/oTjB7tzPCThAg4\nyxpJhqry2GOP8Y9//INgU8rt2rXjkUceoXPnzmGZ8RvFIy5WlIX1rwBeAoaranfvvHtxZq2FNjSx\nNYww2L7dWTK++KJTbq1aOd8wgLffdgrtqqty+5pVrJgzggO3rnbjjS7SyGmnOevIJMfW4Iw77rgj\n3xpbkyZNaNOmDRdeeCE33XQTqampcZIuNMm+Bhcv37QmFM6/ogzwM86jvhzOQ75lmH2pcRS2bVN9\n6CHVqlXV74t22WWq+/erPv54Tplve/BB1U2bVNescb5mbduqpqaqPvtsvigfpQHMD65Us2bNGhUR\nvz/aueeeq5mZmfEWq1BE89lNhC3mU5SB/hUiso48/hWBqOoREbkFmEaOm8DyYOcaYbJ+vTPPz8hw\nkUN8UTsuvBCGD4ezznLHd93lYjU+84yL+ThoEAwdmnvKcd482L0bqleP+W0YRrwZOXKk72OEzp07\nM23atLia/icTItIaGILzx/sN+Jeq/lTo633/McmKTfEEYe9eZwSycKE7FnEGJYMHQ7BQQKqwYoWb\ndrQ/3HzYFGXpZf/+/TRq1Ijt27cDMGXKFP74xz/GWarCk+hTlCIyQFVHBhzfpLkDORdIQhiZGFFg\n7Vo34vKtkW3d6ralS90627Jlzgdt1iyoWtVF7Q+FCLRoETvZDaOEMGHCBL9yS0tL4+KLL46zRJEh\n1gE5CiBv3rmwZvBMwSUjL78MAwc6xXXyyS448JEjuc9p3NgFKC5CiCDDMJz9wksvveQ/HjBgQNwd\ntyPIWGAE8EaQOl/C06rgV3j/JiAgh4hM1jACchTAvSJyLS7fHEB9XFaDQmEKLhlYvRp++skptOnT\n4eGHXfmuXfDtty4GY1oa1KzpFFv37tCnDxxzTFzFNoxosHTpUho0aECNGjWi0r6qsmrVKm6//Xa+\n/fZbAFJTU+nXr19U+osHmjgBOR5X1a8C+u8czn2YgiuJHDrkYjsuWgQVKsC77+YfoT35pFNkv/zi\nEoOaAYiR5KgqN998My+//DI1a9Zk6tSpnBGQ1qg4fPPNNzz00EOsXr2aDRs2cODAgVz1119/PXXq\n1IlIX9EkSglPgwXk6BBwXJyEp/298I4+WgJfhTo5H/E244z2RjKZWR84oPrll6rnnqu5zPZTUlwC\nz7POUu3bV/Wjj+ItaakCcxNICIYOHZorfUzVqlV1zpw5Qc/Nzs7Wl19+WR966CFdsWKFv3zr1q06\nbtw4veKKK7Rjx4768ccfq6pq27ZtgyYYFREdOHCg7tmzJyb3GGkKenY5esJTnzvXFbgs3r7zrgFe\nCNVuOBtwlSeHb/tLWNdHQohE3kr8C2LHDtXJk1VvuUX12GNzlFq9eqqvvOL81JYujbeUpRpTcPFn\nxIgRQRVQ5cqVddasWfnOf++993Kd16VLFz377LNz+bMBeuyxx+rBgwe1W7du+ZTneeedp999910c\n7jZyhKng/gWsA9YAmXhB84GzgM8CzrsXuCdUu7Hc4i5A1G+wJL8gvv8+t1ID1VatVO+6yzlaGwmB\nKbj48t577+VSTOnp6VqrVi3/ccWKFXXmzJm5rrnggguCKsRg2xNPPKENGzbMpQyThaMouCbEKCBH\nYTbgD8DLYV0TaSESbStRL4jPPlNt00a1enXVtDTVSpXcf1G7dqpDhqguXBhvCY0gmIKLDzt27NAR\nI0ZoamqqX/mcddZZunv3bv3pp5+0bt26/vL69evr/v37VVV1w4YNuRRiSkpKrv1OnTppp06dgiq7\nlJQU/fDDD+N855Ej1LMLvI2ziDzgjdpuyFPvV3DecXdgBbAKuDdYm0XZgNp5jtPDuj5SgiTqVmJe\nEBs3OsWWNyxWr16qhUhlb8QPU3Cx4/Dhwzpt2jS9+uqrtXz58rmUT4sWLXT79u3+c1esWJFrJDdq\n1ChVVX388cf9ZV27dtV169bpK6+8om+99ZZu27ZNVVVnzZqVT7lVq1ZNP/vss7jcd7SI5rMbiQ14\nC7jP2/4BjAnr+njfQAx+oBD/tQnEzz+rXnih+++4+GIXGzIjQ3X+fNUjR+ItnXEUTMHFhsmTJ2uj\nRo2CjqzS0tJ07dq1+a554okn/OeceOKJunPnTm3ZsqW/7PXXXw/a1+HDh3MpxxNPPFGXL18e7VuM\nOYmm4ICmwD04N4S/4Sw208gxMvlzOO1ZqK54889/utiO4PzYli0z5+sShoXqij5HjhyhXr16bAtM\nx4RLQ9OnTx+uv/56qgXJN5iVlUVaWhr/+9//AKhatSq7du0CoGLFimzZsiVk+prx48dzxx130Llz\nZ0aNGsWxSZgZI9FCdYnIjcBMnFIbCJQH6gLPqur4cNtLpISnpY8NG3Kcsv/yF/jiC1NuhhGE2bNn\n+5VbhQoVuO2221i0aBELFizg1ltvDarcAKpUqcKgQYP8xz7lBjB48OACc7P17t2bLVu2MGHChKRU\nbglKGaCyqk7HpVTrAZwNHBGRv4XbmI3g4snNN8NLL7kca++9F29pjCJiI7jIcuDAAVJTUzlw4AAf\nfPABr732GtOnT/fX//Wvf2X8+MJ/zG/fvp3GjRuz18tbWKdOHf71r3/Rp08fUlJK9zd+Ao7gBLgV\n+COwC3gN55ZwALhcVZ8Mq71k/+NJyBfEtm0u4PEzz7g0NUuXuuSiRonEFFzkePDBB3n44YcREVJT\nU9m3L3/gi4kTJ3L55ZeH1e7777/PM888Q5cuXbjnnnuoWrVqpEQu0SSagvMhImWBLrjRWz1gO/CO\nqi4Nq51k/+NJuBfEkiUuNc2GDe749tvh2WfjK5NRLEzBRYbly5fTunVrsrOz89WlpKRw4YUXMnDg\nQHr06BEH6ZKTRFVwkcJiUcaCXbtczMipU+GaayAryyUUfe45OPPMeEtnGAnB8OHD8ym3E088kb59\n+3LdddfRwNanjTCxEVykWb8evvwSVq6EevVctus334Ty5cE33fKXv8C4ca7MKPHYCK74LF68mLZt\n2/qPv/rqK5o1a0adOnUQSdoBRtyxEVyECZZIz0vB8AZQB8gGRqvqC15dtBLpRZadO+Hvf3eK6/Dh\n3HUpKU65paTAo4/CXXe5JKKGYQDwwAMP+Pd79uzJOeecE0dpjBDv6WNwkfzL4XTHBFV90Kv7FdiJ\ne38fUtUOwdqNNTEfwXmpD3YDbwT8cHWBuqq6SEQqA/Nx+YRWeps/kR7QS8NIpBeTL2BVuOIKmDTJ\nKbEePaBtW9i0CcqVgzvvdD5uInDccdGVxYg5NoIrHu+++y5XX321/3jRokW5RnNG9Aj17AZ7T3vl\nFVV1r4iUAb4BblXVuSKyBjhNVXcEaSvshKeeNWVv4ARVfUhE0nA6Ym449xfzEZwGSaSnqpuBzd7+\nbhFZjssxVJ3oJdIrPvv3u3W1OXOccqtaFb75Blq3jqkYhlESOXjwIH//+98ZMWKEv+zKK6805ZYA\nBHtPe+V7vd1jcPrD9/UlhParLsp7+iXcaLAL8BCQBUwEwkrwl3BGJiLSBDgV+B64kAgk0hs+fLh/\nPz09nfT09OILumqV819btCinbORIU26lgAgkjSxVrF+/nieffJKTTz6ZSpUqMWvWLJYuXcpPP/1E\nVlaW/7wTTjiBp59+Oo6SJj/FfXZFJAU3w9YUeFFVf/CqFPhcRI7gcsONDrisKAlPz1TV9iKyEEBV\nd4hIubAFjlO8sVx5hgLKKwPzgEu942In0iPSsfyys1XHjFGtXFkVVI8/XrVfP9W33opsP0aJAYtF\nWSCFSU1z2WWX6f/+9794i1rqKOjZDfWe9uqqAtNxozOAet6/tXDpcjqFarcwG26AUwZYENDuwnDb\nSZgRnOfYNwF4U1Une8UbcTHJfDT0yuLD4sXOb23GDHd81VUwahSECBNkGKWd1atX8/nnn4esr1On\nDvfddx+DBg0ya8kShKruEpEZuDQ5y1Q10yvfJiKTcDNts4vRxQvAJKC2iPwT+DMwNNxG4qXgxNsC\neQ33Qz0fUPYDcKI3F5wJ9AKuJtZkZcF997mwWtnZUKMGvPAC9O6dlNaQTZo0Ye3atfEWI+Fo3Lgx\nv/76a7zFKFG89tprQcuvvfZannjiCXMDSGxyvadF5DicheROEakAXAA8JiIVgRR19hOVgG7Ag8Xp\nWFXHi8h8nIGhAH9S1eVh34DG3orybSAdqAlsAR7AWUp+BSwhZ9riPlX9TEQuIrebwGNh9qfFuscD\nB+CCC+Drr52F5MCB8OCDkMTBVz3LqniLkXCE+l3MijI4hw8fpnHjxmzatAmAe++9l1mzZlGlShXe\nf/99C5eVABRgRRnsPT0PeB33Lk4B3lPVf4rI8bjRluIGTePDfU9HC3P0DsacOTB3LmzdCt9/D9On\nuyj/n3zizP+THFNwwTEFVzCqyrhx4/jhhx8QEb788ktWrFgBQO3atdmwYQOpqalxltIIJNEcvUUk\nC6cofTIFWmmqqob1VZQwa3AJw9ix0Ldv7rLKlUuNcjOMopCdnc2KFSvom/dvx+OGG24w5WYcFVWt\nEsn2TMEF8vHH0L+/27/2WmjWDKpUccGRmzePr2yGkaBkZmbSsWNHfvnll3x1VapU4frrr2fo0LDt\nA4xSjIgMDlK8E5ivqouC1AXFFBy4MFoDBriYkQCDB7t0NoZhHJWPPvooqHIbPXo0vXr1KjCpqGGE\n4HRv+8g77gEsBv5PRN5X1ScK04gpuOxsuP56eP99F/x42DC4++54S2UYJYbt27fnK+vevTs33nhj\nHKQxkoSGQHtV3Q0gIg8AnwCdcY7mkVFwItIU54NQFRdJ5NtwhogJz9ChTrlVqeIsJW2drUQwZ84c\nMjIy2LFjB/369aN69erxFqnUsmNHTvjB+vXrc91113HzzTfHUSIjCaiNy+Lt4xBQR1X3iciBENfk\nozD52s/DxQD7EueT8IiIzBOR3uFIm5D85z8uun+ZMjBhgim3EsLq1asZN24cffv2JS0tjQkTJsRb\npFLLoUOH+OSTT/zHQ4cO5dFHH6Vhw4ZxlMpIAsYD34vIAyIyHJgDvO352S0rbCOFUXBlgMqqOh34\nSFV74NKIHxGRv4Uvd4KwcCHccIPbf+YZ6NYtvvIYheaee+5hwIABAKxcuTLoFJkRGz766COWL8/x\nv73gggviKI2RLKjqw0B/4H/A78AAVX1IVfeoaqEHV4VRcKOAc0XkC+BSEekBnIiLMlIyV49XrnSW\nkXv3uvW3QYPiLVGJQSRyW1HIzMxk7ty5zJ8/n5EjR/LWW2+RlpZ29AuNqJDXWdtGbsmBiLwqIltE\nZHFA2TEi8r2ILBSRJd66mK+uu4hkiMhKEbknAv0fAzQHKuGyylwsIsPCbeeoCs6Lyfk8LubYKzjL\nltuAvsDUcDuMG6rw1lvO3P+kk2DzZjjvPBdL0kIFlRimT59Ojx496N+/P9deey2bNm0iLS2NVq1a\nMXeuSxU1fPhw7r//fqZPn06zZs1yhR0bOXIkPXr0YMOGDf76KVOm8O677/LUU0/F67ZKLB07dsx1\nvHr16jhJYkSYsbhsLn5U9QBwnqq2w2V8uUhEOngZBv7tnX8ycLWItPBdJyJ1RCRdRM4Oo//JuNRo\nh4E9AVtYFNqKUlUPA9O8rWTRvLlTaL7UHFWqwIUXwujRLiGpUWjiHVRj/fr1tGzZEoDJkydz8cUX\n06lTJ4YPH86QIUP473//S9++fTl8+DAnnHAC/fv354033uD+++8H3JpRhw4daNiwIQ0bNuT444+n\nZ8+eAFx11VUsXbqU1pbyqNDMmzfPv9+oUSNatGhRwNlGSUHDywfXgcjn7Wyoqt2Lex+FmaIs+axa\n5ZRb/fouUsmOHc5y0izvShy1atWiYsWKALz55ps8/vjjANSsWZMBAwbw8MMP5zq/ffv2LPJy9i1Z\nsoSTTz45V31gGKy9e/dSu3btaIqfdPzwww/+/a5du1KmTJk4SmNEGxFJ8XK0bQY+V5cPrgH583Y2\nCDj25YNLKaRyA5gjIqcUV97S4Qe3dKlTbtWr23RkCefKK6/k/vvvZ8yYMQwbNsy//iYiXHXVVVx1\n1VUsXryYVq1a+a8566yz+O6770hNTaVcnhH7vn37+Oqrr8jKyuLRRx81BVdIZs6cyTXXXMPGjTnZ\nq844I6xky0YcKG7CU1XNBtqJSFVgkoi0KsQ1W3ABm8OhE9BHRH7BuQv4YlG2CaeRQis4cTktegMn\nqOpDIpIG1FXVueF0GBfyfLUbJZeqVavy/PPP5yv3jcSee+45LrroIiZNmuSv6927N0OHDuWuu+5i\n//79uUZtFSpUoHPnztEXPMm48847cyk3MAVXEkhPTyc9Pd1//OCDRctq4+WDm4mzzZhD5PN2XlTM\n64HwpihfAv5ATj62LODFSAhhGMVh6tSpjB07lt9++4369etz6623IiL+8nLlylGtWjWaNm3Kxx9/\nzLx58/xGJr/88gvTp0+P9y2UOLZu3ZrruFq1arRpE9bHtZH45MsHJyLVvH1fPrjlBOTtFJFyuLyd\nU4rTsbeetwuog8ss7tvCu4HCpuIQkQWq2l5EFnpWNIjIj6qa0N7RJSndSKJg6XKCY+lycmjfvj0L\nFy4EoEWLFrzwwgvmA1cCiUQ+OO/87sDzFDFvZ5D+b8RZ6zcEFgFn4aJodQmnnXDW4A6JSBm8/Dwi\nUgvIDqczwzCSg+OOO86//8wzz5hySzJU9a8hqtqHOP8z4KQIinAbcAbwnaqe57kd/CvcRsKZonwB\nl7W1toj8E5hdlA6DORB65UEdBSPtQGgYRvGpVauWf98iyRhRYL+q7gfn9K2qGRRBgYbjBzdeRObj\n4lEK8CdVXX6Uy4IxFhgBvOErCHAU7ApsAn4QkcnAymDl3s0iInWAlsBBVZ1TBFkMwygCgSO4bdu2\nxVESI0nZICLVgQ+Bz0VkB7D2KNfkIyw3AU+xZBz1xILbCOZAGMpRcFaI8uI4EBqGUUyys3NWJ0zB\nGZFGVS/zdoeLyAygGvBZuO0UJl1OFm7dzbcQ6Vvx9vklVA16YXgEcxTsUEC5D58D4b6ClNvw4cP9\n+3nNZA0jXIrrS1TS+e6773jxxRwD6po1a8ZRGiPZUdVZRb32qApOVasUtfFoU1gHwkAFZxjFJVK+\nRCWVt99+229NeuaZZzJw4MA4S2QYwQnH0XtwkOKdwPwIJEDdSHBHwVDlhmHEicWLc+zDhg4d6g+d\nZhiJRjhWlKcD/4ebNmwADMB5sY8WkbvD7DeXAyHBHQUnhygvlgOhYRhFR1VzKThz7jYSmXAUXEOg\nvareqap3Aqfh0op3BvoUthHPgXAO0FxE1onIDap6BBiEy1TwE/CuqmZ45bfkKS+K5aZhGBFg06ZN\n7NixA3DRSxo1ahRniYxkInCwJCJX5qkL2y0tHCvK2riglz4OAXVUdZ+IHAhxTT5CORCq6lSC5JeL\nggOhkQTMmTOHjIwMduzYQb9+/ahumSFiQuDo7ZRTTkEseHlSIiKvAj2ALb4AxyLSEOfeVQcX5GO0\nqr7g1f2KW7LKBg6paodg7RaCXsAT3v4Q4P2Auu7AfeE0Fs4IbjzwvYg8ICLDcaOwt0WkErAsnE4N\nozisXr2acePG0bdvX9LS0nj//fePfpEREWx6stSQL+EpLvnoYFU9GReX+OaAxKbZQLqqtsur3MJM\neCoh9oMdH5VCKzhVfRjoD/wP+B0YoKoPqeoeVe0dbseGUVTuueceBgwYAMDKlSv57bff4ixR6WHJ\nkiX+/VPaTS3KAAAgAElEQVROKXa6LiNBUdXZwI48ZZt9BoWquhsXaNmX900IrU9aqepMINsL0lxg\n1yH2gx0flUIrOBE5BmgOVAKqAxeLyLBwOzRKOCKR24pAZmYmc+fOZf78+YwcOZK33nrLnxPOiD42\ngjMARKQJcCrwvVekuIgjP4jITXlODyfhaVsR2eX5X7fx9n3HYX9RhbMGNxnPLYDca3GGETOmT59O\njx496N+/P3v37uXuu+8mNTWVpk2b8tRTT3HZZS4AwrRp0xgwYABjxoxBRBgwYABPP/00v/32Gxs2\nbOC+++6z7NNhoqqsXLnSf5w3O7qR+EQiSIGIVAYmALd5IzmAjqqa6QXh/1xElnujwLASnqpqRP8o\nw0mXs1RVW0ey81iQqOlGEplETpfz2GOPUalSJQYNGsQ777zDlClTeOedd+jWrRvTpk3LdW5gWeD+\ntGnTWLJkCXfeeWdYfZf2dDm///67P2pJlSpV2LVrV5wlMopLQc+uF1Lxo8As2iJSFvgYmKqq+TMP\nu3MeALJU9ZkIypkCXK2q48O5LhwjkzkiYpPuRlypVauW37H4zTff5IknnMFVMCUQSjF07NiRiRMn\nRk/IJCUzM9O/X69evThKYsSIvP7KAK8BywKVm4hU9EZ1eEaH3YClRepQpKqIDBGRf4tIN3EMAtYA\nV4XbXjgKrhMwX0RWiMhiEVmSN+WNYUSbK6+8ksWLFzNmzBiGDRvm98MKNFf//fff85UFkpKSwuHD\nh6MvbJKxefNm/37dunXjKIkRbYL5K4tIR6A30EVEForIAi/RaR1gtogsBL7DjfqmhW69QN7EuYUt\nAW4EZgB/xmWvuTTcxsJZg7so3MYNI9JUrVqV55/PPzMSOFqbNWsWl112Wa6ywP25c+fypz/9KbqC\nJiFbtuQso5iCS24KSHgaao3s1Ah1fYKqngIgImOATCDNlxsuXMLJB7dWRGoAzYDyAVVh5+gxjEgy\nffp01q5dy+TJk9m4cSPz58+nUqVKrF69mi+//BIR8dfv2LGDdevWMWTIkHiLXeLYtGmTf79OnTpx\nlMRIYg75dlT1iIhsKKpyg/CCLd+ISyPeEFgEnAV8C3QpaueGEQm6dOmSy7rPx5o1a/z7weqN8Hjz\nzTf9+82aNYujJEYS01ZEfNZLAlTwjouUni2cNbjbgDOAtap6HtAO5/RtGEaS895775kPnBF1VLWM\nqlb1tiqqWjZgP+zco+Gswe1X1f0igogco6oZImIxIg0jidm/fz/Nmzdn/fr1ucotiokRTUSkPHCi\nd/hz1NfggA0iUh34EOfItwNbfzOMpGbw4MH5lFvz5s059thj4ySRkcx4fnb/Avri9IsAjURkLPAP\nVT1U0PX52iuKE6mInAtUAz5T1YNhNxBDEslRtqSQyI7e8aS0OXrv3r2bKlWq5CobNmwYAwcONCOT\nJCGaz25REJFngSrAHaqa5ZVVBZ4C9qnqbeG0F84Izo+qzirKdYZhlBxee+21XMeLFi2ibdu2cZLG\nKCX0AJoHftWp6i4R+RuQgbMFKTThGJkYhlGKCDQq6devnyk3IxZosCkLL/l19LIJxAIRuUNElnqR\nUsaLSDmvvLuIZIjIShG5J95yJjuNGzfGMyayLWBr3LhxvP9rYsrll1/u3//Pf/7Dzp074yiNEUtE\n5FUR2RIYrUpEGorIdBH5yYtkdWtAXaTe0ctE5Log8lyDG8GFRZHW4KKBiNQHZgMtVPWgiLwHfAK8\nBawEugKbgB+AXqqa4V1XB2gJHFTVOUHatTU4I6ok6xpcdnY2J598MhkZ7r3y9NNPM3jw4LjIYkSH\nUM+uiHQCdgNvBGT0rgvUVdVF4mJPzgcuxb2fi/yOztNvA+ADYJ/XPsDpQAXgMlXdGM79JdQIDhcG\nppJnSVMR92N1AFap6lrPguZd3I/qI5xkeoZhFJKUlJRcCu3555+3GJ6lhDATnkbsHa2qG1X1TOAh\n4Fdve0hVO4Sr3KCIRibRQFU3icjTwDpgLzBNVb8QkSuAQDvlDbgf1Icvmd6+UMn0hg8f7t9PT08n\nPT09wtIbpYlI5NQqKVxzzTXcd999bN++nXXr1tGvXz/Gjh1LSkqifRsbhSFSz67kTnh6IcV8Rwe0\neyJQR1WnA9MDyjsCm1V1dViCqmpCbLgs4V8Cx+JGcpNwkauvAEYFnHcN8EIY7aphRBPvGYvW30UM\n7yQ4w4YNU9wCvwJ69913x1skI0IU9OwCjYHFQcorA/OAS73jYr2j87T9MXBKkPJTcFkKwmovkT7D\nzgfWqOrv6ixmPgD+AGwE0gLOa+iVGYYRAypXrpzreMOGDXGSxIg33vLRBOBNVZ3sFUfyHV1HVZfk\nLfTKmoTbWCIpuHXAWSJSXkQEt2C5DLdgeaKINPasKnsBU+Iop2GUGp588knuvvtu/3GTJk145JFH\n4iiREUMKlfCUyL6jqxdQF7aNRcIoOFWdi/syWAj8iPthR3ujuVuAacBPwLuqujxughpGKWHSpEm5\nlFvbtm2ZM2cOxx9/fBylMmKBhJHw1HtHDyIy7+h5InJTEHluJMeqsvD3oUluQm9uAka0SVY3gccf\nf5x7770XgJo1a7J69WqqVasWF1mM6BDNZ7coeC4Fk4CD5HYTKIdzE9gc6tqg7SX7y98UnBFtklXB\n/frrrzRr1szvGjBz5kzOPffcuMhiRIdEU3A+ROQ8oLV3uF1V3ylKOwkzRWkYRmLRpEkT+vTp4z9+\n+umn4yeMUapQ1RmqOkJVRwB3FbUdU3CGYYTkpJNyUj4eOHAgjpIYpZgijzATxtHbMBKNw4fhk09g\n/Xr4wx9gzx5YtgyuvRYqVYq3dNHn999/Z9SoUf7jXr16xVEaoxQzuqgX2hqckRQsXAh33eWUUbVq\ncPvt0KsXBAbcyMqCd96B7duhfn2oUgUqVoRjjoEDB6B8eWjUCH77DaZOhddec+3l5Ycf4PTTc46T\ncQ1u3bp1dO/eneXLnTFcxYoV2bx5c778cEbJJlHX4ABEpBaAqm4rahs2gjNKLJs3wyuvwKJF8Omn\ncCgg12/v3tCvHzRsCA0auLqlS2HXrvD6aNYMzjwTvv8eatSAVq2gQpJHPM3IyKBr165s2rTJX/bc\nc8+ZcjOijucD/QDONSzFKzoMjFDVh8JuL9lHNzaCSx527YLJk93oassW+PZb2BcQ2W7QIPjb31z5\nI4/AL7/kb6NTJzjrLKccd++GvXvd6K1CBTfCW78eataEU06B666D887LPQoMRrKN4NLT05k1y+U0\nTk1N5Y033rDpySQl0UZwIjIYuAjor6q/eGUnAC8Dn6nqs2G1l+wvf1NwJZO1a2H2bGjRwimf996D\n55+HvCnJ/vQn+MtfoF07CLCHAJwCW78eNm2C1FRIS4MmTSIvazIpuJUrV/oNS8qUKcO0adPo0qVL\nzPo3YksCKriFwAWquj1PeS1cAP524bRnU5RGQrB3L/z3v2509u23bjoxGB07urW1Fi2gcWM3hRiK\nypWhZUu3GYVj3Lhx/v1LLrnElFspRUReBXoAW9TLB3eU8l+BnUA2cEhVO1A0UvMqN3DrcCKSGm5j\npuCMuLFunZsWzMx004GZmTl1lSrBOefAihVuGrJrV7jpJjA/4+hx+PBhXn/9df9x37594yiNEWfG\nAiOANwpZng2kq+qOPOVhJTzFRTApSl1QTMEZMWfVKnjySXj1VcjOzik/5RQ3OktPh/btnVWjET32\n7dvHBx98QEZGBuvXr+fTTz9l2zZnsFa7dm0uvvjiOEtoxAtVnS0ijQtbjvNVC7Va3UpVZ4jIWSJS\nQQvOCddWRIKZggkQ9hvBFJwRM+bMgcGDnUUiQNmyOWb5d9wBw4a5MiM23HzzzYwdOzZo3XXXXUdq\natgzQkaCE8VkvQp8LiJHcLnhAn3XCp3wVFXLRFIoMzIxokp2Nnz2Gbz8Mnz8sSurWtUZhwwZ4tbS\nVEESZpk7fEqikUlmZiZpaWn+OJOBNG/enBkzZlC/fv2I92skFgU9u95I7aPAtbZQ5SJST1UzPWOQ\nz4FbVHV2NGUvDPa9bESFXbtg3DgYMQJ+/tmVlSvnnLGHDMkdCaQkK7eSyquvvupXbq1bt2bQoEE0\natSI448/nubNm5NyNN8IwwhAVTO9f7eJyCSgAxC2ghORLNxoMPCt4DtWVa0aTnum4IyIsWqVG43N\nmAH33JNj0p+WBgMHwg03QO3a8ZXRgCNHjuQKwTVkyBD++te/xlEiI0EJlvA0X7mIVARSVHW3iFQC\nugEPFqVDVQ0ZTUBE7gi3PZuiNIrN1q1uZPZGHruqzp3httugZ8/kXlsraVOUH330ET179gTguOOO\nY8OGDRxzzDER7cMoGYR6dr2Ep+lATWAL8ICqjg1WDszE5XBT3KBpvKo+FgVZ16lqWjjXJNRrR0Sq\nAWNweYCygb6q+r2IdAeew1npvKqqj8dRTMNj3jxnur9okTsuXx7q1XNO1Y88An/+s00/JiKTJk3y\n799www2m3Ix8qGrQIX2ocuDUKIrjI+y3SUIpOOB54FNVvVJEygIVRSQF+DfQFdgE/CAik1U1A8L2\nsTAigKqLLNKvn3PQLl/ehbR69tn80USMxCPQOtKUm1GCCHsqI2FWkkWkKnCOqo4FUNXDqroLt1i5\nSlXXquoh4F3g0oBLW6nqTCBbRJI8DG582b4dXnrJRdK/+mqn3Pr0gR07XLBjU24lg27duvn3p06d\nGkdJDCM3IpIlIruCbFlA2Ga9iTSCOx7YLiJjgbbAPOB2oAEQmLRkA07p+Si0j4VRNDZvhhdecLEg\n9+51ZXXrOr+1//s/m4YsaZx6as5s0po1a+IoiWHkpiAjk6KQSAquLNAeuFlV54nIc8C9wI8FXaSq\nW3CLnSEZPny4fz89PZ309PTiyloq+P13uPNOeOstl/wToFs3l4rmz392udRKI1F0lo0JCxcu9O+3\nbt06jpIYRnRJJAW3AVivqvO84wnAPcAnQKDlTENgYzgNByo4o3DMnOkU2aZNUKaMc8y++26X2bq0\nk/cj6cEHi2QRHTe+/vpr//4555wTR0kMI7okzBqcNxJbLyLNvaKuwDLcVOWJItJYRMoBvYApcRIz\n6VGFp592wY03bYKzz4bly2HSJFNuycLs2Tn+t6bgjGQmofzgRKQtzk0gFVgD3KCqOz03gefJcRMo\ntI+F+cEVniNHYOhQeMz7dYcMgYcfdiM4IzQlyQ8uKyuL6tWrk52djYiwY8cOqlWrFrH2jZJFouWD\nizSJNEWJqv4InBGk/DPAbPSixC+/wLXXwnffOSVXtiy8+aaL7G8kF1u2bCHbS+GQlpZmys1IahJm\nitKIPVlZLl7kmWfCN9845da4sZuONOWWnGQH5CeybAFGKETkVRHZIiKLC1neXUQyRGSliNwTW2lD\nYwquFLJtm5t+bNDAxYfctg0uuMBZTf76K/ToEW8JjWhx5MgR/34Zm3s2QjMWuLAw5QHBOC4ETgau\nFpEWAfV1RCRdRM6OorxBMQVXiti5E+69F5o0cetsWVnQsSOMHg2ffAI1asRbQiPamIIzCoOX6iZf\ndu4Q5QkbjCOh1uCMyHHgALz/Pixc6PKvZWbCxIkuGgnAJZfA/fe76Umj9BA4RSnmoV/qiJIPZ8IG\n4zAFl2R8/bVbV5s82WXKzkunTvDUU6bYSivVq1f3769evZo9e/ZQKTA5n5HUxMOHszDBOKKFKbgS\nzM6dMHUqHDzowml9+inMmpVTf+qpcMUVbjR33HHuuHNnC61VmklLS6NVq1YsW7aM/fv389///pfL\nL7883mIZJZuNFDMYR7QwBZeg7NwJX30FGRnO4Xr3bpeK5tAhlyE7Kwtmz4Y9e3JfV60a3HKLs4I8\n+WRTZkZ+LrvsMpYtWwa41Dmm4IwQFCrhKfADXjAOIBMXjOPq6It3dBLK0TsaJJqjtyrs3w+7djkl\n9e23TpEtWODqa9RwymzBAqfMjkbnzi5jdtWqbtqxZ08ImIUyYkBJcvQGmD9/Pqeffjrgpiy3bt1q\nLgOllEgkPPXKLyJ3zs6IJzwtCqbgwmTnTti4EZo1c9ODo0dD7dpO0fTq5ZJ9/vabM7k/fBjmz3fG\nHlu3urING5yCOxopKc7CsV07aNQIKlVyhiJlyri+q1eHpk0tRU0iUJIU3O7du+nTpw8TJ070ly1Y\nsIB27dpFrA+j5JDskUxMwRWCI0ecBeIjj8CSJa6sQgXYl8ce6LjjnPLavbvg9sqVcyOuihXdNGLX\nrm70dcwxLrdaxYpOcdWqVSyxjRhRkhTc6NGj6d+/v/+4Vq1aLF26lNq1a0esD6PkkOwKztbgQrBu\nHbzzDqxY4UZqmze78vLl3Yht3To3qho2zCmm0aPhp5/cOTVqQM2abjRXp45LDtqmjRt1NWgAVSKa\n8cgwCsehQ4f48MMP/cfnnXceY8aMMeVmJC2lUsHt3+/WvwJHSPv3O4W0bx+MHQv33Zd7JNa0Kdx1\nl4v8Ua6cmy6sXDlHWQ0aBMuWuUSgxx0X2/sxjGBkZ2ezePFivvzyS7744gu++uor9voy1gK9evXi\nhBNOiKOEhhFdSsUU5fr1ysSJsHixc3T+8ktnfdiihbNM3LrVpYQpW9atffnWyC691CX47NABTjvN\nLBKN4CTqFOW+ffuoUaMGBw4cCFo/d+5czjgjX2xzoxSR7FOUpULBiSh5bzPvGlpKCviCPHTqBHfc\nAWY9bRSGRFVwAF27dmX69On+47S0NNq1a8eVV15J7969IyGiUYJJdgVXKqYoU1NdAOHzz3frYH/4\ng1sLmz/fjeSqVoXWrZ2C27fPphiN5OGKK66gdu3anH/++XTt2pUmTZrEWyTDiBmlYgSXmanUrRtv\nSYxkJZFHcIZREMk+gku4bAIikiIiC0RkSkBZsXINxUu5RSGoaUL3G8++43nPJZFo/F7R+j+wdqPb\nbjDCzfsmIr+KyI8islBE5sZM0KOQcAoOuA1Y5jtI5FxDR6M0vuxL4z2XREzBWbtHIdy8b9lAuqq2\nU9UOea6zfHAAItIQuBgYE1CcsLmGDMMwkpEi5H0TQuuTuL2jE0rBAc8CdwGBiw7Bcg01CDj25RpK\niXWuIcMwjFJEQe9iBT4XkR9E5KY818XvHa2qCbEBlwD/9vbTgSne/hXAqIDzrgFeCKNdtc22aG9R\n/LuI+73ZltxbAc9eY2BxwHHIdzFQz/u3FrAI6BRvnaKqCeUm0BHoKSIXAxWAKiLyBvASxcg1pEls\nIWQkP/b8GglEyLxvqprp/btNRCbhpjNnx1zCPCTMFKWq3qeqaap6Ai6f0HRVvY6AXEMiUs6rm1JQ\nW4ZhGEaxCZn3LfBdLCIVRaQygIhUAroBS2MubRASRsGFQlWPALcA04CfgHdVdXl8pTIMw0hevLxv\nc4DmIrJORG7w3sWDyP8urgPMFpGFwHfAR6o6LV6yB5LUjt4i0p3cSfgej1I/DYE3cP/R2cBoVX1B\nRGoA7+Hmsn8FrlLVnVHoPwWYB2xQ1Z4x7LcazuK1Ne6++wIrY9T3HUA/r98lwA1ApUj3LSKvAj2A\nLaraxisL+fuKyBDc73AYuC1R/tANozSS8CO4onI0/7kIcxgYrKonA38Abvb6uhf4QlVPAqYDQ6LU\nfy7fwRj2+zzwqaq2BNoCGbHoW0Tq474k23tKpyxwdZT6Hksef6BQ/YhIK+AqoCVwEfCSiIXoNoy4\nEW8rl2htwFnA1IDje4F7YtT3h8D5uBd+Ha+sLpARhb4aAp+T2/I0Fv1WBVYHKY9F3/WBtUANnHKb\nEs3fm/zWZEH7yfuMAVOBMwvRfgqwwPf/55V19/pZWdTnNkS7vwI/AguBuUVoM+j1xZG3gDaLK2s1\n4H1gOW5K7cxI/LYFtFtkeYHm3nULvH93ArdG4LctqN1i/b4lYYu7AFG7sWK6FxSj3ybeg1MZ2JGn\n7vco9Pc+cCpwLjkKLhb9tgW+x41wFgCjgIqx6Ntr91YgC9gCvBnN+ya/gvs9WD/ACOCvAeVjgMsL\n0f4dwFsB/38pwM9ev6k4s+sWAefXwX3QnB1Ou17ZGqBGiPOP2m6w64srbyiZIiDrOOAGb78s7qOs\n2L9tsHYjIW+e33MT7uM1Is9C3nYjKW8ib0k7RRkPPEuiCbi1l904P5NAIrrgKSKX4NaGFpHb2ikv\n0VhoLQu0B15U1fbAHtwIJqr3DCAi1XERFBrjRnOVRKR3LPoOQZH7iVb0nhDtQvEjTgS7vrjyhpKp\nyLKKSFXgHFUdC6Cqh1V1V3FlLaDdYsmbh/NxsyMbiitvAe1GUt6EJZkVXEifjWggImVxyu1NVZ3s\nFW8RkTpefV1ga4S79fkOrgHeAbqIyJvA5ij3Cy6KwXpVnecdT8QpvGjfM7g/1DWq+rs6y65JwNkx\n6psC+tkINAo4rzDPXLSi9wRrF++4OBEngl1fXHlDyVQcWY8HtovIWC94+ygRqRgBWYO161MAxf1t\nffwF9/dMBOQN1W4k5U1YklnBxdp/7jVgmao+H1A2Bejj7V8PTM57UXHQ4L6D1wIfRbNfr+8twHoR\nae4VdcWtR0T1nj3WAWeJSHnPiKMrzsgmWn3n9QcK1c8UoJeIlBOR44ETgZCR1cMYgedCVbeo6peq\nOqeQ7Qa23dEbcV+MM4bqVNh2j3Z9UeUtoM3iyJp3hmEvboahuLIW1G5xf1tEJBXoiVt6OCrFbLfY\n8iY6SavgNIb+cyLSEeiNG0Et9L7sugOPAxeIyArcS/ixaPQfhMdi1O+twHgRWYRbk/sXMbhnVZ2L\nGy0vxC2SC24NMOJ9B/MHIsTvq6rLgP/glO2nwEBVLWj6MtgI/A2KP/uQt93zvHbRgIgTuJFvh5Ct\nBCHE9cWSN5RMxZQ17wzDBKCdV16c3zZYu+0jIK+Pi4D5XhsQuZmovO1GSt7EJpYLfrbZZlvwjdxG\nQmXIMSwohzMsaBmBdisClb39SsA3QLcw2gp2/QXFkTeUTMWV1btuFtDc238A9wFU7N82RLsViiuv\nd+07wPUBxxF5FoK0W+zftyRsiRSL0jAM3OyDiPhmH3xBCiIx+1AHmCQiiptqG6/hOaIHu/5zgGLI\nG1Qmb4q3OLJCzgxDKs5i8IYI/bb52sW5ixRLXm+N8Hygv68sEvIGa5fiPwslgqSOZGIYhmGUXpJ2\nDc4wDMMo3ZiCMwzDMJISU3CGYRhGUmIKzjAMw0hKTMEZhmEYSYkpOMMwDCMpMQVXghCRaiLyt4Dj\n2XGQobyIzCxKnjMRSRWRWV6uPsMwjKhiL5qSRQ1goO9AVY8aB7AoiEgLLzN1MPoCE7UIDpTqoqF/\ngYubaRiGEVVMwZUsHgWaerEunxCRLAAvoPRyL8L5ChF5S0S6ishs7/h0XwMi0ltEvvfaeDnESOw8\nXJzHYPTGCy4c0O9bIrJMRP7jjfAqisjHXlzOxSJyZcD1k702jDgjItki8mTA8Z0iMiyg7o2AujIi\nsk1EpgSUHfGeoyUi8p6IlA/SRx0ReUdEVnlR6z8WkRO9ugYi8qGIrPTqnxWXlSNY+5PFparJW+eL\n/Xp35H+hXPeRFeb5IiIDRORGEWkaLbmMgjEFV7K4F/hZVdur6t3kToXSFHhSVU8CWgBXeyO8u4B/\ngBuZ4VJmnK0uing2eZSNFyT6RqCReClhAupSgeNVdV1A8UnAv1W1FbALuBmXgXijqrZT1TbAZwHn\nLwXOKM6PYESMA8DlInJskLo9QGsROcY7voDcaVsA9njP4inAIeD/grQzCZflopmqngEMwYWJAvgA\n+EBVm+MyT1fBBewO1v4O3LOVt66d9+8Thb7rohHujMWtuITAM4E/R1wao1CYgkseflEXzR5c9oQv\nvf0luECt4CLftwd+EJGFQBfghMBGVPUznHIarS4lTiDHAf/LU7ZOVb/z9sfjItkvBrqJyKMi0klV\n/V+/qpoNHBCRSkW9USNiHMZlYRgcov5T4BJv/2py5xLLy9e49EB+ROQ84KCqjvaVqeoSVf1GRLoA\n+1TVl+VAcRnI+wYbCQLfkjsPWqHWgEVkkjdyXCIiN3pljb0Zh1EislREPvMpchG5X0QyROQrEXlb\nRHy/jQS0WeAsiDcK/aO6dEVpuEziRhwwBZc8HAjYzw44zgZ/UG0BXg/48m2pqg8FNuKN2jaH6GMf\nEOzlE4iq6s+41CRLgEdEZGiec44B9h+lHSP6KPAi0FtEqgSpfxe42nv5t8GNSAIR8L/QL8L9fwfS\nGpgfou+T89Z5H0JryVGUvvbL4D7OAvM5VsgzRRk4DR7IDd7I8QzgNhGp4ZWfCIxQ1dbATuAKbyr/\nMuAUXI600/M2VphZENyH4y4RuR63Zr4BIy6YgitZZOGmcXxIiP28+Oq+BP4sIrUARKSGiKTlObcD\nMFdETpc8qepV9X9AGXEJZH2kiciZ3v5fgdkiUg/3df428CReviyvz2OB7ery9RlxRlV3A68Dt+Wv\n0qVAE9zo7RPyP2MVRGQBLqnrWuDVCIgU2Iev/UygNvB5QN3ePFOUoRKE3i4uX+F3uFxqzbzyX1TV\np5Dn4+6zIzBZVQ95v8tHAe34piiPOguCyy7/mqq+jvsg/FZEeojIVBEpKyIdReQf3t/YN/4bF+ku\nIm96a9ini8gn3rn3+v5mjfAwBVeCUNXfgTme4cYT5F4XCLXvP/bSbAwFponIj7gUHHXznLsJNxVU\nRYOnqp8GBFpvrsBlA14GVAdexn0Bz/VeAMOARwLOPw/3sjQSh+eBfri8YHmZgvtICTY96VMy7VX1\nNlU9LCIDfaMqYDtBRkEey/LWeUYkjXD5z/zt46b5BJfAuNCIyLk4BXSmqp6Ky6Xmm4EInPE4gpvl\nUI4+9XnUWRCgHrDG+xCs601VLgFeAe5R1W+AGeqSpn4rLmEyngzLVXWvV7fROzcLOBMjfOKViM62\nkrnhph5f9/YbA0vCvH4icGK878M2BcgK2H8cNwobFliH+9i5xdv3J0/Ne30BfXwL3BhwfArQ0duf\nC87osAIAAAGbSURBVFzj7ZfBrQc+EUK+U4FfgZQw+u6JG5GBM7zaB3TO+9wCd+I+xE7HjeaOASrj\nPt4G5/k9WnrltbzjGkBann7/iTP66gX09srSvN/yEa+Ns73yLriM8PWAtsB9Ae1M9O67e7yflZK6\n2QjOCAtVXQjMCFhYL7R1mWeFOUndGp0RfwL/754GagaU+Ub9G1X134W4PhSXAReIyM8isgRnJbk5\noO4qEVkJZOAU0D+Cta9uFPQjbroUoHyeNbhA60sfnwGpIvKT1++3BcmubtQ02evnE5yx1M7A87Vw\nsyDvAJcDNVR1vFfm+3v5J86S1McB3AiyHs6AK3AE+buqLlJn+GUUAUt4ahiG4SEilVR1j7f+/BVw\nk6dci9vudcAhVX1HRNJxFqwpuHW/6bjp/SrApcAAoBXO77WXqv5W3P5LK6bgDMMwPERkPE65HAOM\n0+j71xlRxBScYRiGkZTYGpxhGIaRlJiCMwzDMJISU3CGYRhGUmIKzjAMw0hKTMEZhmEYSYkpOMMw\nDCMpMQVnGIZhJCWm4AzDMIyk5P8BBMzYJuBuUeQAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x7f54a34a9f90>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import numpy as np\n",
    "from numpy.linalg import norm\n",
    "\n",
    "def theta_NMP(u):\n",
    "    \"\"\"Calculate the NMP-CORE angle for E. coli AdK in degrees\"\"\"\n",
    "    C = u.select_atoms(\"resid 115:125 and (backbone or name CB)\").center_of_geometry()\n",
    "    B = u.select_atoms(\"resid 90:100 and (backbone or name CB)\").center_of_geometry()\n",
    "    A = u.select_atoms(\"resid 35:55 and (backbone or name CB)\").center_of_geometry()\n",
    "    BA = A - B\n",
    "    BC = C - B\n",
    "    theta = np.arccos(np.dot(BA, BC)/(norm(BA)*norm(BC)))\n",
    "    return np.rad2deg(theta)\n",
    "\n",
    "def theta_LID(u):\n",
    "    \"\"\"Calculate the LID-CORE angle for E. coli AdK in degrees\"\"\"\n",
    "    C = u.select_atoms(\"resid 179:185 and (backbone or name CB)\").center_of_geometry()\n",
    "    B = u.select_atoms(\"resid 115:125 and (backbone or name CB)\").center_of_geometry()\n",
    "    A = u.select_atoms(\"resid 125:153 and (backbone or name CB)\").center_of_geometry()\n",
    "    BA = A - B\n",
    "    BC = C - B\n",
    "    theta = np.arccos(np.dot(BA, BC)/(norm(BA)*norm(BC)))\n",
    "    return np.rad2deg(theta)\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    import MDAnalysis\n",
    "    from MDAnalysis.tests.datafiles import PSF, DCD\n",
    "    import matplotlib\n",
    "    import matplotlib.pyplot as plt\n",
    "\n",
    "    u = MDAnalysis.Universe(PSF, DCD)\n",
    "    data = np.array([(u.trajectory.time, theta_NMP(u), theta_LID(u)) for ts in u.trajectory])\n",
    "    time, NMP, LID = data.T\n",
    "\n",
    "\n",
    "    # plotting\n",
    "    degreeFormatter = matplotlib.ticker.FormatStrFormatter(r\"%g$^\\circ$\")\n",
    "    fig = plt.figure(figsize=(6,3))\n",
    "\n",
    "    ax1 = fig.add_subplot(121)\n",
    "    ax1.plot(time, NMP, 'b-', lw=2, label=r\"$\\theta_{\\mathrm{NMP}}$\")\n",
    "    ax1.plot(time, LID, 'r-', lw=2, label=r\"$\\theta_{\\mathrm{LID}}$\")\n",
    "    ax1.set_xlabel(r\"time $t$ (ps)\")\n",
    "    ax1.set_ylabel(r\"angle $\\theta$\")\n",
    "    ax1.yaxis.set_major_formatter(degreeFormatter)\n",
    "    ax1.legend(loc=\"best\")\n",
    "\n",
    "    ax2 = fig.add_subplot(122)\n",
    "    ax2.plot(NMP, LID, 'k-', lw=3)\n",
    "    ax2.set_xlabel(r\"NMP-CORE angle $\\theta_{\\mathrm{NMP}}$\")\n",
    "    ax2.set_ylabel(r\"LID-CORE angle $\\theta_{\\mathrm{LID}}$\")\n",
    "    ax2.xaxis.set_major_formatter(degreeFormatter)\n",
    "    ax2.yaxis.set_major_formatter(degreeFormatter)\n",
    "    ax2.yaxis.tick_right()\n",
    "    ax2.yaxis.set_label_position(\"right\")\n",
    "\n",
    "    fig.subplots_adjust(left=0.12, right=0.88, bottom=0.2, wspace=0.15)\n",
    "\n",
    "    for ext in ('svg', 'pdf', 'png'):\n",
    "        fig.savefig(\"NMP_LID_angle_projection.{0}\".format(ext))\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Note that one would normally write the code more efficiently and generate the atom groups once and then pass them to a simple function to calculate the angle"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "def theta(A, B, C):\n",
    "    \"\"\"Calculate the angle between BA and BC for AtomGroups A, B, C\"\"\"\n",
    "    B_center = B.centroid()\n",
    "    BA = A.centroid() - B_center\n",
    "    BC = C.centroid() - B_center\n",
    "    theta = np.arccos(np.dot(BA, BC)/(norm(BA)*norm(BC)))\n",
    "    return np.rad2deg(theta)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Bells and whistles"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Quick data aquisition"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Especially useful for interactive analysis in ipython pylab using list comprehensions (implicit for loops):"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "```python\n",
    "protein = u.selectAtoms(\"protein\")\n",
    "data = np.array([(u.trajectory.time, protein.radius_of_gyration()) for ts in u.trajectory])\n",
    "time, RG = data.T\n",
    "plot(time, RG)\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### More on the trajectory iterator"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "One can directly jump to a frame by using indexing syntax:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "< Timestep 50 with unit cell dimensions [  0.   0.   0.  90.  90.  90.] >"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "u.trajectory[50]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "50"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ts.frame"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "You can also slice trajectories, e.g. if you want to start at the 10th frame and go to 10th before the end, and only use every 5th frame:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "9\n",
      "14\n",
      "19\n",
      "24\n",
      "29\n",
      "34\n",
      "39\n",
      "44\n",
      "49\n",
      "54\n",
      "59\n",
      "64\n",
      "69\n",
      "74\n",
      "79\n",
      "84\n"
     ]
    }
   ],
   "source": [
    "for ts in u.trajectory[9:-10:5]:\n",
    "    print(ts.frame)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Note**:  Trajectory indexing and slicing uses 0-based indices (as in standard Python) and MDAnalysis also numbers frames starting with 0. Thus the \"tenth frame\" in a trjectory has `ts.frame == 9`."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Writing Coordinates"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "MDAnalysis also supports writing of data in a range of file formats (see the [Table of Supported Coordinate Formats](https://pythonhosted.org/MDAnalysis/documentation_pages/coordinates/init.html#id1) for details). MDAnalysis supports both single frame writers (such as a simple PDB or GRO file) and trajectory writers (e.g. XTC, DCD, but also multi-frame PDB files)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Single frames"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The most straightforward way to write to a file that can only hold a single frame is to use the write() method of any AtomGroup as already also shown under Processing AtomGroups. For instance, to only write out the protein without solvent to a file in GRO format:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "from MDAnalysis.tests.datafiles import PDB\n",
    "uPDB = MDAnalysis.Universe(PDB)\n",
    "protein = uPDB.select_atoms(\"protein\")\n",
    "protein.write(\"protein.gro\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "MDAnalysis uses the file suffix to determine the output file format (unless the format keyword is specified) and will raise an exception if it is not suitable for single frame writing."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "del uPDB  # clean up this example\n",
    "os.unlink(\"protein.gro\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Trajectories"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The typical use pattern is to\n",
    "\n",
    "1. Get a trajectory writer with `MDAnalysis.Writer()` (which is the same as `MDAnalysis.coordinates.core.writer()`), typically specifying in advance how many atoms a frame will contain.\n",
    "2. Use the `write()` method to write a new time step to the trajectory.\n",
    "3. Close the trajectory with `close()` (although it is recommended to simply use the writer with the `with` statement and have the context manager close the file automatically)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Example: Protein-only trajectory"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In practice, the second step is typically repeated in a loop as in the example below:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import MDAnalysis\n",
    "from MDAnalysis.tests.datafiles import PDB, XTC\n",
    "\n",
    "uXTC = MDAnalysis.Universe(PDB, XTC)\n",
    "protein = uXTC.select_atoms(\"protein\")\n",
    "with MDAnalysis.Writer(\"protein.xtc\", protein.n_atoms) as W:\n",
    "    for ts in u.trajectory:\n",
    "        W.write(protein)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The loop steps through the input trajectory frame by frame. The coordinates of the selection (the AtomGroup protein) change accordingly and are then written as a new frame into the output trajectory.\n",
    "\n",
    "The output trajectory only contains the coordinates of the protein. For this trajectory to be useful, a protein-only topology file also has to be stored, as in the example under Single frames."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {
    "collapsed": false
   },
   "outputs": [],
   "source": [
    "del uXTC  # clean up this example\n",
    "os.unlink(\"protein.xtc\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Example: Saving dynamic per-atom properties in B-factor"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "It is often very useful to project per-atom properties on the structure. A common approach is to save scalar values in the B-factor field of a PDB file and then color atoms by B-factor (also known as temperature factor or just beta).\n",
    "\n",
    "The following example computes the shift of each atom in AdK relative to a reference structure (line 23). We take as reference the closed conformation (after a structural superposition on the CORE domain with alignto()). The shifts are written into the B-factor with the set_bfactor() method of AtomGroup. Each frame is written out as part of a multi-frame PDB file:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/anaconda2-4.1.1/lib/python2.7/site-packages/numpy/lib/utils.py:99: DeprecationWarning: `set_bfactors` is deprecated!\n",
      "This will be removed in version 0.16.0; use `bfactors` property instead\n",
      "  warnings.warn(depdoc, DeprecationWarning)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Frame 0: CORE RMSD before/after superposition: 0.0 / 0.0 A. min-max displacement: 0.0...0.0 A\n",
      "Frame 1: CORE RMSD before/after superposition: 0.7 / 0.7 A. min-max displacement: 0.0...3.5 A\n",
      "Frame 2: CORE RMSD before/after superposition: 0.8 / 0.8 A. min-max displacement: 0.0...3.6 A\n",
      "Frame 3: CORE RMSD before/after superposition: 1.0 / 1.0 A. min-max displacement: 0.0...4.2 A\n",
      "Frame 4: CORE RMSD before/after superposition: 1.1 / 1.0 A. min-max displacement: 0.0...4.7 A\n",
      "Frame 5: CORE RMSD before/after superposition: 1.1 / 1.1 A. min-max displacement: 0.0...6.3 A\n",
      "Frame 6: CORE RMSD before/after superposition: 1.2 / 1.1 A. min-max displacement: 0.1...6.4 A\n",
      "Frame 7: CORE RMSD before/after superposition: 1.3 / 1.2 A. min-max displacement: 0.1...7.2 A\n",
      "Frame 8: CORE RMSD before/after superposition: 1.3 / 1.2 A. min-max displacement: 0.1...7.3 A\n",
      "Frame 9: CORE RMSD before/after superposition: 1.4 / 1.3 A. min-max displacement: 0.0...6.6 A\n",
      "Frame 10: CORE RMSD before/after superposition: 1.5 / 1.3 A. min-max displacement: 0.1...6.7 A\n",
      "Frame 11: CORE RMSD before/after superposition: 1.5 / 1.4 A. min-max displacement: 0.0...7.6 A\n",
      "Frame 12: CORE RMSD before/after superposition: 1.6 / 1.4 A. min-max displacement: 0.0...8.0 A\n",
      "Frame 13: CORE RMSD before/after superposition: 1.6 / 1.4 A. min-max displacement: 0.0...8.6 A\n",
      "Frame 14: CORE RMSD before/after superposition: 1.7 / 1.4 A. min-max displacement: 0.0...8.7 A\n",
      "Frame 15: CORE RMSD before/after superposition: 1.7 / 1.5 A. min-max displacement: 0.1...8.6 A\n",
      "Frame 16: CORE RMSD before/after superposition: 1.7 / 1.5 A. min-max displacement: 0.0...8.7 A\n",
      "Frame 17: CORE RMSD before/after superposition: 1.8 / 1.5 A. min-max displacement: 0.0...8.5 A\n",
      "Frame 18: CORE RMSD before/after superposition: 1.8 / 1.5 A. min-max displacement: 0.0...9.3 A\n",
      "Frame 19: CORE RMSD before/after superposition: 1.8 / 1.5 A. min-max displacement: 0.0...9.8 A\n",
      "Frame 20: CORE RMSD before/after superposition: 1.9 / 1.5 A. min-max displacement: 0.1...10.0 A\n",
      "Frame 21: CORE RMSD before/after superposition: 1.9 / 1.6 A. min-max displacement: 0.1...10.3 A\n",
      "Frame 22: CORE RMSD before/after superposition: 2.0 / 1.6 A. min-max displacement: 0.1...10.2 A\n",
      "Frame 23: CORE RMSD before/after superposition: 2.0 / 1.7 A. min-max displacement: 0.1...10.3 A\n",
      "Frame 24: CORE RMSD before/after superposition: 2.1 / 1.7 A. min-max displacement: 0.1...11.4 A\n",
      "Frame 25: CORE RMSD before/after superposition: 2.2 / 1.7 A. min-max displacement: 0.0...11.5 A\n",
      "Frame 26: CORE RMSD before/after superposition: 2.2 / 1.7 A. min-max displacement: 0.1...11.5 A\n",
      "Frame 27: CORE RMSD before/after superposition: 2.2 / 1.8 A. min-max displacement: 0.1...12.3 A\n",
      "Frame 28: CORE RMSD before/after superposition: 2.3 / 1.8 A. min-max displacement: 0.1...12.5 A\n",
      "Frame 29: CORE RMSD before/after superposition: 2.4 / 1.8 A. min-max displacement: 0.1...12.7 A\n",
      "Frame 30: CORE RMSD before/after superposition: 2.4 / 1.9 A. min-max displacement: 0.1...13.2 A\n",
      "Frame 31: CORE RMSD before/after superposition: 2.5 / 1.9 A. min-max displacement: 0.1...13.1 A\n",
      "Frame 32: CORE RMSD before/after superposition: 2.6 / 1.9 A. min-max displacement: 0.1...13.7 A\n",
      "Frame 33: CORE RMSD before/after superposition: 2.5 / 1.9 A. min-max displacement: 0.1...14.4 A\n",
      "Frame 34: CORE RMSD before/after superposition: 2.6 / 1.9 A. min-max displacement: 0.1...14.8 A\n",
      "Frame 35: CORE RMSD before/after superposition: 2.6 / 1.9 A. min-max displacement: 0.2...15.0 A\n",
      "Frame 36: CORE RMSD before/after superposition: 2.7 / 1.9 A. min-max displacement: 0.2...15.6 A\n",
      "Frame 37: CORE RMSD before/after superposition: 2.7 / 1.9 A. min-max displacement: 0.1...15.4 A\n",
      "Frame 38: CORE RMSD before/after superposition: 2.7 / 2.0 A. min-max displacement: 0.1...15.8 A\n",
      "Frame 39: CORE RMSD before/after superposition: 2.8 / 2.0 A. min-max displacement: 0.1...16.1 A\n",
      "Frame 40: CORE RMSD before/after superposition: 2.8 / 2.0 A. min-max displacement: 0.1...16.4 A\n",
      "Frame 41: CORE RMSD before/after superposition: 2.9 / 2.0 A. min-max displacement: 0.1...16.6 A\n",
      "Frame 42: CORE RMSD before/after superposition: 2.9 / 2.0 A. min-max displacement: 0.1...16.9 A\n",
      "Frame 43: CORE RMSD before/after superposition: 3.0 / 2.0 A. min-max displacement: 0.1...17.2 A\n",
      "Frame 44: CORE RMSD before/after superposition: 2.9 / 2.0 A. min-max displacement: 0.1...17.7 A\n",
      "Frame 45: CORE RMSD before/after superposition: 3.0 / 2.0 A. min-max displacement: 0.0...17.9 A\n",
      "Frame 46: CORE RMSD before/after superposition: 3.0 / 2.0 A. min-max displacement: 0.1...18.5 A\n",
      "Frame 47: CORE RMSD before/after superposition: 3.1 / 2.0 A. min-max displacement: 0.2...18.4 A\n",
      "Frame 48: CORE RMSD before/after superposition: 3.1 / 2.1 A. min-max displacement: 0.1...18.7 A\n",
      "Frame 49: CORE RMSD before/after superposition: 3.1 / 2.1 A. min-max displacement: 0.1...18.5 A\n",
      "Frame 50: CORE RMSD before/after superposition: 3.1 / 2.1 A. min-max displacement: 0.1...19.4 A\n",
      "Frame 51: CORE RMSD before/after superposition: 3.1 / 2.1 A. min-max displacement: 0.1...20.0 A\n",
      "Frame 52: CORE RMSD before/after superposition: 3.1 / 2.1 A. min-max displacement: 0.1...19.6 A\n",
      "Frame 53: CORE RMSD before/after superposition: 3.2 / 2.1 A. min-max displacement: 0.1...19.3 A\n",
      "Frame 54: CORE RMSD before/after superposition: 3.2 / 2.1 A. min-max displacement: 0.1...19.9 A\n",
      "Frame 55: CORE RMSD before/after superposition: 3.3 / 2.2 A. min-max displacement: 0.1...20.1 A\n",
      "Frame 56: CORE RMSD before/after superposition: 3.3 / 2.2 A. min-max displacement: 0.1...20.4 A\n",
      "Frame 57: CORE RMSD before/after superposition: 3.3 / 2.2 A. min-max displacement: 0.1...20.3 A\n",
      "Frame 58: CORE RMSD before/after superposition: 3.4 / 2.3 A. min-max displacement: 0.1...20.9 A\n",
      "Frame 59: CORE RMSD before/after superposition: 3.4 / 2.3 A. min-max displacement: 0.1...21.0 A\n",
      "Frame 60: CORE RMSD before/after superposition: 3.4 / 2.3 A. min-max displacement: 0.1...20.6 A\n",
      "Frame 61: CORE RMSD before/after superposition: 3.4 / 2.3 A. min-max displacement: 0.1...20.8 A\n",
      "Frame 62: CORE RMSD before/after superposition: 3.4 / 2.3 A. min-max displacement: 0.1...21.3 A\n",
      "Frame 63: CORE RMSD before/after superposition: 3.5 / 2.3 A. min-max displacement: 0.1...21.4 A\n",
      "Frame 64: CORE RMSD before/after superposition: 3.5 / 2.3 A. min-max displacement: 0.1...21.9 A\n",
      "Frame 65: CORE RMSD before/after superposition: 3.6 / 2.3 A. min-max displacement: 0.2...22.1 A\n",
      "Frame 66: CORE RMSD before/after superposition: 3.6 / 2.3 A. min-max displacement: 0.1...22.4 A\n",
      "Frame 67: CORE RMSD before/after superposition: 3.6 / 2.3 A. min-max displacement: 0.1...22.4 A\n",
      "Frame 68: CORE RMSD before/after superposition: 3.6 / 2.4 A. min-max displacement: 0.1...22.7 A\n",
      "Frame 69: CORE RMSD before/after superposition: 3.7 / 2.4 A. min-max displacement: 0.1...22.9 A\n",
      "Frame 70: CORE RMSD before/after superposition: 3.7 / 2.4 A. min-max displacement: 0.1...23.4 A\n",
      "Frame 71: CORE RMSD before/after superposition: 3.8 / 2.4 A. min-max displacement: 0.1...23.2 A\n",
      "Frame 72: CORE RMSD before/after superposition: 3.7 / 2.4 A. min-max displacement: 0.2...23.4 A\n",
      "Frame 73: CORE RMSD before/after superposition: 3.7 / 2.4 A. min-max displacement: 0.1...23.2 A\n",
      "Frame 74: CORE RMSD before/after superposition: 3.7 / 2.4 A. min-max displacement: 0.1...23.5 A\n",
      "Frame 75: CORE RMSD before/after superposition: 3.8 / 2.4 A. min-max displacement: 0.2...23.5 A\n",
      "Frame 76: CORE RMSD before/after superposition: 3.8 / 2.4 A. min-max displacement: 0.2...23.7 A\n",
      "Frame 77: CORE RMSD before/after superposition: 3.8 / 2.4 A. min-max displacement: 0.1...23.4 A\n",
      "Frame 78: CORE RMSD before/after superposition: 3.8 / 2.4 A. min-max displacement: 0.1...23.5 A\n",
      "Frame 79: CORE RMSD before/after superposition: 3.7 / 2.4 A. min-max displacement: 0.1...23.6 A\n",
      "Frame 80: CORE RMSD before/after superposition: 3.7 / 2.4 A. min-max displacement: 0.1...23.4 A\n",
      "Frame 81: CORE RMSD before/after superposition: 3.7 / 2.4 A. min-max displacement: 0.1...23.3 A\n",
      "Frame 82: CORE RMSD before/after superposition: 3.7 / 2.4 A. min-max displacement: 0.1...23.3 A\n",
      "Frame 83: CORE RMSD before/after superposition: 3.7 / 2.4 A. min-max displacement: 0.1...23.7 A\n",
      "Frame 84: CORE RMSD before/after superposition: 3.8 / 2.4 A. min-max displacement: 0.0...23.7 A\n",
      "Frame 85: CORE RMSD before/after superposition: 3.7 / 2.4 A. min-max displacement: 0.1...23.5 A\n",
      "Frame 86: CORE RMSD before/after superposition: 3.7 / 2.4 A. min-max displacement: 0.0...23.7 A\n",
      "Frame 87: CORE RMSD before/after superposition: 3.8 / 2.4 A. min-max displacement: 0.2...24.0 A\n",
      "Frame 88: CORE RMSD before/after superposition: 3.8 / 2.4 A. min-max displacement: 0.1...24.2 A\n",
      "Frame 89: CORE RMSD before/after superposition: 3.8 / 2.4 A. min-max displacement: 0.1...23.8 A\n",
      "Frame 90: CORE RMSD before/after superposition: 3.8 / 2.5 A. min-max displacement: 0.1...24.2 A\n",
      "Frame 91: CORE RMSD before/after superposition: 3.8 / 2.4 A. min-max displacement: 0.1...24.2 A\n",
      "Frame 92: CORE RMSD before/after superposition: 3.8 / 2.4 A. min-max displacement: 0.1...24.0 A\n",
      "Frame 93: CORE RMSD before/after superposition: 3.8 / 2.4 A. min-max displacement: 0.1...24.1 A\n",
      "Frame 94: CORE RMSD before/after superposition: 3.8 / 2.4 A. min-max displacement: 0.1...24.2 A\n",
      "Frame 95: CORE RMSD before/after superposition: 3.8 / 2.4 A. min-max displacement: 0.1...24.2 A\n",
      "Frame 96: CORE RMSD before/after superposition: 3.9 / 2.5 A. min-max displacement: 0.1...24.5 A\n",
      "Frame 97: CORE RMSD before/after superposition: 3.9 / 2.4 A. min-max displacement: 0.1...24.5 A\n",
      "Wrote PDB trajectory adk_distance_bfac.pdb with distances in bfactor field\n"
     ]
    }
   ],
   "source": [
    "# project a dynamic property on the structure using the B-factor field\n",
    "\n",
    "import numpy as np\n",
    "import MDAnalysis\n",
    "import MDAnalysis.analysis.align\n",
    "\n",
    "from MDAnalysis.tests.datafiles import PSF, DCD\n",
    "\n",
    "\n",
    "u = MDAnalysis.Universe(PSF, DCD)\n",
    "ref = MDAnalysis.Universe(PSF, DCD)  # copy of u\n",
    "\n",
    "CORE_selection = \"resid 1:29 or resid 60:121 or resid 160:214\"\n",
    "pdbtrj = \"adk_distance_bfac.pdb\"\n",
    "\n",
    "with MDAnalysis.Writer(pdbtrj, multiframe=True, bonds=False, n_atoms=u.atoms.n_atoms) as PDB:\n",
    "    # reference coordinates: set to first frame\n",
    "    ref.trajectory[0]\n",
    "    # iterate through our trajectory\n",
    "    for ts in u.trajectory:\n",
    "        # superimpose on the reference CORE (at t=0)\n",
    "        rmsd = MDAnalysis.analysis.align.alignto(u.atoms, ref.atoms, select=CORE_selection)\n",
    "        distances = np.sqrt(np.sum((u.atoms.positions - ref.atoms.positions)**2, axis=1))\n",
    "        # project displacement on structure via bfactor field\n",
    "        u.atoms.set_bfactors(distances)\n",
    "        PDB.write(u.atoms)\n",
    "        print(\"Frame {0}: CORE RMSD before/after superposition: {1[0]:.1f} / {1[1]:.1f} A. \"\n",
    "              \"min-max displacement: {2:.1f}...{3:.1f} A\".format(ts.frame, rmsd, distances.min(), distances.max()))\n",
    "\n",
    "print(\"Wrote PDB trajectory {0} with distances in bfactor field\".format(pdbtrj))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Using the MDAnalysis.analysis Modules"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "MDAnalysis comes with a number of existing analysis code in the MDAnalysis.analysis module and example scripts (see also the Examples on the MDAnalysis wiki)."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### RMSD"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "As an example we will use the MDAnalysis.analysis.rms.rmsd() function from the MDAnalysis.analysis.rms module. It computes the coordinate root mean square distance between two sets of coordinates. For example for the AdK trajectory the backbone RMSD between first and last frame is"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "6.8527748446562393"
      ]
     },
     "execution_count": 58,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import MDAnalysis.analysis.rms\n",
    "u = MDAnalysis.Universe(PSF, DCD)\n",
    "bb = u.select_atoms('backbone')\n",
    "A = bb.positions  # coordinates of first frame\n",
    "u.trajectory[-1]      # forward to last frame\n",
    "B = bb.positions  # coordinates of last frame\n",
    "MDAnalysis.analysis.rms.rmsd(A,B)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Superposition of structure"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In order to superimpose two structures in a way that minimizes the RMSD we have functions in the MDAnalysis.analysis.align module.\n",
    "\n",
    "The example uses files provided as part of the MDAnalysis test suite (in the variables PSF, DCD, and PDB_small). For all further examples execute first"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "import MDAnalysis\n",
    "from MDAnalysis.analysis import align, rms\n",
    "from MDAnalysis.tests.datafiles import PSF, DCD, PDB_small"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "In the simplest case, we can simply calculate the C-alpha RMSD between two structures, using rmsd():"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "28.20178579474479"
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ref = MDAnalysis.Universe(PDB_small)\n",
    "mobile = MDAnalysis.Universe(PSF,DCD)\n",
    "rms.rmsd(mobile.atoms.CA.positions, ref.atoms.CA.positions)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Note that in this example translations have not been removed. In order to look at the pure rotation one needs to superimpose the centres of mass (or geometry) first:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "21.892591663632704"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "ref0 =  ref.atoms.CA.positions - ref.atoms.CA.center_of_mass()\n",
    "mobile0 =  mobile.atoms.CA.positions - mobile.atoms.CA.center_of_mass()\n",
    "rms.rmsd(mobile0, ref0)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The rotation matrix that superimposes mobile on ref while minimizing the CA-RMSD is obtained with the rotation_matrix() function"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "6.80939658647\n"
     ]
    }
   ],
   "source": [
    "R, rmsd = align.rotation_matrix(mobile0, ref0)\n",
    "print rmsd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[[ 0.14514539 -0.27259113  0.95111876]\n",
      " [ 0.88652593  0.46267112 -0.00268642]\n",
      " [-0.43932289  0.84358136  0.30881368]]\n"
     ]
    }
   ],
   "source": [
    "print R"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Putting all this together one can superimpose all of mobile onto ref:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": [
    "mobile.atoms.translate(-mobile.atoms.CA.center_of_mass())\n",
    "mobile.atoms.rotate(R)\n",
    "mobile.atoms.translate(ref.atoms.CA.center_of_mass())\n",
    "mobile.atoms.write(\"mobile_on_ref.pdb\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Exercise 5"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Use the above in order to investigate how rigid the CORE, NMP, and LID domains are during the transition: Compute time series of the CA RMSD of each domain relative to its own starting structure, when superimposed on the starting structure."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The code contains a function superpose() and rmsd(). The latter is marginally faster because we only need the calculated RMSD and not the full rotation matrix. (We are calling the lower-level function MDAnalysis.core.qcprot.CalcRMSDRotationalMatrix() directly, which has somewhat non-intuitive calling conventions). superpose() also does the superposition of the mobile group to the references (but alignto() is actually a more flexible tool for doing this). Otherwise it is mostly book-keeping, which is solved by organizing everything in dictionaries with keys CORE, NMP, LID."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {
    "collapsed": false
   },
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/anaconda2-4.1.1/lib/python2.7/site-packages/MDAnalysis/core/qcprot.py:9: DeprecationWarning: qcprot has moved to MDAnalysis.lib.qcprot and will be removed from here in release 1.0\n",
      "  DeprecationWarning)\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAVkAAAFLCAYAAAB87P9LAAAABHNCSVQICAgIfAhkiAAAAAlwSFlz\nAAALEgAACxIB0t1+/AAAIABJREFUeJzsnXd4FNXXx79304GQ0IvUANI7CNKbID0ggoCiglJEBV9Q\nQZSi8hMLFkSaBAQEKdIEEZASmvTQi1Qh9IRACOnJnvePs7Ozm0Y2O7uz2dzP88wzfebMZvPdO+ee\ne44gIkgkEonEMRj0NkAikUjcGSmyEolE4kCkyEokEokDkSIrkUgkDkSKrEQikTgQKbISiUTiQHQV\nWSFEGSHEDiHEGSHEKSHEu5kcN0MIcVEIcVwIUc/ZdkokEklO8dT5/ikA/o+IjgshCgA4KoTYSkTn\nlQOEEJ0BVCKiKkKIJgDmAGiqk70SiURiE7q2ZInoDhEdNy0/BnAOwFNpDusJYLHpmIMAAoQQJZxq\nqEQikeQQl/HJCiEqAKgH4GCaXU8BCLdYv4n0QiyRSCQuiUuIrMlV8DuAUaYWrUQikbgFevtkIYTw\nBAvsEiJan8EhNwGUtVgvY9qW9joyCYNEInEIRCRyeq4rtGQXADhLRD9ksv8PAIMAQAjRFMBDIrqb\n0YFE5LbTpEmTdLdBPp98trz4fPaia0tWCNEcwEAAp4QQxwAQgI8AlAdARDSPiDYJIboIIS4BiAXw\nun4WSyQSiW3oKrJEtA+ARzaOe9sJ5kgkEjdl8eLFWLVqFTw9PeHl5YW2bdti+PDhECLHXoBso7tP\nVpI92rRpo7cJDsWdn8+dnw1w/eeLiYnBiBEjEBcXZ962atUq7Ny5EwsXLkT+/Pkden+hhc/BFRBC\nkLs8i0SSGyEiGI1GeHg88eUU4eHhuHDhAtq1a+fw1uTChQsxePBgNGrUCB999BFu376NcePGISYm\nBnXq1MG6detQsWLFTM8XQoDs6PjS3amsoXOaJJK8SHx8PB0+fJiSk5N1s8FoNFK7du0oKCiIIiMj\nMz0uOTmZpk+fTvny5SMA1LNnzwyPj46OplWrVlFERIR5W1JSEg0bNoyeeeYZunbtWrZta926NQGg\n+fPnm7edO3eOnn76aQJAQUFBlJSUlOn5Jm3JuTbZc7IrTVJkJXmN5cuXU3BwMOXPn58A0EsvvURG\no1EXWw4cOEDgjmsaNmxYhsccO3aMGjVqZD7O19eXANBTTz1FS5Ysob1799Lx48fpww8/pICAAPO+\n/fv3U2JiIvXu3dt8bsOGDSkuLu6Jdl25coUAkJ+fH0VHR1vte/DgAXXt2pU2bNiQ5TWkyEqRleRB\njhw5YhYcAOTh4UEA6JdfftHFnqFDh5ptEULQoUOHzPtiYmLo//7v/8w2lilThjZs2ED//fcfNWvW\nzOo5LKcSJUoQAPLy8qLGjRsTAAoMDKTy5csTAOrTZzhFRlr/qNy7d49mzJhB169fJyKiyZMnEwAa\nMGBAhnZn50dJiqwUWUkeZMqUKebX7fDwcFq4cCEBoAIFCtClS5eIiOjGjRt08eJFh9sSGxtL/v7+\nBICCg4MJADVq1IgSEhJo/vz5VKZMGQJABoOBRo0aRY8ePTKfm5ycTF999RV169aNmjRpQpUrV6YX\nX3yRDhw4QImJifTuu++aRbdw4cJ09OhROnnyJPn5FSPgFhUo8IguXOAW6saNG6l48eIEgIoVK0Z7\n9uyhoKAgAkBbtmzJ0PZjx4gePMj6+aTISpGV5EGUFuC6deuIiFtkffv2JQBUq1Yt834vLy8KCwtz\nqC2LFi0iANS0aVOKiYmhp556igBQkSJFrF7vjx49mqPrK26REydOmLcNHXqEAFYwIdbTM880Md3r\nOfL0PEbAMBLCYHY5pKSkWF0zNZVo+nQiLy+ifv2IsmrQSpGVIivJY0RFRZHBYCBPT0+rVmFUVBSV\nLVs23Wt3z549c3yvy5cv0507d7I8RulYmjdvHhERrVy50nzvqlWr0tKlS9OJnD0YjUR16pBJZFNN\n81fI07MHeXgkm8UXmE2AF40bN87q/Dt3iDp1IvNxb71FlFWfoRRZKbISNychIYFu3bplXldErHXr\n1umOPXbsGA0fPpyWL19Oly5dIj8/PwKQo1bkxYsXyc/Pj/z9/WnPnj2ZHgOA8uXLZ+5YMhqNtGDB\nAlq+fLmm4qqwdy8rV/HiRHPn8rKfXxJ5ebHgdu1K5OPD2wsXPkmnTqk/EufOEZUqpewjMr0IZIkU\nWSmyEjdm//79VKFCBfL29qa9e/cSEdGQIUMIAH3xxRdPPH/MmDEEgLp3757lcdu3b6eXX37ZSsz7\n9+9vbpHmy5ePtm3bZnVOQkICDRo0iADQq6++avvD5ZD+/Vm5JkzgVm2PHmRulY4ezdsOHlTFtEIF\nolOnWGBLlOBtLVoQ3biRvftJkZUiK3FDkpOT6csvvyRPT0+z0NWvX59SUlLMPs9jx4498Tp37941\nt2YPHz6c4TFJSUnmHvsWLVpQcnIyHTt2jACQt7c39enThwCQj48PTZw4kf7++2/asWMH1ahRw9yh\ndfDgQa0/ggy5c4f9qAYDkRIqe+cOUefORJ9/bu1bvXmTqHFjVjl/f275AkTt2hHFxmb/nlJkpchK\n3ITk5GQKCQmhXr16UcGCBc3iOmrUKLOvVeltL1GiBKWmpmbrumPHjiUA1KFDhwyD7kNCQqx8uB98\n8AF17tyZANB7771HqampNGLEiAzDrKpUqUKhoaFafxSZ8vnnrFrZdTPHxRG99BKZW7q2CiyRFFkp\nshK3ICkpydxiVKZq1aqZA+VXrFhhtW/QoEHZvvbdu3fNIVZt2rSxGkWVnJxMlSpVIgA0cuRIcywr\nAPL396d79+4REftZ161bR6NGjaImTZpQyZIlady4cdkaEKAVa9eqvtZMIrIyxGgkmjGDaNQo2wWW\nSIqsFFlJruTQoUO0aNEiunHjBiUmJlKvXr0IABUsWJBmz55N//33n9XxRqPR3IsPgJYuXWrT/fbv\n308lS5YkAFShQgXzYIElS5YQAKpUqZI5ZlW5x5QpUzR7XntZuJBdBEo0gDMHtkmRlSIryWU8evTI\n3LK0HNkUGBiYqd+UiOjEiRNkMBjIw8PD3MK0hRs3blgNaw0ODqYqVaoQAAoJCSEiFvMhQ4ZQixYt\nrMLD9GTFCjK/7n/yiXMFlkiKrBRZiUsSGxub6ZDNWbNmkTIqSUmUooxmehJ//vnnE8faZ0VcXByN\nHTvWnDcAAJUvXz7LBCl60749K5VeDWspslJkJS7GoUOHyMfHh4YMGZJun9FopNq1axMA+u233ygh\nIYFCQ0PNY+2dxa1bt2j06NFUtmxZWrNmjVPvbQuJiUR+fqxUd+/qY4O9IivzyboZGzYAjx8D/fvr\nbUnepXPnzti8eTMAYPXq1ejdu7d53759+9CiRQsUK1YMN27cgLe3t15m5gr27QNatACqVwfOntXH\nBnvzybpCIUWJRiQkAH37AgMHAvfu6W1N3uTIkSNmgQWA4cOHIyIiwrw+Z84cAMCQIUOkwGaDXbt4\n7uLFF7JEiqwbcfQoCy0RcPKk3tbkTT7//HMAwJgxY9CmTRtERETg7be5RF1kZCRWrlwJIQSGDRum\np5m5htBQnrdurasZdiFrfLkR//yjLp88CXTooJ8teZGTJ09i/fr18PX1xdixYxEfH4/atWtj5cqV\nOHLkCLy9vZGUlIQuXbqgQoUKepvr8iQnq9/p3CyysiXrRliK7IkT+tmRV5k6dSoAYOjQoShZsiQq\nVqyIOXPmwMfHB1euXMH58+cBwNyylWTN0aNAbCxQtSpQsqTe1uQc2fHlJhABJUoAivuvXj3g2DF9\nbcpL3Lx5E2XLloWnpyeuXLmCMmXKmPfFxsbixo0bCA8Ph5eXF1rn5maZE5k2DRg/Hhg6FJg7Vz87\n7O34ku4CN+HyZRbYwEDg4UPuiU1OBry89LYsb7B+/XoQEbp27WolsACQP39+VK1aFVWrVtXJutyJ\nO3R6AdJd4DYoroI2bYCgICApCbhwQVeT8hTr1q0DAPTq1UtnS9yDlBRg715ezu0NfymyboIiss2a\nAXXq8LL0yzqHBw8eYOfOnfDw8EDXrl31NsctCAvjeO/KlYHSpfW2xj6kyLoJ+/bx3FJkZRiXc9i0\naRNSUlLQqlUrFClSRG9z3IKDB3nesqW+dmiB9Mm6AQ8fAmfOAN7eQMOGwJ07vF2KrHOQrgLtMQVi\noGZNfe3QAtmSdQMOHuTogoYNAV9f2ZJ1JvHx8fjrr78AAD179tTZGvfh33957g59hbIlm0s5eRIY\nNw4oWhS4dYu3NWvG80qVgHz5gJs3gfv3AfkG6zi2b9+O2NhYNGzYEOXKldPbHLdBEdlq1fS1Qwuk\nyOZSvv8eMDWgzDRvznODAahdm1u4J08Cbds63768wtq1awEAwcHBOlviPsTGAjducPihOwyMkyKb\nS9m/n+cffwzExwOenkD37ur+OnWkyGrJ1atXMXPmTHTs2BEdO3YEEeGzzz7DwoULAUh/rJYooYeV\nK/P3OrfjBo/g3qSmAu++C1SpAowezduiorhjwNcX+OQT7vBKi/TLakdYWBi6dOmCu3fv4ttvv0WD\nBg1QpEgR/P333xBCYNq0aajpDj00LoI7+WMB2fHl8uzbB8yaBXzwARATw9sOHOB5o0YZCywA1K3L\n8/XredLCjkGDgLt37b9WbmLLli1o3bo17t69i8aNG6N48eIICwvD33//jcKFC+Ovv/7Chx9+qLeZ\nboUSWeAuIitbsi7Oxo08T04Gtm0DevVSXQVKR1dGNG3KI2V27QKCg4GePYE33wRq1eJkG6dPA4cO\n8bHDhrEfNzMePwb69eOONABYvNj+58oNnDx5Et26dUNKSgoGDhyIBQsWIDU1FQsXLsTRo0fx8ccf\no2LFinqb6Xa4U6cXAFl+xtWpXp3MReQGD+Zt7drx+tq1WZ+bksKlkP391WtkNM2ZY33e/v1EN2+q\n6++/rx4rBNHx49o+o6vSr18/AkCvvvoqpaam6m1OnqF+ff6u/fOP3pYwkDW+3FdkL13iv5CnJ89L\nliRKTiYqUIDX79zJ3nVu3GChbN+eqEQJPvfpp4k6d+blQoWIlOKn8+bxtoAAor/+Ijp9mu8vBFGn\nTrzv+ecd98yuwsWLF8lgMJCXlxeFh4frbU6ewWgkyp+fv2f37+ttDSNF1o1F9vvv+S/Urx9RmTK8\nHBLC86CgnF83MZHnRiNRhw5kbiXv3Uvk5UVWrdYKFXh5+HCiiAiiggV5ffv2zK9/+DDRd9+5zj9J\nThg2bBgBoNdff11vU/IU4eH8/SpaVG9LVKTIurHIKgL4669Ew4bxcrlyPB84UJt7nD+vCmuhQjx/\n910uv6yIbdGiqmBOncrbGjYksnyDNhqJliwhatRIPe+TT7SxMScYjUZq27YtNW3alBKVX5Vscvv2\nbfLx8SEhBJ09e9ZBFkoyYts2/u40b663JSpSZN1UZKOjWfwMBqLISKI//lDFCyCaOVO7e02YoF63\nXTt2SRARrVtH1KQJ0YYN6rGxsUSlS/OxX32lbp8+Xb2GwaDtD0FOOHr0KAEgAPTTTz/ZdO64ceMI\nAPXq1ctB1kky46ef+LuTQTV13ZAi66Yi+/vv/Ndp0YLXY2OJfH1VIQsL0+5esbFEtWoR1ajBgv4k\nNm4ks6/40CGifftUv/H33xOtX8/LbdtqZ6OtTJo0ySyypUqVotjY2Gydt23bNsqfPz8BoAMHDjjY\nSkla3nkn/Q+43tgrsjJO1kVRQre6deN5vnzqyK38+XnYrFbky8eDFk6dyl6eg65dgVGjOLHySy9x\nGfKUFGDMGN5eqRIfp+RU0IMNGzYAAAICAnD79m3MmjXriecsXrwYzz//PGJjY/HKK6+gSZMmjjZT\nkgZ3G4gAQLZkXZGYGNU/evq0ul15ldKzhaiQkEBUr57asm7enCgpifdFRfE2f399bAsPDycAlC9f\nPlqzZg0BoCJFilB0dLTVcVu2bKFu3bpR+/btqWXLluaW75gxY2TIlhO5fJnDtVJSiMqX5+/O+fN6\nW6UC6S5wP5H95hv+yzz7LHcoKcTFEY0Zw733rsC5cxxtUKIE9worGI2qayMmxvl2zZ49mwBQz549\nyWg0UosWLQgAjRgxguLj44mIaP78+eTh4WEWVgAkhKAZM2Y43+A8TGIiUfHiZA5RFIJdT8oPtisg\nRdbNRDY+nr9sAPs+XZ2ICO6kS0vFivwMFy4436YuXboQAAoJCSEiot27d5uFtGzZsvTaa6+Z1z/4\n4APaunUr/f3333Tp0iXnG5vH2bGD0g2OqVFDb6uskSLrZiI7ezb/VerWtW7F5jaaN+fnCA117n0f\nP35sDr+6YzFaY9OmTVSnTh2rVuusWbOca5wkHWPH8vdk7FiiI0eIPvuMO1JdCXtFVuYucCGSk4Ev\nv+Tljz4CRI4rveuPUvzOGZ1fRqMR58+fR/HixbF3714kJiaiSZMmKFGihPmYzp07o1OnTli1ahUW\nL16MYcOGoUePHo43TpIlSk7kLl24skfDhvra4wikyLoQy5cD//3HPasvvKC3NfahlcheuHABy5Yt\nQ7FixTBy5Mh0+yMiIvDiiy9i165dAABPUwLS7pbJdU0YDAb069cP/fr1s88oiSZcv8616fz91YTz\n7ogUWRdi6VKejx0LeHjoa4u9lCrF85yI7NWrV7F+/XqsWLECB5S8jgA6dOiAqhaxPcePH0fPnj1x\n/fp1FCxYEEajEY8fP4a3tzdefPFFex9B4mCUVmyHDpmn7HQHpMi6EKdO8dwdKhkoLdnbt7N/zrlz\n5zBgwAAcP37cvK1AgQIoUqQIrl27ho0bN5pF9tSpU2jWrBni4+PRpEkTrFmzBqVKlcLt27chhEAp\nReUlLsumTTzv0kVfOxyNHIzgIkRFcasvXz7AHVKU5sRdMHr0aBw/fhz+/v7o27cvli1bhjt37mDa\ntGkAgI3KCA0AM2fORHx8PHr37o3Q0FCULl0aQgiULl1aCmwuIDER2L6dl59/Xl9bHI1syboIp0/z\nvGbNrBNo5xZsFdlz585h69at8PPzw9WrV1HEYuhZp06d4OHhgT179uDhw4fw9fXFypUrAQCffvop\nfH19tTZf4mD27OGCiXXqAGXK6G2NY5Ei6yIoIlurlr52aIWlyBI9OVJi5syZAIBXXnnFSmABoFCh\nQmjZsiVCQ0OxefNmeHl54eHDh6hfv76srZVLUVwFnTvra4czkCLrIribyBYsCPj5cWslJobXMyM6\nOhqLFi0CALzzzjsZHtOtWzeEhoZi48aNePz4MQAWZEnuZPNmnucFkXWDF1P3QOn0cheRFSL7LoOF\nCxciNjYW7dq1Q61MPoBupkw5GzduxKZNm2AwGNC/f38tTZY4ievXgXPnOHQrqzp17oIUWReASG3J\napldS2+yE2GQmpqKH3/8EQDw7rvvZnpc1apVUaVKFURHRyM5ORkdO3ZEyZIltTRX4iS2bOF5u3aA\nl5e+tjgDKbIuwK1bwMOHQOHCXEnWXXhSS3bv3r1o06YNrly5ggoVKphbq5lhuV+6CnIvish26qSv\nHc5CiqwLYOmPzc1DadOSmciGh4cjODgYLVu2xN69e1G0aFHMnj0bHk8YgaGIbIECBRAcHOwIkyUO\nJiWFS9sDeUdkZceXC+BunV4KaUd9GY1GzJs3Dx988AFiYmKQP39+jBkzBmPGjEHBrHrGTLRp0wYT\nJkxA7dq1kS9fPgdaLnEUhw4B0dFAlSpAUJDe1jgHKbIugNLp5U7+WMC6JZucnIwXXnjBXLGgV69e\n+Omnn2waOGAwGPD55587wlSJk8hrrgJAugtcgtzakk1MTESPHj3QpUsXnDx5Mt1+VWQJw4cPx4YN\nG1C4cGGsXLkSq1evliOz8iB5UWT1zgEbAuAugJOZ7G8N4CGAMNP0cRbX0ip9pFNJSSHy8+OcmlFR\neltjG1OmTDHnZ/Xw8KB3332XHj58aN5//jw/V+HC9wkA+fn50cGDB62u8e+/RKVKET3/PFdakLgv\nkZFcydjbW5+KGTkFuTlpN4AWAOo9QWT/yOa1NPlAnc3Fi/xXeOopvS2xjfPnz5O3tzcBoL59+5LB\nYCAA1KhRI4qLiyMiokePyJTtPpYAQWvXrk13HSVps1L9duxY1yo9ItGO5cvJXHY+N2GvyOrqLiCi\nvQAePOEwN+pvT09u8ceeO3cOkydPxtGjR0HEr/9JSUl4/fXXsWLFCoSFhaFixYo4cuQI3nrrLRAR\ndu/+E8BjAPkwdepMVKkSjLAw9ZpEwO+/83LnzkBqKvDNN8C8eXo8ocTR7N3L8w4d9LXD6dij0FpM\nAMoj65ZsJIDjAP4EUCOL62jxo+V0lDrz77+vtyWZc/78eSpWrJjZNRAUFEQAqGjRohQZGWk+7vjx\n4+Tn50cAaPjw4eTr60vAvwQQtW/Pz+nhQXTqFB9/+DBvK1WKKDWVaPp0Xh88WKcHlTiUjh3577th\ng96W2AZys7uAniyyBQDkMy13BnAhi+to84k6kagoovz5+a9w/Lje1mTMtWvXqGzZsgSA6tSpQ4UL\nFzaL7ZIlS9Idv2zZMqsKsKVKnU9XKG/AAD72ww95/e23eX3rVl5v1cqJDyhxGhUq8N/333/1tsQ2\n7BVZlw7hIqLHFst/CSFmCSEKE1FURsdPnjzZvNymTRu0adPG4Tbaw+zZnEDlueeAunX1tiY9kZGR\neO655xAeHo7mzZtjy5Yt8PDwwLp165CYmIiBAwemO6d///4ICwvDN998g379+qFy5Sr44gvgtdeA\n11/nhOTLlwNTpgCrVvE5ffrwvHJlnl++7Jzn05LERKBFC6B8edUFIlFJSACuXeOKH66eLzk0NBSh\noaHaXdAehdZiAlABwKlM9pWwWH4GwH9ZXMf+nywnEh9PVKIE/7Jv3aq3NRnz4osvEgCqW7cuPXjw\nINvnGY1GunjxIhmNRjIaiWJj1X2DB/MzP/ssz0uU4AgLIqLkZO78AohMfWe5htBQtaV++LDe1rge\np0/zZ1O5st6W2A5yc8eXEGIZgH8APC2EuC6EeF0IMUwIMdR0SB8hxGkhxDEA3wNwmwp4v/4K3L0L\n1Kvnmh0B69evx6pVq5A/f36sW7cOgYGB2T5XCIHKlStDCAEhuNqDwrhxnJR8/35e791brWfm6am2\ncq5c0ehBnMTu3erynDn62eGqXLzI86ef1tcOPdA7umAAEZUmIh8iKkdEC4loLhHNM+3/iYhqEVF9\nImpGRAf1tFcrjEbuRQe4aKKr5SuIjo7GW2+9BQCYOnUqKlSooNm1q1QB+vZV19PWO6xUieeXLml2\nS6dgKbLLlnHCH4nKhQs8lyJrI0KI/EKIXF5X1fns2QP8+y9Qtqy14LgK48aNw61bt9CkSRO8/fbb\nml//o4+4NVu6NNCypfW+3OiXTU4G/vmHl+vVA+LjgSVL9LXJ1VBEtkoVfe3QA5tEVghhEEIMEEL8\nKYS4B+A8gNtCiLNCiK+FEJUdY6Z7ocQL9urlevk0z549izlz5sDLywvz589/YmasnFC7NrBvH7Bj\nB7sILHlSS/bePe4sdCXCwoC4OKBqVeDjj3nbnDnsoZUwsiWbfXYCqARgPICSRFSWiIqDR24dAPCl\nEOJljW10Ow4c4HnTpvrakRGrTF3+gwYNyrRKgRY0bcqilJasWrK3b3NLqFcvh5mVIxRXQatWQI8e\nnBP47Fl+Y5Ew0iebfToQ0WdEdJKIjMpGIooiotVE9AKAFdqa6F4QubbIrl69GgDQR4mrcjJZtWQ3\nbQIePeIWcEKCc+3KCkuR9fIC3niD17/7Tj+bXIlHj4A7dwBfX/evTJsR2RJZIbhrhoiSszimzpOO\nkXCveWQkULw4oGF/Uo6YMmUKmjRpgnv37gEALl68iFOnTiEgIADt2rXTxaaKFbkj8No19nVasnUr\nz1NT1cxlemM0qu6fVq14PmIER1SsWwdoGW6ZW1FasZUru0e5e1t54iMLIfoB6C+E6C+E8BBCVBJC\ndBRCvCWEmC6EWC+EOAvgiOPNzf0ooUtNm+obVXD9+nV8/vnnOHToEP73v/8BANauXQsA6N69O7y9\nvXWxS2ntpKay0CqkpgJ//62uW+ZA0JPTpzmSoHx5oFw53la6NIeqAcDo0Wx7XiYv+2OB7LVkYwH4\nAEgCMBXAMQCjAFQEcBnANQA9ADzvIBvdCsVV8Oyz+trx3XffISUlBQAwe/ZshIeHY82aNQCA3r17\n62lahn7Zo0eBBxaphFxFZC1dBZaMHcuie+IEEBLifLtcibzsjwWyIbJEtNEUv7qaiMaBxXULgDsA\n1gE4TUSXiGiHg211C1zBHxsVFYWff/4ZANC4cWMkJSVhxIgROHjwIPz8/NBJ54zKGfllFVdBjRo8\ntxTZM2eAQYM48sDZZCayfn7A11/z8oQJeTtuVrZkbYSI7hPRDADzAPQH0FUIESSEKKS5dW5GXBy3\nbAwGoFEj/eyYPXs2YmNj0bFjRyxZsgQGgwF//vknAKBz5866189SRNayJauI7Pvv8/zkSdVnO348\nx6UuWOA8GwHu0FHsSiuyAA+0aNmSffBz5zrXNlciL8fIAnYMRiCiGCKaDuAlcIasjZpZ5aaEhXG1\nztq1gQIF9LEhPj4eP/zwAwDggw8+QNWqVTFo0CDz/hdeeEEfwyxQ3AVKS/bRI/Zle3hw+FZQECdk\nOX8eePxYFTpnD8WdNYuLArZqlXErTQjgnXd4eft259rmKhDJlqytgxHSddUQUTwR/QRgWGbHSBhX\ncBWEhIQgIiICDRs2NEcQTJo0Cd7e3vDz80PXrl31M85E2pZsaCj/ODVtCgQEAA0a8PawMK4ZlZjI\n61evOs/GuDjg2295ecKEzI9r3Zrn+/YBSUmOt8vViIzkH6KAAKBYMb2t0QebByMIId4RQpSz3CiE\n8AZQXAixCMCrmlnnZugtsjdv3sQEkyKMHz8eyu9hhQoVsGfPHoSGhiIgIEAf4yywFFmjUS2+17Ej\nzy1F1hQQAcCxLdm7d4HPPuNBBgAwfz4QEcFun+eey/y84sWB6tVZlI/kwfib8+d5/vTTrpejw2nY\nkrILgC/rRXBsAAAgAElEQVSAtwDsA3ALwFkAV8ARBj8DqG9PSjB7Jrh4qsOHD4lKluR0b3oUDDQa\njdS1a1cCQN26dSOj0eh8I2ygWDH+rHr2JMqXj5f37+d9mzfzepMmRAEBaopBDw9Ol6g1jx8T1a/P\n9/D2JvriC6IyZXg9g7Jl6Rgxgo+dOlV721yZ27eJqlfnZx86VG9rcg70qowAwAtAKQCB9hig1eSq\nIpuYSPT990RFipC5YGJqqvPtWLx4MQGggIAAunHjhvMNsJFmzVTxBHhdEdC7d6331axJVLo0L1+9\nqq0dKSks9ABRYKD1fWvVyt7fcsUKPv6557S1zV6MRqJjxxzzw2QpsLVrE927p/09nIVuIutqk6uK\nbJ8+6j9lq1ZEJ0449/5RUVG0ePFiKlSoEAGgBQsWONeAHLJlC9FLLxH98APRhQssCJY89ZT6uX78\nMVGLFry8fbu2dijVdAMD+Q1k0yZV0Fetyt417tzh4/Pl4x9dV2HxYrZr2jRtrxsdTVSjhnsILJEU\nWZcW2UeP+BXW05Pojz/SC4UjOXfuHHXt2pU8PDzM9bY6d+7s8m6C7NK9uyqyR44QDRrEyz//rM31\nHzwgGjiQzKXKLcU7Otr2H0ulVbdvnzb2aUG/fmxTy5baXdNoJOrfn69bo0buF1gi+0U2D44kdh57\n9vCQysaNge7dneP4T0pKwqeffoq6deuaY1/bt2+PmTNn4vfff4e7BH8onV9ly/KyUlFBiwiDvXu5\n5trSpZyDYMkSwDKVQ8GCQJ06tl1TKTfnSrkMlBy4hw+rERr2smgR8NtvQP783CmZVyMKLJEi60B2\n7uR527bOu2f37t0xadIkJCUlYciQIbh9+za2bduGkSNH6j7IQEt69OC42bfe4h+voCDebm+EQWQk\n0KkTcP06Rw4cOwa89JL99rqayN64AYSH83JCgjbDlP/9Fxg5kpd/+invxsWmxaWr1eZ2dpgGGjtL\nZMPDw7F161bkz58fGzdudPlqvfbQoAGHRSlJz7US2UWL+Lpt23LomFZJ1dPGy+qUf8eMkqhIYd++\njPNp7N8PrF/PI+wuX+acDG++mf44IuDVV/mzGzCAhzlLTNjja0g7AXgFwEgtr2nDve13vmhIVBSR\nEEReXtbVWh2JEkHQtWtX59zQhbhxg/2AxYrl/BpGI9HTT/N11q3TzjYFV/LLjh5N5uqxAFFwcPpj\nDh0iMhhU3zfAn09GHDrE+4sWZZ+1OwEX88keBSeNyfPs3s1fy6ZNrau1OhKlVnxbZ/onXIRSpQAf\nHx4gEBOTs2vs3s1DQEuXBhwx8K1ZM54fPqz9tW1FacmOGcPzffusy+WkprIrxmgEXngB+P13wN+f\nP58bN9JfT8k0NmgQ+6wlKlqLbBkAee8/PAP08McqIuvOboLMMBjs7/xSkrgMGWJRe2zXLk7rrwEN\nG/L86FFNLmcTRCyYABd6DAtjX3b//jwqLSLCOuvZzz/zCLUyZYBffmGhVVweO9Lk24uL484uABg8\n2OGPkuuwt1rt+0KIYNNybQBhRPSrJpblchSRdZbeXb9+HVeuXEHBggVRr14959zUxbBHZCMjgdWr\nWXiGDDFtPHKE/4CvvaaJfXqJrNEIPPMM+1zj4vj+ycmcqCggAGjenI/bt4/nERFcURjgEjpKMiMl\nwiJtspvVqzmJT5MmQM2ajn+e3Ia9Ldn9RLQOAIjoFDjXbJ4nMpI7Cnx8nJece9euXQCAVq1aOaTC\nbG7Ans6vxYu5Q6pzZ65yAEB9r9+5U5OiYnXqcAtZyR7mCBYsAAoVAo4fV7ddu8a/F4cOcTVdxVWg\nfDdbtOD53r0syO+8wwnSO3bkFqxC+/Y837HD2rWgpJg0/zhJrLBXZH3SrOufXcQFMOkdnn2Wy6k4\nA6f6Y5OTuUnkYtgjsr+a3r+ses7PneN5UhJw8OCTL3L/PrBsGTBsGMd/pUlw6+sL1KrFQmYpgloy\nZw4nCF+/Xt2mPAYAfP89uwIA1UestGR37wYGDgRWrGBbf/zROra7Vi2gaFH2ySquhcuXOSwtXz6g\nXz/HPFNux16RrSKE6CqEqCyE6AoggyLPeQ/ldcqZ/tidJv+E5v7YtAWqUlI47VRgIMfsOEotckBO\n3QVxcfzm4eGhZvoCYK1OSgmEzLh/n5uqAwcC8+bxO/nIkemMcaTL4OFD9bpnzqjblUxYAQHcAlXK\nwSgiW78+i+rFi8Dy5dzBtWlT+jhXg0H9Tivf8XnzeP7ii7LDKzPsElkimgNuvb4DThTzkyZW5WKS\nk7knFnBMD3VGXLt2DVevXkVAQADq1q2r3YWVd8+JE9Vt33zDTfXkZH7Hrl+fe09coFpgTluyJ06w\n+TVrpokEUdQJUF9PAFaqtKV0Q0KAW7eAqlWBadN4tERCAjBqlNVhjhTZXbvUzq2MRHbiRHWkWtGi\nakpJb2/22Srbd+7MvIGg+GV37OARY0pO3WHDtHsOtyOnsV8AxgLYCg7bmgjAw55YMnsnuEic7MaN\nHC9YvbqduQpOncp2QOWiRYsIAHXv3t2OG6Zh2TIO9FUCJL/7jm3y9ub1BQs42DJ/fl4fN067e+eQ\n6Gg2xdfXts9+xgw+b/Bgi42PHpE5cQFA5OenZneZMIFvsmsXr6ekEJUvz8dt2sTbbt0i8vfnbX/8\nwduSkujI1vvmcf1a8+676p/L05MoKYm3K8lztm0jOn6cqHBholGjrM/dupWod2+i8+ezvseFC3yt\nwoXVJD1jxmj/LK4EdEx12Ns0F+BqtR/aY4i9k6uI7EsvkX25Qx8+JBo5kgXO0zNbGTYGDRpEAGj6\n9Ok5vGka1q3jzDYAUa9e6n9u2bKULjnojh3qsatXp7/WuXNEf/7ptOw4SkrJefOI9u5lrXwSSnKZ\nWbMsNirR9bVqEVWrxsv//MN/D19fXq9alSghgT8vgKhSJevch99/z9vLlyd65RWiwEAyenlRV8Mm\nMhg4T2122L6daOLEJ6ckrFmTb6cMIDhzhrcXLcrrSobLlJTs3TcjjEY1l66SglIRc3dFT5EdCqA7\nAINpvYc9htg7uYLIRker/385ymv6zz9qZm9lekLuvkuXLpG3tzcBoLNnz+bIbiuuXyfy8eF7jx/P\n277+WrWnfPn0yvXNN7yvQAGi335j4V23juj559Xz1q+337Zs0Ly59cf31FNPFlolLd+hQxYblTyA\nL75INGwYmXMCTppkfYPPPiPq0IGXv/3W+sLJyUR161ofD1C0RyBVxGXau/fJzxMdrf5wZJVaUUmn\n6OfHeWsBopUriSIi1D+NVr9zyo9S0aJE4eHaXNOV0U1k+d4YAC6g+BeAtQBaAvCy55p22KLNJ2oH\nCxfyJ9qqVQ4voGSqbtpU/cf98ccsT+nTpw8BoEGDBuXwpmmYO5fv+/zz1v+VEydyE2b37vTnGI1E\nffumExOradgwbex7AqdPE33wAZuj5H2dNCnz42Ni1OHPCQkWO8aP55MnTiRaupSX27RRm4WK2Cru\nk3z5eCx1Wk6d4vfwL77gVr0pR2MY6tGPX8XxZ/fff0RxcRna9+mn6kc4YIC6PTGRKy788guvL1vG\nx3TsyM8PEE2ezK15gKhRI1s/ycw5fJhdEKGh2l3TldFVZM0X4bI0rQFMADBfi2vmwAYtPk/biI/n\nrNzffUdkNFL79mR+VbUiKYlbRo0bc0uwSRMeLL5jh3rMgwdq8tlHj4imT+eLjRiR6e337NlDAMjP\nz4/CtWpSKMlAZ85Mvy+rplBMDNH77xN160bUujX/YHzxBdFff5H5VdrJ7N7Nt86fn1t6WR3ToEGa\nHcHBvOO337i5ZvmD0agRfxavvKJuy259lQcP6EExThhwI6C6uZlqbJE+qWtUlHV5nYIF1R+CRYt4\nmxDs8n3jDV7/8kt134svcn5dgE2V5AyXEFlXmHQR2S1bzP8B0SPHkYCRvL2Jou4kEu3Zw2r73nuq\nLzPtVLWqKlxr1pBVM1gRpzZtMrx1amoqNW7cmADQZ+PHE126ZP/zGI1EpUrxfU+ftv96ROwAVJTi\nv/+0uaYNKMm9R47MeP+332aikVWr8o5jx3g9KEj9u61Ywdvu3iUqVIiVzoYs3ieXnqRY+KX/PqQp\n/jZhAm9u146oTh1eVvrVmjZVT/P3JypRgpcPH+YJYDfImDG8nNfqi2mJFFk9RfbHH63+SX7CCFpT\n5X21CqDlVL06UUgId8/u2aO+dh4/ztcaPpzXP/+c169d4/XixTO89fLlywkAlSpVipK7duXejsOH\n7Xsepeu4aFFtO6qUIlkhIdpd80kYjUTvvEORA94hgzCSpyfRxYvpDxswgNK/fSQl8RuFEOpr/Ouv\n84FBQdY9UGfO2PzeHB9P1NDjGPXBSiqPq7QArxEBlDB+svmYiAj2owIcZDJ5Mi+/8QbrPsC/Xb17\nq1+xwED+TXv8mNc9Pdl9AGTcJynJHlJk9RRZJWamUydKNnhZi2q1avyONnUqx8ekrbindKYonUtK\nS0npfTEa1fCoyMh0t+7fvz8BoF+U5g7ArWZ7mDePr9Onj33XSYvyY2TpVHQ0ipMSoE+7HSSAvTRp\n+waV1IZhYRYbz57ljRUrqtt27+ZX++wW9noC337Lr/OLFhG9FcRvLTFlq5l/3BSXb+fOfPzJk7xe\nrJjqGnj3XRbUevV4vWdP9foVK5KVy1iJNJDYjkuILLhqrY8W17LDBvs/TVtRes/XraNv2m+i06hB\n/7YYzLWrn9QS3LGDzC2jixd5uXBh6/iaRo14+5496U6vUKECAaB7SlcvwP5ee1qgSrMuI3+sPSii\nVaKEat+9e9mLr8oJjx6pvV4AxQx4k4oXJ3PrbuxYPuThQ97m45OmwOHq1byjSxfH2JeGsaOSKAKm\nEAKT26FBA7JyDxiNau5XJXxZ+cEID+ffbMWzQcSuceVr4eHhWgUccxuuIrLbAFwF8I0W18uhDRp8\nnDZSqRIpzQSl02vz5myem5KihmsNHszzvn2tj1EEdO5cq823b98mAFSsQAEyFirEx+TLx/MjR3L2\nLEajKkxa+WMtr634ek+dIjp6lFvpNWs6ph610r1epQrPCxSgiKsxNHSoKlBlyhB9+CEvP/NMmvOn\nTuUd//d/2tuWAWvWEM3BUL7nRx+ZS577+loHHbz/viqcrVtnfU3lIwAyT7QtyR72iqwm+WSJqAOA\nIAALtbheriApicelCwFUqoRbt3hz6dLZPN/DA+jTh5eVRCJWA+cBVK/O87Nn1W137+LAnj0AgDHl\nykE8eMDJSF59lfevWWP7swCc6ePWLR5XWaNGzq6RGUKoKZyWLQN69gRiY3ns58qV2b9OZCSPnU1b\n2+T334GSJYFu3YDp0zk/nxBcCbF5c+DxYxTduQpz53JircaNOcnJl1/y6cpQVzNKzgLl83cwzZoB\nK8DZVWjFCmzdQgCAVq0APz/1uN691eURI7K+pmXKwWrVtLJUkiNsVWUALcDDaGcB+NG03NEepddi\ngrNbsufPczOhQgUiUjvQM3CfZs6ePWpzA+DOLkvWr+ftHTvy+u7dRB4edK9QIWoL0HVlXGNICI+Z\nVHzBRNwEGj2a6Ndfs2eLEuvzwgs2PIAN/PKL9bMqHX+1aqX3V2fGlCl8jsHAPfsKll3tyvTGG7xP\nCV5u1sx8eEoKu4mVUa9LlqS5T8OGvCM7owU0omrlFLoNDhH4pPNhAtKPbUhNZQ9SrVpPfv0/ckT9\nKD74wHF25wXgTHcBgI8ATAIPo20H4DkALwD4EsA0ewyxd3K6yP7xh1kAld5cHx8bXaKpqeoAcEUc\nLVF6+8uU4XXLIa6WXcqxsdwjXrgwmV/JlW5nX19rQcqMgQP5+CcMfsgxlrGmZcrwyDJlfGZ2CmrF\nx5PZsQoQzZ/P22/eVD/82bOJunYlattWHY78+LGqpml6f27e5H4sq2GhqalZdjg6itdfJ5qBt4kA\n+tHvfWtzx4/nH/Pr18lozN5vUmys6hpZuNCRlrs/zhbZTIfOAuhjjyH2Tk4XWWWwwMiRZi207IzO\nNkog4+jR6fclJ6tDXM+fJ/LwIKOnJ33p4UFxithYnvfaa2TuALMU4iclb9m/Xx27eepUDh4imzRt\nynFJR4/yujK2/5lnnvzrFBJC5p4rgHt2iDjhAMABsZmhRHJkJ/ri6FE+tmzZ7D2TRoSEELUDv40c\nRkMqU8b0kRiNaqvfxiap0mVw4IBjbM4rOFtkPzFN3cC1vFoB6AzgQ+jY6UV6iKzyj/vDD7RzJy+2\naJGD6yijuzIakkmkRqGbourvt21LAKh9xYpEP/1kXQpXaV0DHLuj5Bzw9+cRZWm5e1cVZoBfk7P7\n6p4TYmOtW4exsWpMcUgId/enphIdPMg/DCNGEF2+zEKjZD/56ituovn48GenDNRfsCDz+yrvzgUL\nWn8Oq1dzXJdlAK0SkDp8uPbPnwX//kuUD48pGR6UDA8a+WoM71DcUgC/qWQy/DYjNm/msGsn5eZx\nW5wqsnw/tDf5YWea/LKTTNuEPYbYOzldZNu1449v0ybz0Pa0wQGa0K+f+k8G0NoRIwgAvfbaa+mP\njY9n94EQ6qgkJezhs8+sj338WM2M4u1N9NFH2U8LpSX/+5/V81mNIzVFBtBbb/FyqVLsjFRyPMyb\nxy1bDw+O3s8K5XOYMoXX799X3SuWeRWUsLmNGx33zBlgNPLvzWGwP3jXxG28Q2nBK5N893c6ThdZ\nV52cLrLKUNlLl8wNxoze+O1G6ewBiIKCqH+/fgSA5qYJ6zJz7Jh1Epft2/ncIkWsRfTNN3l71aoZ\nD4VyFgkJLPCNG6uukTJliN5+m6P1LQVGGRuqfOCKi6Nt2yffZ9cuMvuwo6Otk68GBHAL8dYtXvfz\ns6nFqBXBwUTfYRQRQHHjTT8GSnifEjjbsKFsmjoZKbJ6iGxsLJn9g8nJNHo0r379tQPutWqVKgZf\nfEHly5cnAHTy5MnsnW80qr3vr7zCmVJWriRzZ5ENY+4dTlISR1goImI0cgasgADu9FJcDcrgDWWa\nMSN712/dmo9/9VVu/RoMqv966VLuTLP09zqZb74h6gPT30aJKFGGpO3apf6oSCerU9FVZAEEmOaB\n9lxHi8mpIquMcTRFeStZ/pYudcC9zpwhAijVw4PWzp5NAMjf359SbMm8vG2bmsk5Xz51ULzWI7sc\nxePH/HpvSa1aqshev5696yitemUaOpT92gCnllQyb82Zo/0zZIPISKI3u5qiJfz9iW7fVv9mSUnq\n6ImXX9bFvryK3iL7ruVcz8mpIvv772TZ4lESRe/cqf2t7ty+TV97e9ObAME0Pffcc7Zf6Phxoh49\nVIHp0SN3v3Z+/DE/hy2JUo1G9Y9VsCB3/EVFcYteCDXjulJCQC+UxANKJ5ySie3qVbbT25sFWOIU\n7BVZTUZ8gUvQ5B2Ucp9VqgCAebTXU09pf6uQBQvwflIS/ipTBl27dsWgQYMwbdo02y9Uty7XiT50\niIc6LV5sXe85t/HWW0CXLsDUqdk/Rwjg6695dNj33wPFi3OhyOBg/ulJSODCkI74Q9pCixY8//FH\n6/UKFdjWpCTgpzxfszT3YI9CQ23JjrLnOlpMcGZLVkl7N3s2GY1qf43WnfMpKSlmH+xff/2l7cUl\nKps3qy38Tz7R2xp2V1i6NSz/9kqpgyJFrMP3JA4DLtKSzVtYtGSjooDERK5pnz+/trfZvHkzrl27\nhooVK6Jj2rwGEu3o0AEoW5aXe/TQ1xaA8y0oCAE8+6y63qwZ1+++fx9YtMj5tklsRoqsrTx6pCYQ\nefpp2xPD2MCcOXMAAMOGDYPBIP9UDsPDA/jzT2DtWk62ozc1agCBgbxcuzb/gisIAYwZw8vffQcY\njc63T2IT9v7n5mKnXg7YvJnTG92/zy2fp55ymMheu3YNf/75J7y8vDB48GBtLy5JT+3a7O90BQwG\nbrEC1q1ahd69gfLl+Y1q48asr5WSwp0GYWGcxUzidOwV2b/TzN2XqVOBzp05R94zzwBbtgAGA27e\n5N1a95X8/PPPICL06dMHxYoV0/biEtfnzTeBEiXUFJaWeHoCo0bx8mefAamp1vuJuEHQti3g48Nf\nzoYNgQYN2LclcS72OHRdaYIjO75iYtSk2NOmWSWa/uwz3vykHCzZ5c6dOzR06FAyGAwEgHbt2qXN\nhSXuxaNHaga3adPU7f/8Q1S3rnXHWdGiamz0smX62ZxLgV4dX0KIEhbLflkdm+tZswaIi+NQmg8/\n5JaECS1bsmvWrEGVKlUwb948CCEwbtw4tGzZ0v4LS9wPf38gJISXJ04ETp0C/vqLk6OfOMFhal9+\nCTx8CEREAF99xcfOnaufzXkVW1UZwHgAzwN402JbIwBt7VF7eyc4siXboQO3AjLIF6DE99tbDTQy\nMpIKFSpEAKhr1650Lk15aIkkQ5RscBUqqGkghwzhZEGWREereXLTVpOUZAl0aMmuBVARwHAhxB9C\niHkA6oHTHrofN24A27cD3t7Aiy+m252Tjq+9e/ciODgYCxcuNG+bNGkSHjx4gA4dOmDDhg2oJmuG\nSLLD118DFSsC//3HnVzvvw/8/DPg62t9XMGCwIABvDxvXtbXnDuXIxi48SKxl5yqM4DnTfMSALoD\naGiP2ts7wVEt2S+/5F//TMqyKPUBszN8/urVq9S3b1/z8FgANGvWLDp16hQZDAYyGAx0ypFJsyXu\nyT//cL7dr7/Oeqi0klc3MDDzLGN//01mX252kxC5OZBZuBwoskajmogkgxIpyclq3hWrEiZp2L9/\nP7300kvk6elJAMjPz89KbIOCgggAjRw5UvtnkEgsadyYv7CffsoiaplE/cEDtSSQFj4wN0GKrCNF\n9tgxMg9hzKBynVJeqnjxjE8/fvw4tTVVMgBAHh4e9PLLL9N1U7P322+/Ne8rVKgQRTqxppQkj6Kk\nc7ScOnbkop5KnTdlsoxayMNIkXWkyCqF7jNpYS5ezLsbNLDe/uDBAxo+fLg5DCswMJA+/PBDs7ha\n8vXXX5O/vz8tWrRIe/slkrQkJBC98w5X9qheXc08pkx+fkSjOHG4ueJvHidXiyyAEAB3AZzM4pgZ\nAC4COA6gXhbHafSRWqDUlcogh+GFC2oRVMv0o0ajkTp37mxuuY4aNYqiMqvfZcKm3LASiZZERnJS\nHKXsz+zZRFu28HLr1npb5xLoIrLgkK21AMIAnARwKiuhzOI6LcCRCRmeCy7S+KdpuQmAA1lcS9MP\nlm7c4I8nf/50roL4eKJ69Xh3gQJ/0tChw8ho6nDYsGEDAaCAgAA6ffq0tjZJJI7i4UPOOUxEdOUK\nf7lLl9bXJhfBXpFVo+ptYymA903imuMMFUS0VwhRPotDegJYbDr2oBAiQAhRgoju5vSe2WbrVp63\nbcvhWxa89x5w/DhQqNB9PHjwEubNi0HFihXw3nvvYfTo0QCAyZMno2bNmg43UyLRhIAAzjkMAOXK\nAV5eHJ8YG6t9erk8Rk5FNoKI/tDUkox5CkC4xfpN0zbHi+yWLTzv1Mlq85UrwJw5gLc3wWB4CUAM\nAOCjjz7CgQMHcPnyZVSvXh0jR450uIkSiUPw8ACCgoB//wUuXwbq1NHbolxNTkV2khBiPoDtAMwZ\nJ4hojSZW5ZDJkyebl9u0aYM2bdrk7EKpqcDfppw3afK4LljA8yZNrmHPnm2oWbMm+vfvj48//hjr\n168HAPzwww/w8vLK2b0lElegShUW2YsX85zIhoaGIjQ0VLPr5VRkXwdQDYAXVHcBAdBaZG8CKGux\nXsa0LUMsRdYuwsKAqCgu92EqMQOw9v7yCy8/fDgdADBy5EgMHz4cJ0+exMqVKxEcHIznnntOGzsk\nEr1QvvdKgvo8RNoG2pQpU+y6Xk5FtjERVbXrzioCmeel/QPASAArhBBNATx0ij/W0lVgUQdryxZO\nCFOuXCJOnZoJf39/vPzyyxBCYPHixQgODkaXLl0cbp5E4nAqV+Z5HhRZrcmpyP4jhKhBRGftubkQ\nYhmANgCKCCGuA5gEwBvcmzePiDYJIboIIS4BiAW3oB2PIrJpXAVK0qOSJf/C9evAoEGD4O/vDwDw\n8fFB//79nWKeROJwlJbspUv62uEGCI5QsPEkIc4BqATgKtgnK8DCqJvzRghBOXmWdDx6BBQuzMv3\n75tLf9y7x+kMiQje3lUQH38ZZ86cQY0aNey/p0Tiavz3HyeeKVVKzYKURxFCgIhyXAUmpy3Z53N6\nQ5dnxw52vjZvblVbafFiTnLUsOEtHD16GU2aNJECK3Ffypbl0MXbt4HHj4ECBfS2KNeSo6TdRHQN\nQCA4+1Z3AIGmbbmfnTt53r691ebVq3nu5fUrAKBfv37OtEoicS5KGBcgXQZ2kiORFUKMAg9IKG6a\nfhVCvKOlYbqhhG60bWu1+ZrpJ+TYsfkAgBczyC0rkbgV0i+rCTl1FwwB0ISIYgFACPElgP0AftTK\nMF2IiuIyHj4+QNOm5s2pqcBdU0xDYuJ/aNGiBcqUKaOTkRKJk8jDYVxaktMaXwKAZYnMVOTG8uAj\nRwL9+6u163fv5lxETZtaZZa/d48P8fZ+CCBFugokeQPLMK5t24CXXgL+/FNfm3IhOW3JLgRwUAix\n1rQeDM6olXsIDwdmzeLl117jmFjFVZBmpJjSuZqcfA1CCLzwwgvOslIi0Q+lJfvrr4BSKmnfPuDq\nVatiopKssbklK4QQAFaBY1ajTNPrRPS9xrY5FiUBDKDWPHqCyBLdROvWrVGqVCmHmyeR6M7TT/M8\nORkIDASKF+ead7I1axM2i6wpGHUTEYUR0QzTdMwBtjkWZcABAPzxB3D2LHDyZDp/LGAZJnhLugok\neYdy5YApU3i6ehX44APePmeOvnblMnLqkw0TQjTW1BJnkprKPiYAaNCAA2CHDMnQHwsAu3ZdAAB4\ne0fJqAJJ3mLiRJ4CA9mt5uPDDZQrVzI/JyEBaNkSePddp5npyuRUZJsA2C+EuCyEOCmEOCWEOKml\nYaFB9lcAAB8ISURBVA7l8GHgwQOOA/zf/3jbgQM8b93a6tCEhARs3BgGAOjRoxGKFCniTEslEteh\nSBGgXz9ujGRVVvzQIWDvXm7xxsU5zz4XJaci2wk8rLYdeDBCN9M8d6D4Yzt1Ap57DihvkTc8jT/2\nq6++QkwMj3bp399agCWSPMfw4TwPCQESEzM+5pjJe5iczIKbx7FJZIUQS0yLwUR0Le3kAPscg2WW\nLYMBePNNXvf2tvLHXr16FV988QWA0gCAsmU9nGyoROJiNG3KFRQiI4FPPmEhTcsxiy6avXudZ5uL\nYmtLtqEQojSAwUKIQkKIwpaTIwzUnIcPgYMHOQRFGdX1xhuc/aV/f8DPDwCQmpqKIUOGICEhAb6+\nFQEApUvrZbRE4iIIAYwfz8tff82iezKNp9BSZPfscZ5tLoqtIjsHXA2hGoCjaaYj2prmILZv546v\nZs2AggV5W4kSHDerZOQGJ+rduXMnihcvjcTEQAjBh0kkeZ5+/bhySPnynOD+mWeA8+d5X0ICR+oo\neZj/+Yc7lgHgzBnglVf4fy0PYZPImsK1qgNYQERBRFTRYgpykI3akkntLsvk3Fu3bsXnn38OIQRm\nzFgBIoHixWX8tURipkMHHoLepQv7Zpcv5+2nT7OoVqvGHcuPH6st3VGjeGCDVhVMcgk5zcI1QmtD\nnMbhwzxPkwBG4e7du3j55ZdBRJg0aRKCgloAkK4CiSQd/v7AW2/x8qZNPFdcBfXrcxgXwH7ZM2f4\nLRJgQY6Odq6tOpLT6ILcCZGa7KJatQwPWbJkCSIiItCmTRt8/PHH5oEIUmQlkgxo25ZjZw8f5iQf\nliLbghso2LMHmDlTPScuDli2TF2/fx+IiHCezU4mb4ns7dtcR75IEaBQoQwP2WNy1L/22mvw8PDA\n7du8XY6klUgyIF8+9a1wy5aMW7I7d3LWe4AjEgBg7lxu9Fy7BtSoAdSsya4FNyTHIiuEKCaEKKal\nMQ7nAo/cMo/JToPRaMReU8hJS9MXRLZkJZInoBQP3bBB9b/Wr8//Z8WKcUs1Lo4T4U+YwI2cEye4\nhdu7N7eAIyLUzPhuhq1xskIIMVkIEQngXwAXhBARQoiJjjFPYxRXgUWZb0vOnTuHqKgolC5dGhUr\nctiWFFmJ5Al07szztWtZTMuV4zp5QqguAwB45x12Lbz6Kq9368bRCd7evG4R3eNO2NqSfQ9Ac3BJ\n8MJEVAg8xLa5EOI9za3TGkVkM2nJKq6Cli1bQpiiDaTISiRPoHJl/p9SQrXq11f3KS6DChVYVAF1\n8E9MDMelb9vG+UJCQ7mAo5thq8i+AqA/EV1VNhDRFQAvAxikpWEOQXEXZNKStRRZBSmyEkk2UFqz\ngLXIvvIK0KMH5272MI2YrFaNh7MDwPz5LMS9e/O64rt1I2wVWS8iiky7kYgiAHhpY5IDsaElqyA7\nviSSbKD4ZQFrkS1aFFi/3lqEAWDVKg7rGjCA1xUXwqJF3CHmRgiy4YGEEGFE1MDWfc5ACEFZPktq\nKveEJiXxa0qaEsfXrl1DhQoVEBgYiPv378NgMCA5md1FBgPHW8vBCBJJJiQkcCdXbCxw/Tpgaw28\n1FQeQXbzJpeBatkSePSIBzycOMG+3sGD2dfrZIQQIKIcl9eyVTbqCiEeWd7fYtk37cEuRXg4C2zp\n0hnWkFdasc2bN4fBwA38O3d4X4kSUmAlkizx9eWOr/v3bRdYgF0JgwYBX3wBBAezf/fRI+tjfvmF\nh/PmstdKW4fVehBRQYvJ32JybXeB9MdKJI6lQwfOa5BTXn+dXx2jolhgfXyAevU4WXj16uxeaNGC\nqzRYkpoKLFmi5k9wMWxqn5mqIYQT0R3T+iAALwD4D8AUIorS3EKteEL4Vtr4WEAV2Vz2wymR5E6q\nVOHGUEwM/9MpYWAAp1Z8/nng6FEW2hUreJ6YyJ1rq1YBjRu7ZP5aWzu+5gJIAgAhRCsA0wAsBvAI\nQBap0l2ALDq9IiIicPbsWfj6+qJRo0bm7Uqnl2zJSiROonx5oFYtHrBgkbQJRYsCO3Zw5ZJbt3g+\nZQpHLqxaxcccOcJi7GLYKrIeFq3VfgDmEdFqIvoEQGVtTdOYLNwFK1asAAC0adMG3kpgNKS7QCJx\nKQoWZJ/shx8CRiNn89q6lavo1qzJUQk7duhtZTpsFlkhhOJiaA/A8olcu2sok5YsEWH+/PkAgNdf\nf91qn5JvOCh3JHGUSNwfLy9g2jQW11KlgEqV+B/15Zd5v1Ig1YWwNYRrAoAuACIBlAPQgIhICFEZ\nwCIiau4YM7NlW+YhXMnJPLLEaORQEItqtEePHkWjRlwg8ebNm/Dx8QHAkSN16nAgws2ban5viUTi\nIigjzDw92VXQuDFQsWLWlXRzgL0hXLZGF0wFMAbALwBaWKiaAcA7OTXC4Vy9qsbhpSn3HRISAqAw\nfHwOYsIEH/P22bN5PmiQFFiJxCXx9FRjK+vX546yq1eBy5f1tSsNNmfhIqIDRLSWiGIttl0gojBt\nTdOQTCIL4uPjsWzZMgA/4tatSpg+HfjtN44eWWIqGankJJZIJC6MhwfQrh0vu5jLwNYQrj+y2k9E\nPewzx0Fk0um1evVqREe3BTAABgN7E4YPB4YN49SWrVuzP10ikeQCOnQAfv+dRXbYML2tMWNrZ9Wz\nAMIB/AbgIKxHfLkuGbRkiQizZq0E8DMA4Icf+G+zfj0X4QSAkSOdbKdEIsk5HTrwXCmWqiSk0Rlb\n3QUlAXwEoBaAHwA8ByCSiHYR0S6tjdMMpSVbtap509KlS7F/f18AJdCiRQreeosTApUsyftLleLR\nfRKJJJdQqRJ3fD14YF2WXGds7fhKJaLNRPQqgKYALgEIFUK87RDrtCJNRYTr169jxIiPAQyAh0cq\nfvnFEwYDxzsvXcrziRM5WkQikeQilNasC1VZsCmECwCEED4AugLoD6ACgD/AJcJvam6dbXZlHMIV\nFwfkz8+KGRcHo8GA9u3bIzS0BIDlaN+esG1b7vB6SCSSJxAayjXHPDyAXbuA5vZHlTo1hEsIsRjA\nfgANwLkKGhPRZ3oLbJZcusTzSpUAT0/MmDEDoaGh8PXtCQDo1EkKrETiNrRpA4wdyz7Zfv1cYpit\nrYMRjACU0C3LEwUAIiLdIkozbcmuWgX07Qv06AHj2rUoV64cbt68iUKF4vDggR9OnOBBBxKJxE1I\nTmax/ecfTha+cSMnhc4hzh6MYLBIbZg25aFrhuxb+GP37duHmzdvolSpTnjwwA+lSgG1a+trnkQi\n0RgvL2D5ck4y89dfwJo1upqTc3m3QAhhEEIM1OJammMhssuXLzct8uC0Tp2sE/1IJBI3oWxZ4KOP\neFnnTjBbfbIFhRDjhRAzhRAdTSXC3wFwBUBfx5hoJyaRTalUCatMKdHi41sBYJGVSCRuihKDuWkT\nV0XRCVtbsksAVAVwCsAbAHYC6AMgmIh6amybNphEdt+9e4iIiECVKvVw/HgBCKEWzJRIJG5IUBD7\nAx894qgDnbBVZIOI6DUimgsO4aoBoBMRHdfeNA24f59LWRQogMVbtwIAGjUai6QkgcaN2WUjkUjc\nGKU1u26dbibYKrLJygIRpQK4QUQJ2pqkIaZWrLFKFaxZuxYAEB/fFYB0FUgkeYKephfs9es5OYkO\n2CqydYUQj0xTDIA6ynKaKraugUlkbxcogIcPkxEY+CfWrQsEIIfMSiR5ggYNuHrurVtcH+zxY2DU\nKE78beNArJxiU4IYInKNjAvZxSSyh6MJwCE8fFgD+fMDP//Mn71EInFzhODW7E8/AXPnck6DMFNW\nVg8P4P33HW6CJiFcLotJZP++XgdADZQv/xiHDwP9++trlkQicSLKa2tICAusUrTvww8BkxvRkeQJ\nkT36qBoAYPLkFFSvrqdBEonE6bRuDQSymxDNmgEnTgBffMHugoED2Y3gQGxOEOOqpBtWazRyga74\neATiCKLREKdPyyTcEkmeZN06bsV+9BGXoCIChgwBFi7kUK/z5zNNu2fvsFpNRFYIUZSIIoUQXkSU\n/OQztCedyIaHA+XKISEgAH7RkQAMiI83pC3xJZFI8ipJSVzE7733gCZNMj3MXpHVqox3cSHEOHAV\n22kaXdM+TK6Cm/mKANGeKFjwAXx9C+lslEQicRm8vTnHgYPRSmRrgyvYFtPoevZjKgt8Ppnz1pQt\nm6inNRKJJI+iVceXAHAGQLRG17Of69cBAGceBwAAataUZQ4kEonz0Upk9wGYAaCuRtezn/BwAMC/\nCU8BABo1CtTTGolEkkextSR4ZQAliGhfml3lAXxPRJc1s8xeTC3Z66gFAKhWLXeNo5BIJO6BrS3Z\n7wFkNHw22rTPdTCJbDgaAbCqBi6RSCROw9byM4eJqHEm+04RkW51BqxCuIxGwM8PSEpCfkQjXhRA\nQoIB3t56WSeRSHIrTi0/AyArx6ZfTo3QnHv/3969B0dVpnkc/z4JQghJ5wKEAEKiMKNgjbtSgCCj\nolguGIXsMnhZpZyxZsvyLlSx4s7oYpVVVnTiICgzI8VO6TgsDMsKrIrLAoIiMF5ZRy6CuiDGGCeE\nq4SE0M/+cU6aJumEdJLO6X55PlVd9Ln06edNhx8vp8953++gvp6D6d05TojCwjoLWGNMIOIN2Q9E\n5J+arhSRnwPtujdNRCaKyC4R2S0ij8TYfrWIHBKRj/zHL896UP9Uwd5wJmDnY40xwYn3OtmHgVf9\n+bwaQ3Uk0B34+3jfXETSgOeBCcA3wPsislJVdzXZ9W1VndzmA/shu0+9UbmHD7durDEmGPEOdVgF\nXCEi14D/tT28rqrr2/n+o4E9qroPQESWAFOApiEb3/mQyJUFgwD70ssYE5x23fGlqm/hze/VUQOB\n/VHLX+MFb1NjRWQbUAHMUtUdrR41ErI/BCxkjTHB6azbahPpQ2Cwqh4XkUnACvDTsyWRkPXujbCQ\nNcYEJeiQrQAGRy2f76+LUNVjUc9Xi8gCEclX1ZqmB5szZ473ZOtWxgNfMYL0dKW4uN1XXxhjzjEb\nNmxgQyfObhvoeLIikg58hvfFVyXwHnCbqu6M2qeffy4YERkN/ElVi2McK3KdbLhvX9KqqxlABb2G\n9mfPHgtZY0z7JMtQh+2iqqdE5H5gDd7lZItUdaeI3O1t1heBn4jIPXgz5dYCt7R60Npa0qqrqUf4\nlkIm/sAC1hgTHPdmRti9Gy66iC/JZQgHyc8vpqZmX9DlpZSioiL27t0bdBnGJIWU7skmRORLr0IA\namr24co/JF1FxHr/xnQW9yZSjITskIALMcYYB0M27P839ytsxkRjTPCcC9mj27cDXk924MCAizHG\nnPOcC9kTe/YA8BWDGT484GKMMec850I2vcK7lyGVQnbx4sWMGjWK7OxsBg4cSElJCZs3bwZgx44d\nTJkyhdzcXHJycpgwYQJbtmyJvHbfvn2kpaURCoUIhUJceOGFlJWVnXH84uJiMjMzCYVCZGdnEwqF\nePDBB7u0jcacs1TViQegGg5rXVqaKmgWR/R3v1P1mpi8ysvLtV+/frpixQo9fvy4NjQ06Ouvv66P\nPPKIfvHFF5qXl6ePPfaYHjx4UI8dO6bz5s3TrKws3bp1q6qq7t27V9PS0jQcDquq6gcffKC9evXS\ntWvXRt6juLhY169f3+aakv1nZkxX8v8+tD+bOvLiZHoAqlVVqqA1ZCuovvNOcgfG4cOHNSsrS5cv\nXx5z+x133KElJSXN1t9zzz169dVXq+rpkD116lRk++jRo/VXv/pVZLm4uFjXrVvX5rqS+WdmTFfr\naMg6dbrgqH/51l8pAEj60wVbtmyhrq6O0tLSmNvXrl3LtGnTmq2/+eabeffdd6mrq4us834XYOvW\nrWzfvp2hQ4cmpmhjTFycCtmafd6dXccIUVgI+fmt7y8infZojwMHDtCnTx/S0mJ/DNXV1fTv37/Z\n+v79+xMOh6mp8cbIUVX69u1LZmYm48aN495772XKlClnvKa0tJT8/Hzy8vLIz89n0aJF7arZGBMf\np+74OlFdDcBRspO+FwvQu3dvqqurCYfDMYO2T58+VFZWNltfWVlJWloaeXl5VFVVISIcOHAAgOee\ne47FixfT0NBAt26nP96VK1dyzTXXJK4xxpiYnOrJ1vs9u2NktSlkO3KepemjPcaOHUuPHj1YsWJF\nzO3XXXcdy5Yta7Z+6dKljB07loyMjDPaIiI8/PDD9OjRgwULFjRrqzGm6zkVsifjDNmghUIhnnji\nCe677z5WrlxJbW0tDQ0NvPnmm8yePZs5c+awefNmHnvsMQ4ePMixY8eYP38+r7zyCk8//XTkOE0D\ndPbs2ZSVlVFfX9/VTTLGNOFUyDYcOgSkzukCgJkzZ/Lss8/y5JNPUlBQwODBg3nhhRcoLS1lyJAh\nbNq0iW3btlFcXMyAAQN49dVXWbNmDWPGjIkco+k54ZKSEvLz81m4cGFk3U033RS5ljYUCjF16tQu\na6Mx5zKnhjrcPG0aY5ctYy4Pcft3c+nbNzJMWdDlpRT7mRlzWkeHOnSqJ3virye8P9O707dvwMUY\nYwyOhWzD4VoA6nv0CLgSY4zxOBWyHPV6sid7ZJxlR2OM6RpOhWz6cS9kGzJ6BlyJMcZ4nArZbie8\nS5bCvTIDrsQYYzxOhWz3eu9efsm2kDXGJAenQjbjpBey6TlZAVdijDEet0L2lHe6oFtedsCVGGOM\nx6mQzQx7PdmefXMCrsQYYzxOhWwvP2R79UudkC0uLqZfv37U1tZG1i1atIhrr70WgLS0NAoLCwmH\nw5HtDQ0NFBQUkJ6eHlk3fvx4evbsSSgUoqCggKlTp1JVVdV1DTHGxORUyGbhXcKVc37vgCtpOxEh\nHA4zd+7cFvfJy8tj9erVkeXVq1eT32SwXBFhwYIFHDlyhN27d3Po0CFmzJiRsLqNMW3jVMj2oIEG\n0uk9IHV6sgCzZs2ivLycI0eOxNw+ffp0Xnrppcjyyy+/zJ133tlsv8bxBnJzc5k6dSqffvppYgo2\nxrSZUyEL3jCH+b3PC7qMuIwcOZLx48fzzDPPxNxeWlrK22+/zZEjRzh06BCbNm1qNvNBtOrqapYv\nX86IESMSVbIxpo2cC9mjZJPTxo6sSOc9OuqJJ57g+eefj8xwEC0jI4PJkyezZMkSli5dyuTJk+kR\nY3yGBx54gPz8fC677DIGDhxIeXl5xwszxnSIU9PPgNeTDYWCriJ+l1xyCTfeeCNPPfUUw4YNa7Z9\n+vTpPProowCUlZXFPMb8+fO56667ElqnMSY+zvVk4wlZb0r0znl0hjlz5rBw4UIqKiqabbvyyiup\nrKzku+++Y9y4cZ3zhsaYhHOuJ3uUbLJS9IavIUOGcMsttzBv3jwuvfTSZttfe+21M5ZtYG1jkp9z\nPdnvpSctzLCdlJpOHfP4449z/PjxyPro7cOGDTvjVEL0tvZOS26MSSynpp9RYGn6P3BLw/Lo9dbj\ni5P9zIw5zaafaeJ4uo0la4xJHs6FbF03mxXBGJM8nAvZ+u4WssaY5OFeyNrUM8aYJOJcyIZ7Wsga\nY5KHcyGrvSxkjTHJw7mQJZSidyIYY5zkXMh2z+sVdAnGGBPhXsjmp+DoMMYYZzkXsj0LUmvA7gsu\nuID169efsW7jxo0MGjQostw4tUxOTg65ubmMGjWKsrIy6uvru7pcY0ycnAvZrMLcoEvoFE3HJViw\nYAGHDx+msrKS8vJylixZwg033BBghcaYtnAuZPMG5Z99pxTUOJZAz549ueqqq1i1ahVbtmzhjTfe\nCLgyY0xrnAvZ3kV9gi6hSwwaNIiRI0fyzjvvBF2KMaYVTo0nW0d3+g6M44uvzhweMIBRqwYMGEBN\nTU2Xv68xpu2c6skeI4ucnHNnXNWKiopmU4MbY5KLUyEb96wIyTb/TBz279/Phx9+yFVXXdXl722M\naTunQvYYmSk1K0Kj+vp66urqIo+TJ0+2uG9tbS0bN26ktLSUMWPGMGnSpC6s1BgTL6fOyX4vqTlu\nQUlJyRnL48aNazadzP3338+MGTMAGDp0KNOmTWPmzJldVqMxpn2cmn5mXdo4rj21qel6m0olTvYz\nM+Y0m34myvF0G7DbGJNcnArZEzb1jDEmybgVsudZyBpjkotTIWvzexljko1TIXvK5vcyxiQZt0LW\npp4xxiQZp0JWsjKDLsEYY87g1M0IaaHmIVtUVNTswn7TuqKioqBLMMYZgYesiEwE5uL1qhepalmM\nfeYBk4DvgZ+q6rZYx+qWl91s3d69ezuzXGOMiUugpwtEJA14Hvg74BLgNhG5uMk+k4AhqvoD4G7g\nty0dL6NP85B1xYYNG4IuIaFcbp/LbQP329dRQZ+THQ3sUdV9qnoSWAJMabLPFOBlAFX9M5AjIv1i\nHaxXv9Sa3yserv8iu9w+l9sG7revo4IO2YHA/qjlr/11re1TEWMfALIH5HVqccYY01FBh2yncnV+\nL2NM6gp0FC4RGQPMUdWJ/vJsQKO//BKR3wJvqepSf3kXcLWqVjU5lg0bZYxJiI6MwhX01QXvA0NF\npAioBG4FbmuyzyrgPmCpH8qHmgYsdOyHYIwxiRJoyKrqKRG5H1jD6Uu4dorI3d5mfVFV3xCRG0Tk\nc7xLuH4WZM3GGBMPZwbtNsaYZOTEF18iMlFEdonIbhF5JOh6OkJEzheR9SKyXUT+IiIP+uvzRGSN\niHwmIv8tIil9vZqIpInIRyKyyl92pn0ikiMiy0Rkp/85Xu5Y+2aIyKci8omI/FFEuqdy+0RkkYhU\nicgnUetabI+IPCoie/zP9/qzHT/lQ7YtNzSkmAZgpqpeAowF7vPbMxtYq6oXAeuBRwOssTM8BOyI\nWnapfc8Bb6jqMOBvgF040j4RGQA8AIxQ1UvxTjneRmq37/d4+REtZntEZDhwMzAM7y7UBXK2+/ZV\nNaUfwBhgddTybOCRoOvqxPatAK7D+4vaz19XCOwKurYOtOl84H+A8cAqf50T7QNCwBcx1rvSvgHA\nPiAPL2BXufD7CRQBn5zt82qaL8Bq4PLWjp3yPVnadkNDShKRYuBvga14H3gVgKp+CxQEV1mH/RqY\nBUR/IeBK+y4AqkXk9/7pkBdFJBNH2qeq3wDlwFd4NwYdVtW1ONK+KAUttKfNN0c1ciFknSQiWcB/\nAA+p6jHODCRiLKcEESkBqtQb5Ke1/2alZPvwencjgBdUdQTeFTGzcefzy8W71b0Ir1fbS0Rux5H2\ntaLd7XEhZCuAwVHL5/vrUpaIdMML2D+o6kp/dVXjmA0iUgh8F1R9HTQOmCwiXwL/DlwrIn8AvnWk\nfV8D+1X1A395OV7ouvL5XQd8qao1qnoKeBW4Anfa16il9lQAg6L2O2veuBCykRsaRKQ73g0NqwKu\nqaP+Ddihqs9FrVsF/NR/fiewsumLUoGq/ouqDlbVC/E+q/WqOh34L9xoXxWwX0R+6K+aAGzHkc8P\n7zTBGBHJ8L/wmYD3BWaqt084839WLbVnFXCrf0XFBcBQ4L1Wjxz0CedOOmk9EfgM2APMDrqeDrZl\nHHAK2AZ8DHzkty8fWOu3cw2QG3StndDWqzn9xZcz7cO7ouB9/zP8TyDHsfb9K7AT+AR4CTgvldsH\nLAa+Aerw/hH5Gd4XezHbg3elwef+z+D6sx3fbkYwxpgEcuF0gTHGJC0LWWOMSSALWWOMSSALWWOM\nSSALWWOMSSALWWOMSSALWWOMSSALWWOMSSALWZMU/IGu74la3hRwPRkisuGsY4XGfu15IrLRH+vY\nnOPsl8Akizzg3sYFVf1xot9QRC4WkZYGl74LWK7tuCVSVU/i3ZJ5a0fqM26wkDXJ4ilgiD8G69Mi\nchTAH/hnpz8+62ci8oqITBCRTf7yyMYDiMjtIvJn/xi/aUMv9Bq88SFiuR1/UJCoGl4RkR0i8ie/\np5spIq+JyMf+VCzTol6/0j+GOcdZyJpkMRv4XFVHqOo/c+b4nUOAZ9SbCuRi4Da/pzsL+AV4vVLg\nFuAK9cZxDdNKyInIRODnwKDGIe2itp0HXKCqX0Wtvgh4XlWHA0fwpqmfCFSo6mXqTcXyZtT+nwKj\n4v0hGPdYyJpU8H+q2jgf2HZgnf/8L3iDR4M35N4I4H0R+Ri4FriwpQOq6pt4AblQ/RHwo/QBDjVZ\n95WqbvWf/xFvtLRPgOtF5CkR+bGqHo06fhioE5Fe8TTUuKdb0AUY0wZ1Uc/DUcthTv8OC/CSqv6i\nLQf0e6/ftrC5Fsg4yyFUVT8XkcuAG4AnRWStqj4ZtU8P4ERb6jHusp6sSRZHgeyoZWnheVON29YB\nPxGRvhCZ0nlwyy9jNPCeiIwUkZ7RG1T1EJDuDwLfaLCIXO4//0dgk4j0B2pVdTHwDF5PGv/984Fq\n9WYPMOcwC1mTFFS1Btjsf4H0NGeek23peWRZVXcCvwTWiMj/4g20XNjKW36DNwFetqrWxti+Boi+\nwuEzvOnZdwC5wG+AH+EF9cfA40B0L/Ya4PVW3t+cI2zQbmNi8E8DPKyqd4pIEfCaqv4ojtcvx5s6\n+vOEFWlSgvVkjYlBVT8G3oq6DKzNvRH/6oRXLWANWE/WGGMSynqyxhiTQBayxhiTQBayxhiTQBay\nxhiTQBayxhiTQBayxhiTQBayxhiTQBayxhiTQP8Pb7UIrneJdOwAAAAASUVORK5CYII=\n",
      "text/plain": [
       "<matplotlib.figure.Figure at 0x7f54a6513a50>"
      ]
     },
     "metadata": {},
     "output_type": "display_data"
    }
   ],
   "source": [
    "import numpy as np\n",
    "from MDAnalysis.analysis.align import rotation_matrix\n",
    "from MDAnalysis.core.qcprot import CalcRMSDRotationalMatrix\n",
    "\n",
    "def superpose(mobile, xref0, xref_com=None):\n",
    "    \"\"\"Superpose the AtomGroup *mobile* onto the coordinates *xref0* centered at the orgin.\n",
    "\n",
    "    The original center of mass of the reference group *xref_com* must\n",
    "    be supplied or the superposition is done at the origin of the\n",
    "    coordinate system.\n",
    "    \"\"\"\n",
    "    # 995 us\n",
    "    xref_com = xref_com if xref_com is not None else np.array([0., 0., 0.])\n",
    "    xmobile0 = mobile.positions - mobile.center_of_mass()\n",
    "    R, rmsd = rotation_matrix(xmobile0, xref0)\n",
    "    mobile.rotate(R)\n",
    "    mobile.translate(xref_com)\n",
    "    return rmsd\n",
    "\n",
    "def rmsd(mobile, xref0):\n",
    "    \"\"\"Calculate optimal RMSD for AtomGroup *mobile* onto the coordinates *xref0* centered at the orgin.\n",
    "\n",
    "    The coordinates are not changed. No mass weighting.\n",
    "    \"\"\"\n",
    "    # 738 us\n",
    "    xmobile0 = mobile.positions - mobile.center_of_mass()\n",
    "    return CalcRMSDRotationalMatrix(xref0.T.astype(np.float64), xmobile0.T.astype(np.float64), mobile.n_atoms, None, None)\n",
    "\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    import MDAnalysis\n",
    "    import matplotlib\n",
    "    import matplotlib.pyplot as plt\n",
    "\n",
    "    # load AdK DIMS trajectory\n",
    "    from MDAnalysis.tests.datafiles import PSF, DCD\n",
    "    u = MDAnalysis.Universe(PSF, DCD)\n",
    "\n",
    "    # one AtomGroup per domain\n",
    "    domains = {\n",
    "        'CORE': u.select_atoms(\"(resid 1:29 or resid 60:121 or resid 160:214) and name CA\"),\n",
    "        'LID': u.select_atoms(\"resid 122-159 and name CA\"),\n",
    "        'NMP': u.select_atoms(\"resid 30-59 and name CA\"),\n",
    "        }\n",
    "    colors = {'CORE': 'black', 'NMP': 'blue', 'LID': 'red'}\n",
    "\n",
    "    u.trajectory[0]   # rewind trajectory\n",
    "    xref0 = dict((name, g.positions - g.center_of_mass()) for name,g in domains.iteritems())\n",
    "\n",
    "    nframes = len(u.trajectory)\n",
    "    results = dict((name, np.zeros((nframes, 2), dtype=np.float64)) for name in domains)\n",
    "\n",
    "    for iframe,ts in enumerate(u.trajectory):\n",
    "        for name, g in domains.iteritems():\n",
    "            results[name][iframe, :] = u.trajectory.time, rmsd(g, xref0[name])\n",
    "\n",
    "\n",
    "    # plot\n",
    "    fig = plt.figure(figsize=(5,5))\n",
    "    ax = fig.add_subplot(111)\n",
    "    for name in \"CORE\", \"NMP\", \"LID\":\n",
    "        data = results[name]\n",
    "        ax.plot(data[:,0], data[:,1], linestyle=\"-\", color=colors[name], lw=2, label=name)\n",
    "    ax.legend(loc=\"best\")\n",
    "    ax.set_xlabel(r\"time  $t$ (ps)\")\n",
    "    ax.set_ylabel(r\"C$_\\alpha$ RMSD from $t=0$, $\\rho_{\\mathrm{C}_\\alpha}$ ($\\AA$)\")\n",
    "\n",
    "    for ext in ('svg', 'pdf', 'png'):\n",
    "        fig.savefig(\"AdK_domain_rigidity.{0}\".format(ext))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "collapsed": true
   },
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python2",
   "language": "python",
   "name": "python2"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 2
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython2",
   "version": "2.7.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 0
}
